{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Alteration of intercellular signaling in ulcerative colitis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Analysising single cell RNA-seq data for inter and intracellular interactions. \n",
    "The first step is to import the singel cell RNA-seq data file. The import file is the cell cluster averaged gene expression files. OmniPath is accesed by the python client."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import igraph as ig\n",
    "import os\n",
    "from itertools import permutations\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import omnipath as op"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First lets check the OmniPAth version and data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.16.4\n",
      "Options(url='https://omnipathdb.org', fallback_urls=('http://no-tls.omnipathdb.org',), static_url='http://no-tls.static.omnipathdb.org/resources', license=None, cache=<FileCache[size=0, path='/Users/maya/.cache/omnipathdb']>, autoload=True, convert_dtypes=True, num_retries=3, timeout=600, chunk_size=8196)\n"
     ]
    }
   ],
   "source": [
    "print(op.__server_version__)\n",
    "print(op.options)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can donwload OmniPath intercellular network and see what kind of columns are in the intercellular dataframe. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "intercell_network = op.interactions.import_intercell_network()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>is_stimulation</th>\n",
       "      <th>is_inhibition</th>\n",
       "      <th>consensus_direction</th>\n",
       "      <th>consensus_stimulation</th>\n",
       "      <th>consensus_inhibition</th>\n",
       "      <th>dip_url</th>\n",
       "      <th>curation_effort</th>\n",
       "      <th>...</th>\n",
       "      <th>category_source_intercell_target</th>\n",
       "      <th>uniprot_intercell_target</th>\n",
       "      <th>genesymbol_intercell_target</th>\n",
       "      <th>entity_type_intercell_target</th>\n",
       "      <th>consensus_score_intercell_target</th>\n",
       "      <th>transmitter_intercell_target</th>\n",
       "      <th>receiver_intercell_target</th>\n",
       "      <th>secreted_intercell_target</th>\n",
       "      <th>plasma_membrane_transmembrane_intercell_target</th>\n",
       "      <th>plasma_membrane_peripheral_intercell_target</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>P14416</td>\n",
       "      <td>P48995</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>nan</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>resource_specific</td>\n",
       "      <td>P48995</td>\n",
       "      <td>TRPC1</td>\n",
       "      <td>protein</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>P14416</td>\n",
       "      <td>P48995</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>nan</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>resource_specific</td>\n",
       "      <td>P48995</td>\n",
       "      <td>TRPC1</td>\n",
       "      <td>protein</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>P14416</td>\n",
       "      <td>P48995</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>nan</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>composite</td>\n",
       "      <td>P48995</td>\n",
       "      <td>TRPC1</td>\n",
       "      <td>protein</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7</td>\n",
       "      <td>Q13255</td>\n",
       "      <td>P48995</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>nan</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>resource_specific</td>\n",
       "      <td>P48995</td>\n",
       "      <td>TRPC1</td>\n",
       "      <td>protein</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9</td>\n",
       "      <td>Q13255</td>\n",
       "      <td>P48995</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>nan</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>resource_specific</td>\n",
       "      <td>P48995</td>\n",
       "      <td>TRPC1</td>\n",
       "      <td>protein</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   index  source  target  is_stimulation  is_inhibition  consensus_direction  \\\n",
       "0      0  P14416  P48995            True          False                 True   \n",
       "1      2  P14416  P48995            True          False                 True   \n",
       "2      5  P14416  P48995            True          False                 True   \n",
       "3      7  Q13255  P48995            True          False                 True   \n",
       "4      9  Q13255  P48995            True          False                 True   \n",
       "\n",
       "   consensus_stimulation  consensus_inhibition dip_url  curation_effort  ...  \\\n",
       "0                   True                 False     nan                1  ...   \n",
       "1                   True                 False     nan                1  ...   \n",
       "2                   True                 False     nan                1  ...   \n",
       "3                   True                 False     nan                1  ...   \n",
       "4                   True                 False     nan                1  ...   \n",
       "\n",
       "  category_source_intercell_target uniprot_intercell_target  \\\n",
       "0                resource_specific                   P48995   \n",
       "1                resource_specific                   P48995   \n",
       "2                        composite                   P48995   \n",
       "3                resource_specific                   P48995   \n",
       "4                resource_specific                   P48995   \n",
       "\n",
       "  genesymbol_intercell_target entity_type_intercell_target  \\\n",
       "0                       TRPC1                      protein   \n",
       "1                       TRPC1                      protein   \n",
       "2                       TRPC1                      protein   \n",
       "3                       TRPC1                      protein   \n",
       "4                       TRPC1                      protein   \n",
       "\n",
       "   consensus_score_intercell_target  transmitter_intercell_target  \\\n",
       "0                                 1                         False   \n",
       "1                                 3                         False   \n",
       "2                                 3                         False   \n",
       "3                                 1                         False   \n",
       "4                                 3                         False   \n",
       "\n",
       "  receiver_intercell_target secreted_intercell_target  \\\n",
       "0                      True                     False   \n",
       "1                      True                     False   \n",
       "2                      True                     False   \n",
       "3                      True                     False   \n",
       "4                      True                     False   \n",
       "\n",
       "  plasma_membrane_transmembrane_intercell_target  \\\n",
       "0                                          False   \n",
       "1                                          False   \n",
       "2                                          False   \n",
       "3                                          False   \n",
       "4                                          False   \n",
       "\n",
       "  plasma_membrane_peripheral_intercell_target  \n",
       "0                                       False  \n",
       "1                                       False  \n",
       "2                                       False  \n",
       "3                                       False  \n",
       "4                                       False  \n",
       "\n",
       "[5 rows x 46 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intercell_network.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['index', 'source', 'target', 'is_stimulation', 'is_inhibition',\n",
       "       'consensus_direction', 'consensus_stimulation', 'consensus_inhibition',\n",
       "       'dip_url', 'curation_effort', 'references', 'sources',\n",
       "       'references_stripped', 'n_references', 'n_sources', 'n_primary_sources',\n",
       "       'category_intercell_source', 'parent_intercell_source',\n",
       "       'database_intercell_source', 'scope_intercell_source',\n",
       "       'aspect_intercell_source', 'category_source_intercell_source',\n",
       "       'uniprot_intercell_source', 'genesymbol_intercell_source',\n",
       "       'entity_type_intercell_source', 'consensus_score_intercell_source',\n",
       "       'transmitter_intercell_source', 'receiver_intercell_source',\n",
       "       'secreted_intercell_source',\n",
       "       'plasma_membrane_transmembrane_intercell_source',\n",
       "       'plasma_membrane_peripheral_intercell_source',\n",
       "       'category_intercell_target', 'parent_intercell_target',\n",
       "       'database_intercell_target', 'scope_intercell_target',\n",
       "       'aspect_intercell_target', 'category_source_intercell_target',\n",
       "       'uniprot_intercell_target', 'genesymbol_intercell_target',\n",
       "       'entity_type_intercell_target', 'consensus_score_intercell_target',\n",
       "       'transmitter_intercell_target', 'receiver_intercell_target',\n",
       "       'secreted_intercell_target',\n",
       "       'plasma_membrane_transmembrane_intercell_target',\n",
       "       'plasma_membrane_peripheral_intercell_target'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intercell_network.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(30265, 46)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intercell_network.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we can check what are the categories what we can use for intercellular interactions. We can use the intercellular interctions directly, hovewer it contain porteins which are invovled in the adhesion process and binds to the intracelluar domains of receptor proteins. That is why we suggest the fitlering as in the paper.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'desmosome', 'cell_surface_ligand', 'cell_surface_enzyme', 'secreted_receptor', 'ecm', 'receptor_regulator', 'adhesion', 'tight_junction', 'cell_adhesion', 'matrix_adhesion_regulator', 'cell_surface_peptidase', 'ecm_regulator', 'secreted_enzyme', 'ligand_regulator', 'gap_junction', 'ligand'}\n"
     ]
    }
   ],
   "source": [
    "print (set(intercell_network[\"category_intercell_source\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'adherens_junction', 'receptor', 'ion_channel_regulator', 'desmosome', 'adhesion', 'tight_junction', 'cell_adhesion', 'matrix_adhesion', 'transporter', 'ion_channel', 'gap_junction'}\n"
     ]
    }
   ],
   "source": [
    "print (set(intercell_network[\"category_intercell_target\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_source_types = [\"cell_surface_enzyme\",\"cell_surface_ligand\",\"ligand\",\"secreted_enzyme\",\"secreted_receptor\",\n",
    "           \"adhesion\",\"cell_adhesion\",\"tight_junction\",\"cell_surface_peptidase\",\"gap_junction\",\"desmosome\"]\n",
    "filtered_target_types = [\"adhesion\",\"receptor\",\"cell_adhesion\",\"tight_junction\",\"gap_junction\",\"ion_channel\",\"transporter\",\n",
    "                \"adherens_junction\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20784, 46)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_intercell_network = intercell_network[intercell_network.category_intercell_source.isin(filtered_source_types)] \n",
    "filtered_intercell_network = filtered_intercell_network[filtered_intercell_network.\n",
    "                                                        category_intercell_target.isin(filtered_target_types)]\n",
    "filtered_intercell_network.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The next step is to preprocess the data files. The imput file is the cell type specfic expression. The data are from Smillie et al 2019 (https://pubmed.ncbi.nlm.nih.gov/31348891/). Each column is the average per cell type in transcript per million. From that we will build a simple cell-cell interaction count. It is based on expression tresholding. Our assumptation was that the interactions are important if they are exisit in any way, so we have chosen a realtievely low expression treshold. We invite the user for testing various tresholds in their own work. \n",
    "You will need for this the expression data which are in this example's folder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cell_imm = pd.read_csv(\"imm_all_expression_condition.tsv\", sep=\"\\t\", index_col=0, header=0)\n",
    "df_cell_fib = pd.read_csv(\"fib_all_expression_condition.tsv\",sep=\"\\t\", index_col=0, header=0)\n",
    "df_cell_epi =  pd.read_csv(\"epi_all_expression_condition.tsv\",sep=\"\\t\", index_col=0, header=0)\n",
    "df_cell_exp = df_cell_imm.join(df_cell_fib, how=\"outer\") #Keeping all genes\n",
    "df_cell_exp = df_cell_exp.join(df_cell_epi, how=\"outer\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RNA.CD8..IELs_Healthy</th>\n",
       "      <th>RNA.CD8..LP_Healthy</th>\n",
       "      <th>RNA.CD4..Memory_Healthy</th>\n",
       "      <th>RNA.MT.hi_Healthy</th>\n",
       "      <th>RNA.Cycling.T_Healthy</th>\n",
       "      <th>RNA.NKs_Healthy</th>\n",
       "      <th>RNA.CD4..Activated.Fos.lo_Healthy</th>\n",
       "      <th>RNA.CD4..Activated.Fos.hi_Healthy</th>\n",
       "      <th>RNA.CD8..IELs_Uninflamed</th>\n",
       "      <th>RNA.MT.hi_Uninflamed</th>\n",
       "      <th>...</th>\n",
       "      <th>RNA.Immature.Goblet_Inflamed</th>\n",
       "      <th>RNA.Stem_Uninflamed</th>\n",
       "      <th>RNA.Immature.Enterocytes.2_Inflamed</th>\n",
       "      <th>RNA.Goblet_Inflamed</th>\n",
       "      <th>RNA.Tuft_Inflamed</th>\n",
       "      <th>RNA.Enterocytes_Inflamed</th>\n",
       "      <th>RNA.Best4..Enterocytes_Inflamed</th>\n",
       "      <th>RNA.Enteroendocrine_Inflamed</th>\n",
       "      <th>RNA.M.cells_Inflamed</th>\n",
       "      <th>RNA.M.cells_Uninflamed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7SK</th>\n",
       "      <td>0.030182</td>\n",
       "      <td>0.042158</td>\n",
       "      <td>0.016792</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.228991</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015092</td>\n",
       "      <td>0.030573</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.001679</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.005192</td>\n",
       "      <td>0.002465</td>\n",
       "      <td>0.007913</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.062424</td>\n",
       "      <td>0.025848</td>\n",
       "      <td>0.049498</td>\n",
       "      <td>0.208431</td>\n",
       "      <td>0.321939</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.045158</td>\n",
       "      <td>0.018531</td>\n",
       "      <td>0.053037</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.013224</td>\n",
       "      <td>0.002473</td>\n",
       "      <td>0.089265</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>0.276995</td>\n",
       "      <td>0.198592</td>\n",
       "      <td>0.138406</td>\n",
       "      <td>0.295042</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.498747</td>\n",
       "      <td>0.130186</td>\n",
       "      <td>0.256156</td>\n",
       "      <td>0.305482</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>0.002978</td>\n",
       "      <td>0.005669</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.025840</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.079150</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.66123</td>\n",
       "      <td>0.211192</td>\n",
       "      <td>0.848589</td>\n",
       "      <td>0.187466</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.486224</td>\n",
       "      <td>0.778587</td>\n",
       "      <td>2.653654</td>\n",
       "      <td>0.457293</td>\n",
       "      <td>0.737544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>0.023545</td>\n",
       "      <td>0.214442</td>\n",
       "      <td>0.202609</td>\n",
       "      <td>0.167921</td>\n",
       "      <td>0.097767</td>\n",
       "      <td>0.545677</td>\n",
       "      <td>0.144894</td>\n",
       "      <td>0.166956</td>\n",
       "      <td>0.010304</td>\n",
       "      <td>0.243075</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 153 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          RNA.CD8..IELs_Healthy  RNA.CD8..LP_Healthy  RNA.CD4..Memory_Healthy  \\\n",
       "7SK                    0.030182             0.042158                 0.016792   \n",
       "A1BG                   0.062424             0.025848                 0.049498   \n",
       "A1BG-AS1               0.276995             0.198592                 0.138406   \n",
       "A1CF                   0.002978             0.005669                 0.000000   \n",
       "A2M                    0.023545             0.214442                 0.202609   \n",
       "\n",
       "          RNA.MT.hi_Healthy  RNA.Cycling.T_Healthy  RNA.NKs_Healthy  \\\n",
       "7SK                0.000000               0.000000         0.228991   \n",
       "A1BG               0.208431               0.321939         0.000000   \n",
       "A1BG-AS1           0.295042               0.000000         0.498747   \n",
       "A1CF               0.000000               0.025840         0.000000   \n",
       "A2M                0.167921               0.097767         0.545677   \n",
       "\n",
       "          RNA.CD4..Activated.Fos.lo_Healthy  \\\n",
       "7SK                                0.000000   \n",
       "A1BG                               0.045158   \n",
       "A1BG-AS1                           0.130186   \n",
       "A1CF                               0.000000   \n",
       "A2M                                0.144894   \n",
       "\n",
       "          RNA.CD4..Activated.Fos.hi_Healthy  RNA.CD8..IELs_Uninflamed  \\\n",
       "7SK                                0.015092                  0.030573   \n",
       "A1BG                               0.018531                  0.053037   \n",
       "A1BG-AS1                           0.256156                  0.305482   \n",
       "A1CF                               0.000000                  0.079150   \n",
       "A2M                                0.166956                  0.010304   \n",
       "\n",
       "          RNA.MT.hi_Uninflamed  ...  RNA.Immature.Goblet_Inflamed  \\\n",
       "7SK                   0.000000  ...                       0.00000   \n",
       "A1BG                  0.000000  ...                       0.00000   \n",
       "A1BG-AS1              0.000000  ...                           NaN   \n",
       "A1CF                  0.000000  ...                       0.66123   \n",
       "A2M                   0.243075  ...                       0.00000   \n",
       "\n",
       "          RNA.Stem_Uninflamed  RNA.Immature.Enterocytes.2_Inflamed  \\\n",
       "7SK                  0.001679                             0.000000   \n",
       "A1BG                 0.013224                             0.002473   \n",
       "A1BG-AS1                  NaN                                  NaN   \n",
       "A1CF                 0.211192                             0.848589   \n",
       "A2M                  0.000000                             0.000000   \n",
       "\n",
       "          RNA.Goblet_Inflamed  RNA.Tuft_Inflamed  RNA.Enterocytes_Inflamed  \\\n",
       "7SK                  0.000000           7.005192                  0.002465   \n",
       "A1BG                 0.089265           0.000000                  0.000000   \n",
       "A1BG-AS1                  NaN                NaN                       NaN   \n",
       "A1CF                 0.187466           0.000000                  0.486224   \n",
       "A2M                  0.000000           0.000000                  0.000000   \n",
       "\n",
       "          RNA.Best4..Enterocytes_Inflamed  RNA.Enteroendocrine_Inflamed  \\\n",
       "7SK                              0.007913                      0.000000   \n",
       "A1BG                             0.000000                      0.000000   \n",
       "A1BG-AS1                              NaN                           NaN   \n",
       "A1CF                             0.778587                      2.653654   \n",
       "A2M                              0.000000                      0.000000   \n",
       "\n",
       "          RNA.M.cells_Inflamed  RNA.M.cells_Uninflamed  \n",
       "7SK                   0.000000                0.000000  \n",
       "A1BG                  0.000000                0.000000  \n",
       "A1BG-AS1                   NaN                     NaN  \n",
       "A1CF                  0.457293                0.737544  \n",
       "A2M                   0.000000                0.000000  \n",
       "\n",
       "[5 rows x 153 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_cell_exp.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will choose the mean -2SD of the whole data set. First we do a log2 based transformation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cell_exp[df_cell_exp.values == 0] = \"NaN\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RNA.CD8..IELs_Healthy</th>\n",
       "      <th>RNA.CD8..LP_Healthy</th>\n",
       "      <th>RNA.CD4..Memory_Healthy</th>\n",
       "      <th>RNA.MT.hi_Healthy</th>\n",
       "      <th>RNA.Cycling.T_Healthy</th>\n",
       "      <th>RNA.NKs_Healthy</th>\n",
       "      <th>RNA.CD4..Activated.Fos.lo_Healthy</th>\n",
       "      <th>RNA.CD4..Activated.Fos.hi_Healthy</th>\n",
       "      <th>RNA.CD8..IELs_Uninflamed</th>\n",
       "      <th>RNA.MT.hi_Uninflamed</th>\n",
       "      <th>...</th>\n",
       "      <th>RNA.Immature.Goblet_Inflamed</th>\n",
       "      <th>RNA.Stem_Uninflamed</th>\n",
       "      <th>RNA.Immature.Enterocytes.2_Inflamed</th>\n",
       "      <th>RNA.Goblet_Inflamed</th>\n",
       "      <th>RNA.Tuft_Inflamed</th>\n",
       "      <th>RNA.Enterocytes_Inflamed</th>\n",
       "      <th>RNA.Best4..Enterocytes_Inflamed</th>\n",
       "      <th>RNA.Enteroendocrine_Inflamed</th>\n",
       "      <th>RNA.M.cells_Inflamed</th>\n",
       "      <th>RNA.M.cells_Uninflamed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7SK</th>\n",
       "      <td>0.0301819</td>\n",
       "      <td>0.0421578</td>\n",
       "      <td>0.0167924</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.228991</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0150918</td>\n",
       "      <td>0.0305725</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.00167886</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.00519</td>\n",
       "      <td>0.00246488</td>\n",
       "      <td>0.00791332</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.062424</td>\n",
       "      <td>0.0258481</td>\n",
       "      <td>0.0494975</td>\n",
       "      <td>0.208431</td>\n",
       "      <td>0.321939</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0451578</td>\n",
       "      <td>0.0185314</td>\n",
       "      <td>0.0530365</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0132239</td>\n",
       "      <td>0.0024732</td>\n",
       "      <td>0.089265</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>0.276995</td>\n",
       "      <td>0.198592</td>\n",
       "      <td>0.138406</td>\n",
       "      <td>0.295042</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.498747</td>\n",
       "      <td>0.130186</td>\n",
       "      <td>0.256156</td>\n",
       "      <td>0.305482</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>0.00297762</td>\n",
       "      <td>0.00566872</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0258403</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0791505</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.66123</td>\n",
       "      <td>0.211192</td>\n",
       "      <td>0.848589</td>\n",
       "      <td>0.187466</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.486224</td>\n",
       "      <td>0.778587</td>\n",
       "      <td>2.65365</td>\n",
       "      <td>0.457293</td>\n",
       "      <td>0.737544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>0.0235452</td>\n",
       "      <td>0.214442</td>\n",
       "      <td>0.202609</td>\n",
       "      <td>0.167921</td>\n",
       "      <td>0.0977667</td>\n",
       "      <td>0.545677</td>\n",
       "      <td>0.144894</td>\n",
       "      <td>0.166956</td>\n",
       "      <td>0.0103037</td>\n",
       "      <td>0.243075</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 153 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         RNA.CD8..IELs_Healthy RNA.CD8..LP_Healthy RNA.CD4..Memory_Healthy  \\\n",
       "7SK                  0.0301819           0.0421578               0.0167924   \n",
       "A1BG                  0.062424           0.0258481               0.0494975   \n",
       "A1BG-AS1              0.276995            0.198592                0.138406   \n",
       "A1CF                0.00297762          0.00566872                     NaN   \n",
       "A2M                  0.0235452            0.214442                0.202609   \n",
       "\n",
       "         RNA.MT.hi_Healthy RNA.Cycling.T_Healthy RNA.NKs_Healthy  \\\n",
       "7SK                    NaN                   NaN        0.228991   \n",
       "A1BG              0.208431              0.321939             NaN   \n",
       "A1BG-AS1          0.295042                   NaN        0.498747   \n",
       "A1CF                   NaN             0.0258403             NaN   \n",
       "A2M               0.167921             0.0977667        0.545677   \n",
       "\n",
       "         RNA.CD4..Activated.Fos.lo_Healthy RNA.CD4..Activated.Fos.hi_Healthy  \\\n",
       "7SK                                    NaN                         0.0150918   \n",
       "A1BG                             0.0451578                         0.0185314   \n",
       "A1BG-AS1                          0.130186                          0.256156   \n",
       "A1CF                                   NaN                               NaN   \n",
       "A2M                               0.144894                          0.166956   \n",
       "\n",
       "         RNA.CD8..IELs_Uninflamed RNA.MT.hi_Uninflamed  ...  \\\n",
       "7SK                     0.0305725                  NaN  ...   \n",
       "A1BG                    0.0530365                  NaN  ...   \n",
       "A1BG-AS1                 0.305482                  NaN  ...   \n",
       "A1CF                    0.0791505                  NaN  ...   \n",
       "A2M                     0.0103037             0.243075  ...   \n",
       "\n",
       "         RNA.Immature.Goblet_Inflamed RNA.Stem_Uninflamed  \\\n",
       "7SK                               NaN          0.00167886   \n",
       "A1BG                              NaN           0.0132239   \n",
       "A1BG-AS1                          NaN                 NaN   \n",
       "A1CF                          0.66123            0.211192   \n",
       "A2M                               NaN                 NaN   \n",
       "\n",
       "         RNA.Immature.Enterocytes.2_Inflamed RNA.Goblet_Inflamed  \\\n",
       "7SK                                      NaN                 NaN   \n",
       "A1BG                               0.0024732            0.089265   \n",
       "A1BG-AS1                                 NaN                 NaN   \n",
       "A1CF                                0.848589            0.187466   \n",
       "A2M                                      NaN                 NaN   \n",
       "\n",
       "         RNA.Tuft_Inflamed RNA.Enterocytes_Inflamed  \\\n",
       "7SK                7.00519               0.00246488   \n",
       "A1BG                   NaN                      NaN   \n",
       "A1BG-AS1               NaN                      NaN   \n",
       "A1CF                   NaN                 0.486224   \n",
       "A2M                    NaN                      NaN   \n",
       "\n",
       "         RNA.Best4..Enterocytes_Inflamed RNA.Enteroendocrine_Inflamed  \\\n",
       "7SK                           0.00791332                          NaN   \n",
       "A1BG                                 NaN                          NaN   \n",
       "A1BG-AS1                             NaN                          NaN   \n",
       "A1CF                            0.778587                      2.65365   \n",
       "A2M                                  NaN                          NaN   \n",
       "\n",
       "         RNA.M.cells_Inflamed RNA.M.cells_Uninflamed  \n",
       "7SK                       NaN                    NaN  \n",
       "A1BG                      NaN                    NaN  \n",
       "A1BG-AS1                  NaN                    NaN  \n",
       "A1CF                 0.457293               0.737544  \n",
       "A2M                       NaN                    NaN  \n",
       "\n",
       "[5 rows x 153 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_cell_exp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_matrix = df_cell_exp.to_numpy()\n",
    "data_matrix_log2 = np.log2(data_matrix.astype(float)) #We need to change the data file to floats\n",
    "mean_cell = np.nanmean(data_matrix_log2)\n",
    "std = np.nanstd(data_matrix_log2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(-2.160708171397226, 3.00937762482671)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mean_cell, std"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can check the distribuiton of the cell types. Here only for one."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[-5.0501723 , -4.56805614, -5.89604351, ...,         nan,\n",
       "                nan,         nan],\n",
       "       [-4.00175632, -5.27379813, -4.33649916, ...,         nan,\n",
       "                nan,         nan],\n",
       "       [-1.85206773, -2.33211901, -2.8530207 , ...,         nan,\n",
       "                nan,         nan],\n",
       "       ...,\n",
       "       [-3.47290816, -4.19215007, -4.40580567, ..., -2.06536459,\n",
       "         0.35303725,  0.44592408],\n",
       "       [-4.38586399, -2.27399124, -3.85313444, ..., -4.58601467,\n",
       "        -8.31527547,         nan],\n",
       "       [-2.5735855 , -1.76640068, -1.1066736 , ..., -4.14328163,\n",
       "        -7.335609  ,         nan]])"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_matrix_log2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<seaborn.axisgrid.FacetGrid at 0x19ee52d6c88>"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAWAAAAFgCAYAAACFYaNMAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/Il7ecAAAACXBIWXMAAAsTAAALEwEAmpwYAAAXKklEQVR4nO3df6zdd33f8ee7TgyMGjsxnu1eB9kGq11gIk5cHAqq2mRUTrbO6dSmoIpYyJ2nzVh0TO3S7Q9WqZNotS2BKAryCMOZKDTLiOyiKDQ1PyeNDKfHCgfSKreWk/jiX4HgpmTZxfF7f9zPvT45vrGP7fs9n/Pj+ZCu7vf7Od9779vX4cXX7/P5fj6RmUiS+u+nahcgSePKAJakSgxgSarEAJakSgxgSarkitoFXI4tW7bko48+WrsMSbqQmG9wqO+An3/++dolSNIlG+oAlqRhZgBLUiUGsCRVYgBLUiUGsCRVYgBLUiUGsCRVYgBLUiUGsCRVYgBLUiUGsCRVYgBLUiUGsCRVMtTLUWqwTE9P02q1XjW2ceNGFi9eXKkiabAZwFowrVaLXffuZenEegBOTR3inp2wefPmypVJg8kA1oJaOrGe5eveDsCZV07TbrfPuca7YmmGAazGvHj8We4+/DKrJnNuzLti6SwDWJesu+fbbrfJM/mqa5asWjt3Ryzp1QxgXbLunu/UwW+ybMMNlauShocBrMvS2fM99f1DlauRhovzgCWpEgNYkiqxBaGe9fKmm6TeGcDqmW+6SQvLANZF8U03aeHYA5akSgxgSarEAJakSgxgSarEAJakSgxgSarEAJakShoN4IhYFhEPRcRfRcRTEfHuiLg6Ih6LiKfL56vKtRERn4yIyYh4MiKub7I2Saqt6TvgTwCPZubPAe8EngLuBPZn5gZgfzkHuAXYUD52APc1XJskVdVYAEfEUuAXgfsBMnM6M38EbAX2lMv2ALeV463AAznjW8CyiFjdVH2SVFuTjyKvA04C/y0i3gk8AXwEWJmZR8s1x4CV5XgCeK7j64+UsaNoZMy3T5x7xGlcNRnAVwDXA7sy8/GI+ARn2w0AZGZGxEUtpxURO5hpUfCWt7xloWpVn3TvE+cecRpnTfaAjwBHMvPxcv4QM4F8fLa1UD6fKK9PAdd0fP2aMvYqmbk7Mzdl5qYVK1Y0VryaM7tP3PJ1b59bWU0aR40FcGYeA56LiJ8tQzcD3wP2AdvK2DZgbzneB9xRZkPcCJzqaFVI0shpejnKXcDnImIxcAj4EDOh/2BEbAeeAW4v1z4C3ApMAi+VayVpZDUawJl5ENg0z0s3z3NtAjubrEeSBolPwklSJQawJFViAEtSJQawJFViAEtSJQawJFViAEtSJQawJFViAEtSJQawJFXS9FoQ0nm5PrDGmQGsqlwfWOPMANZrmp6eptVqzZ23223yzEWtn9+T2fWBpXFjAOs1tVotdt27d27R9KmD32TZhhsqVyWNDgNY57V0Yv3c3emp7x+qXI00WpwFIUmVGMCSVIkBLEmVGMCSVIkBLEmVGMCSVInT0MZE90MVMJiP/PpossaJATwmuh+qGNRHfn00WePEAB4jnQ9VDDIfTda4sAcsSZUYwJJUiQEsSZXYAx5T8802AGccSP1kAI+p7tkG4IwDqd8M4DHWPdug+664qQXYJc0wgDWn+67YBdilZhnAepXOu2IXYJea5SwISarEAJakSgxgSarEAJakSgxgSarEAJakSgxgSaqk0QCOiMMR8Z2IOBgRB8rY1RHxWEQ8XT5fVcYjIj4ZEZMR8WREXN9kbZJUWz/ugH85M6/LzE3l/E5gf2ZuAPaXc4BbgA3lYwdwXx9qk6RqarQgtgJ7yvEe4LaO8QdyxreAZRGxukJ9ktQXTQdwAn8eEU9ExI4ytjIzj5bjY8DKcjwBPNfxtUfK2KtExI6IOBARB06ePNlU3ZLUuKbXgnhvZk5FxN8HHouIv+p8MTMzIi5qua3M3A3sBti0aZNLdUkaWo3eAWfmVPl8AngYeBdwfLa1UD6fKJdPAdd0fPmaMiZJI6mxAI6IN0bEktlj4FeANrAP2FYu2wbsLcf7gDvKbIgbgVMdrQpJGjlNtiBWAg9HxOzP+ZPMfDQivg08GBHbgWeA28v1jwC3ApPAS8CHGqxNkqprLIAz8xDwznnGfwDcPM94AjubqkeSBo1PwklSJe6IMaKmp6dptVpz5+7vJg0eA3hEtVotdt27l6UT6wH3d5MGkQE8wpZOrHd/N2mA2QOWpEoMYEmqxACWpEoMYEmqxACWpEoMYEmqxACWpEoMYEmqxACWpEoMYEmqxACWpEoMYEmqxMV4hlT3cpMAGzduZPHixZUqknSxDOAh1b3c5KmpQ9yzEzZv3ly5Mkm9MoCHWOdyk5KGjz1gSarEAJakSmxBjIgzr5ym3W7PnbsHnDT4DOAR8eLxZ7n78MusmpwJXfeAkwafATxClqxa6x5w0hAxgDXQulsr4HxnjQ4DWAOtu7XifGeNEgNYA6+ztSKNEqehSVIlBrAkVWIAS1Il9oA1VOabFQHOjNBwMoA1VLpnRYAzIzS8DGANHWdFaFTYA5akSgxgSarEAJakSgxgSarEAJakShoP4IhYFBGtiPhSOV8XEY9HxGRE/GlELC7jryvnk+X1tU3XJkk19eMO+CPAUx3nfwTclZlvA14Atpfx7cALZfyucp0kjaxGAzgi1gD/GPh0OQ/gJuChcske4LZyvLWcU16/uVwvSSOp6Tvgu4HfA86U8+XAjzLzdDk/AkyU4wngOYDy+qly/atExI6IOBARB06ePNlg6ZLUrMYCOCL+CXAiM59YyO+bmbszc1NmblqxYsVCfmtJ6qsmH0V+D/BPI+JW4PXAm4BPAMsi4opyl7sGmCrXTwHXAEci4gpgKfCDBuuTpKoauwPOzN/PzDWZuRZ4P/CVzPwt4KvAr5fLtgF7y/G+ck55/SuZ6b7qkkZWjXnA/xb4aERMMtPjvb+M3w8sL+MfBe6sUJsk9U1fVkPLzK8BXyvHh4B3zXPNy8Bv9KOeQTc9PU2r1Tpn3DVvpdHicpQDqNVqsevevSydWD835pq30ugxgAfU0on1rnkrjTjXgpCkSgxgSarEAJakSgxgSarEAJakSpwFMSTOvHKadrs9d95ut8kzPigI5/5uwDnTGg4G8JB48fiz3H34ZVZNzoTu1MFvsmzDDZWrGgzdvxvnTGtYGMBDZMmqtXNzg099/1DlagZL5+9GGhb2gCWpkp4COCLe08uYJKl3vd4B39PjmCSpR+ftAUfEu4FfAFZExEc7XnoTsKjJwiRp1F3oTbjFwE+X65Z0jP8tZxdVlyRdgvMGcGZ+Hfh6RHw2M5/pU02SNBZ6nYb2uojYDazt/JrMvKmJoiRpHPQawP8D+BTwaeCV5sqRpPHRawCfzsz7Gq1EksZMr9PQ/iwi/lVErI6Iq2c/Gq1MkkZcr3fAs9vF/27HWALr57lWktSDngI4M9c1XYgkjZueAjgi7phvPDMfWNhyJGl89NqC+PmO49cDNwN/CRjAknSJem1B7Oo8j4hlwBeaKEiSxsWlLkf5Y8C+sCRdhl57wH/GzKwHmFmE5x8ADzZV1LiZnp6m1WrNnbvdkDQeeu0B/6eO49PAM5l5pIF6xlKr1WLXvXtZOjEzq8/thqTx0GsP+OsRsZKzb8Y93VxJ42npxHq3G5LGTK87YtwO/B/gN4DbgccjwuUoJeky9NqC+PfAz2fmCYCIWAH8BfBQU4WNMnu+zXKbeg2LXgP4p2bDt/gBbuh5yez5Nstt6jUseg3gRyPiy8Dny/lvAo80U9J4sOfbLLep1zC40J5wbwNWZubvRsQ/A95bXvrfwOeaLk6SRtmF7oDvBn4fIDO/CHwRICL+YXntVxusTZJG2oX6uCsz8zvdg2VsbSMVSdKYuFAALzvPa29YwDokaexcKIAPRMQ/7x6MiN8GnmimJEkaDxfqAf8O8HBE/BZnA3cTsBj4tfN9YUS8HvgG8Lrycx7KzI9FxDpmVlJbXr7nBzNzOiJex8zyljcwM83tNzPz8KX8oSRpGJz3Djgzj2fmLwB/ABwuH3+Qme/OzGMX+N7/D7gpM98JXAdsiYgbgT8C7srMtwEvANvL9duBF8r4XeU6SRpZPT1MkZlfzcx7ysdXevyazMy/K6dXlo8EbuLsE3R7gNvK8dZyTnn95oiIXn6WJA2jRp9mi4hFEXEQOAE8BvwN8KPMPF0uOQJMlOMJ4DmA8vopZtoUkjSSGg3gzHwlM68D1gDvAn7ucr9nROyIiAMRceDkyZOX++0kqZq+rOeQmT8Cvgq8G1gWEbNv/q0BpsrxFHANQHl9KTNvxnV/r92ZuSkzN61YsaLp0iWpMY0FcESsKHvHERFvAN4HPMVMEM8uZbkN2FuO95VzyutfyUyXCJM0snpdjOdSrAb2RMQiZoL+wcz8UkR8D/hCRPwh0ALuL9ffD/z3iJgEfgi8v8HaJKm6xgI4M58ENs4zfoiZfnD3+MvMLPguSWPBNX0lqRIDWJIqabIHrMItiCTNxwDuA7cgkjQfA7hP3IJIUjd7wJJUiQEsSZUYwJJUiQEsSZUYwJJUiQEsSZU4DW0BdD9oAbBx40YWL15cqSJJw8AAXgDdD1qcmjrEPTth8+bNlSuTNMgM4AXS+aCFJPXCHrAkVWIAS1IlBrAkVWIPWCPvzCunabfbrxpzlooGgQGskffi8We5+/DLrJqcWYPZWSoaFAZwA7rvuFyAvb4lq9Y6S0UDxwBuQPcdlwuwDxZbEhoUBnBDOu+4XIB9sNiS0KAwgDWWbEloEDgNTZIqMYAlqRJbEBdpvpXPnOUg6VIYwBepe+UzcJaDpEtjAF+C7pXPnOUg6VLYA5akSgxgSarEAJakSgxgSarEAJakSgxgSarEAJakSgxgSarEAJakSgxgSarEAJakShoL4Ii4JiK+GhHfi4jvRsRHyvjVEfFYRDxdPl9VxiMiPhkRkxHxZERc31RtkjQImrwDPg38m8y8FrgR2BkR1wJ3AvszcwOwv5wD3AJsKB87gPsarE2SqmssgDPzaGb+ZTl+EXgKmAC2AnvKZXuA28rxVuCBnPEtYFlErG6qPkmqrS894IhYC2wEHgdWZubR8tIxYGU5ngCe6/iyI2Ws+3vtiIgDEXHg5MmTzRUtSQ1rfD3giPhp4H8Cv5OZfxsRc69lZkbERW0lkZm7gd0AmzZtchsKXTa3qVctjQZwRFzJTPh+LjO/WIaPR8TqzDxaWgwnyvgUcE3Hl68pY1Kj3KZetTQ5CyKA+4GnMvO/dLy0D9hWjrcBezvG7yizIW4ETnW0KqRGzW5Tv3zd21+13ZTUpCbvgN8DfBD4TkQcLGP/Dvg48GBEbAeeAW4vrz0C3ApMAi8BH2qwNkmqrrEAzsz/BcRrvHzzPNcnsLOpeiRp0PgknCRVYgBLUiUGsCRVYgBLUiUGsCRVYgBLUiUGsCRVYgBLUiUGsCRVYgBLUiUGsCRVYgBLUiUGsCRV0viOGMNuenqaVqs1d95ut8kzbsQh6fIZwBfQarXYde/euUW6pw5+k2UbbqhclaRRYAD3YOnEepavezsAp75/qHI1kkaFPWBJqsQAlqRKDGBJqsQAlqRKfBNO6nLmldO02+1zxjdu3MjixYsrVKRRZQBLXV48/ix3H36ZVZNn53ufmjrEPTth8+bNFSvTqDGApXksWbV2buqh1BR7wJJUiQEsSZUYwJJUiQEsSZX4JpzUg/mmpjktTZfLAJZ60D01zWlpWggGsNQjp6ZpodkDlqRKDGBJqsQWRBe3IJLULwZwF7cgktQvBvA83IJIUj/YA5akSgxgSarEAJakShoL4Ij4TESciIh2x9jVEfFYRDxdPl9VxiMiPhkRkxHxZERc31RdkjQomrwD/iywpWvsTmB/Zm4A9pdzgFuADeVjB3Bfg3VJ0kBoLIAz8xvAD7uGtwJ7yvEe4LaO8QdyxreAZRGxuqnaJGkQ9LsHvDIzj5bjY8DKcjwBPNdx3ZEydo6I2BERByLiwMmTJ5urVJIaVu1NuMxM4KIfMcvM3Zm5KTM3rVixooHKJKk/+h3Ax2dbC+XziTI+BVzTcd2aMiZJI6vfAbwP2FaOtwF7O8bvKLMhbgROdbQqJGkkNfYockR8Hvgl4M0RcQT4GPBx4MGI2A48A9xeLn8EuBWYBF4CPtRUXZI0KBoL4Mz8wGu8dPM81yaws6laJGkQ+SScJFViAEtSJS5HKS2A7oX8wV2TdWEGsHQJurepb7fbfOprkyxb81bAXZPVGwNYugTd29TP7pzirsm6GAawdIk6t6l35xRdCt+Ek6RKDGBJqsQAlqRKDGBJqsQAlqRKDGBJqsQAlqRKnAcsNaD7STnw0WSdywCWGtD9pJyPJms+BrDUkM4n5aT52AOWpEq8A5b6wJ6w5mMAS31gT1jzMYClPrEnrG72gCWpEgNYkioZ+xZE915e7XabPJMVK5I0LsY+gFutFrvu3cvSifXA2a1lpCbNNysCnBkxbsY+gAGWTqx3axn1VfesCHBmxDgauwC25aBB4awIjV0A23KQNCjGLoDBloOkweA0NEmqxACWpErGsgUhDYPuN4zBaWqjxgCWBkT33OB2u82nvjbJsjVvBZymNooMYGlAdM8Nnp2hM/uGsUtajh4DWBognXODu2fodAf0C88+zb+8qc073vGOuWsuNpBtc9RlAEtDpDug7/7yd88byHD+QO2eF2+bo78MYGmInS+Q4dxAne9J0DetXtdzm2O+O+bua9Q7A1gaId2PN1/ojb3uJ0EvtHNH9x3zfNeodwawNMIu9MbefE+Cdob4fAHeecc83zU/+clPALjyyivnxs53hzzOfWgDWBpx53tj70JeK8AvdM2iJctZ9dZrZ35mD22QcZ1uN1ABHBFbgE8Ai4BPZ+bHK5ckjb1eArz7miuWrjrvXfR8bZDz9aFhNO+KByaAI2IRcC/wPuAI8O2I2JeZ36tbmaTLcbFtkPnWSr7YKXfdd9nztUW6v0cvX7PQ/ycwMAEMvAuYzMxDABHxBWArsOABfGrq7F/4352YYtH/fZkfvPHvXdL5QnwPf8Zgfc9R+RkDVfeS5XR68djhi7r+pR8e4w/3THLVz8zcGf/4+aP83gfed86Uu1ntdps//vxjvPHNqwF4/m/aLHrDm7jqZ94yd03397jQ1/z4+aM88B//9YK2RiJzMBYjj4hfB7Zk5m+X8w8CmzPzw13X7QB2lNOfBf66r4XW82bg+dpFVDbuv4Nx//PD8P4Ons/MLd2Dg3QH3JPM3A3srl1Hv0XEgczcVLuOmsb9dzDuf34Yvd/BIC1HOQVc03G+poxJ0kgapAD+NrAhItZFxGLg/cC+yjVJUmMGpgWRmacj4sPAl5mZhvaZzPxu5bIGydi1XeYx7r+Dcf/zw4j9DgbmTThJGjeD1IKQpLFiAEtSJQbwkIiI/xARUxFxsHzcWrumfomILRHx1xExGRF31q6nhog4HBHfKX/3B2rX0w8R8ZmIOBER7Y6xqyPisYh4uny+qmaNl8sAHi53ZeZ15eOR2sX0Q8cj6rcA1wIfiIhr61ZVzS+Xv/uRmQd7AZ8Fuh9euBPYn5kbgP3lfGgZwBp0c4+oZ+Y0MPuIukZcZn4D+GHX8FZgTzneA9zWz5oWmgE8XD4cEU+Wf5oN9T+9LsIE8FzH+ZEyNm4S+POIeKI8jj+uVmbm0XJ8DFhZs5jLZQAPkIj4i4hoz/OxFbgPeCtwHXAU+M81a1XfvTczr2emFbMzIn6xdkG15cwc2qGeRzswD2IIMvMf9XJdRPxX4EsNlzMofEQdyMyp8vlERDzMTGvmG3WrquJ4RKzOzKMRsRo4Ubugy+Ed8JAo/7HN+jXg3BWrR9PYP6IeEW+MiCWzx8CvMD5//932AdvK8TZgb8VaLpt3wMPjjyPiOmb+yXUY+BdVq+kTH1EHZvqcD0cEzPxv9k8y89G6JTUvIj4P/BLw5og4AnwM+DjwYERsB54Bbq9X4eXzUWRJqsQWhCRVYgBLUiUGsCRVYgBLUiUGsCRVYgBLUiUGsCRV8v8BuwKOiAy9i44AAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 360x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.displot(data_matrix_log2[:,0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After checking the histogram we can say that we will use the mean minus 2 standard deviation of the expressed genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\modos\\.conda\\envs\\Single_cell_RNA_seq\\lib\\site-packages\\ipykernel_launcher.py:1: RuntimeWarning: invalid value encountered in greater\n",
      "  \"\"\"Entry point for launching an IPython kernel.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([[ True,  True,  True, ..., False, False, False],\n",
       "       [ True,  True,  True, ..., False, False, False],\n",
       "       [ True,  True,  True, ..., False, False, False],\n",
       "       ...,\n",
       "       [ True,  True,  True, ...,  True,  True,  True],\n",
       "       [ True,  True,  True, ...,  True, False, False],\n",
       "       [ True,  True,  True, ...,  True,  True, False]])"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_matrix_log2 > (mean_cell-2*std)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For simplicity we call \"NaN\" each number which is below the treshold."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cell_log2 = pd.DataFrame(data=data_matrix_log2, index=df_cell_exp.index ,columns=df_cell_exp.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RNA.CD8..IELs_Healthy</th>\n",
       "      <th>RNA.CD8..LP_Healthy</th>\n",
       "      <th>RNA.CD4..Memory_Healthy</th>\n",
       "      <th>RNA.MT.hi_Healthy</th>\n",
       "      <th>RNA.Cycling.T_Healthy</th>\n",
       "      <th>RNA.NKs_Healthy</th>\n",
       "      <th>RNA.CD4..Activated.Fos.lo_Healthy</th>\n",
       "      <th>RNA.CD4..Activated.Fos.hi_Healthy</th>\n",
       "      <th>RNA.CD8..IELs_Uninflamed</th>\n",
       "      <th>RNA.MT.hi_Uninflamed</th>\n",
       "      <th>...</th>\n",
       "      <th>RNA.Immature.Goblet_Inflamed</th>\n",
       "      <th>RNA.Stem_Uninflamed</th>\n",
       "      <th>RNA.Immature.Enterocytes.2_Inflamed</th>\n",
       "      <th>RNA.Goblet_Inflamed</th>\n",
       "      <th>RNA.Tuft_Inflamed</th>\n",
       "      <th>RNA.Enterocytes_Inflamed</th>\n",
       "      <th>RNA.Best4..Enterocytes_Inflamed</th>\n",
       "      <th>RNA.Enteroendocrine_Inflamed</th>\n",
       "      <th>RNA.M.cells_Inflamed</th>\n",
       "      <th>RNA.M.cells_Uninflamed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7SK</th>\n",
       "      <td>-5.050172</td>\n",
       "      <td>-4.568056</td>\n",
       "      <td>-5.896044</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-2.126639</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-6.050093</td>\n",
       "      <td>-5.031619</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-9.218304</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.808425</td>\n",
       "      <td>-8.664267</td>\n",
       "      <td>-6.981502</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>-4.001756</td>\n",
       "      <td>-5.273798</td>\n",
       "      <td>-4.336499</td>\n",
       "      <td>-2.262358</td>\n",
       "      <td>-1.635142</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-4.468881</td>\n",
       "      <td>-5.753884</td>\n",
       "      <td>-4.236870</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-6.240706</td>\n",
       "      <td>-8.659406</td>\n",
       "      <td>-3.485761</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>-1.852068</td>\n",
       "      <td>-2.332119</td>\n",
       "      <td>-2.853021</td>\n",
       "      <td>-1.761008</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.003621</td>\n",
       "      <td>-2.941349</td>\n",
       "      <td>-1.964903</td>\n",
       "      <td>-1.710843</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>-8.391626</td>\n",
       "      <td>-7.462761</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-5.274231</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-3.659258</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.596775</td>\n",
       "      <td>-2.243375</td>\n",
       "      <td>-0.236862</td>\n",
       "      <td>-2.415296</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.040307</td>\n",
       "      <td>-0.361069</td>\n",
       "      <td>1.40798</td>\n",
       "      <td>-1.12881</td>\n",
       "      <td>-0.4392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>-5.408426</td>\n",
       "      <td>-2.221338</td>\n",
       "      <td>-2.303227</td>\n",
       "      <td>-2.574148</td>\n",
       "      <td>-3.354513</td>\n",
       "      <td>-0.873881</td>\n",
       "      <td>-2.786930</td>\n",
       "      <td>-2.582460</td>\n",
       "      <td>-6.600697</td>\n",
       "      <td>-2.040528</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 153 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          RNA.CD8..IELs_Healthy  RNA.CD8..LP_Healthy  RNA.CD4..Memory_Healthy  \\\n",
       "7SK                   -5.050172            -4.568056                -5.896044   \n",
       "A1BG                  -4.001756            -5.273798                -4.336499   \n",
       "A1BG-AS1              -1.852068            -2.332119                -2.853021   \n",
       "A1CF                  -8.391626            -7.462761                      NaN   \n",
       "A2M                   -5.408426            -2.221338                -2.303227   \n",
       "\n",
       "          RNA.MT.hi_Healthy  RNA.Cycling.T_Healthy  RNA.NKs_Healthy  \\\n",
       "7SK                     NaN                    NaN        -2.126639   \n",
       "A1BG              -2.262358              -1.635142              NaN   \n",
       "A1BG-AS1          -1.761008                    NaN        -1.003621   \n",
       "A1CF                    NaN              -5.274231              NaN   \n",
       "A2M               -2.574148              -3.354513        -0.873881   \n",
       "\n",
       "          RNA.CD4..Activated.Fos.lo_Healthy  \\\n",
       "7SK                                     NaN   \n",
       "A1BG                              -4.468881   \n",
       "A1BG-AS1                          -2.941349   \n",
       "A1CF                                    NaN   \n",
       "A2M                               -2.786930   \n",
       "\n",
       "          RNA.CD4..Activated.Fos.hi_Healthy  RNA.CD8..IELs_Uninflamed  \\\n",
       "7SK                               -6.050093                 -5.031619   \n",
       "A1BG                              -5.753884                 -4.236870   \n",
       "A1BG-AS1                          -1.964903                 -1.710843   \n",
       "A1CF                                    NaN                 -3.659258   \n",
       "A2M                               -2.582460                 -6.600697   \n",
       "\n",
       "          RNA.MT.hi_Uninflamed  ...  RNA.Immature.Goblet_Inflamed  \\\n",
       "7SK                        NaN  ...                           NaN   \n",
       "A1BG                       NaN  ...                           NaN   \n",
       "A1BG-AS1                   NaN  ...                           NaN   \n",
       "A1CF                       NaN  ...                     -0.596775   \n",
       "A2M                  -2.040528  ...                           NaN   \n",
       "\n",
       "          RNA.Stem_Uninflamed  RNA.Immature.Enterocytes.2_Inflamed  \\\n",
       "7SK                 -9.218304                                  NaN   \n",
       "A1BG                -6.240706                            -8.659406   \n",
       "A1BG-AS1                  NaN                                  NaN   \n",
       "A1CF                -2.243375                            -0.236862   \n",
       "A2M                       NaN                                  NaN   \n",
       "\n",
       "          RNA.Goblet_Inflamed  RNA.Tuft_Inflamed  RNA.Enterocytes_Inflamed  \\\n",
       "7SK                       NaN           2.808425                 -8.664267   \n",
       "A1BG                -3.485761                NaN                       NaN   \n",
       "A1BG-AS1                  NaN                NaN                       NaN   \n",
       "A1CF                -2.415296                NaN                 -1.040307   \n",
       "A2M                       NaN                NaN                       NaN   \n",
       "\n",
       "          RNA.Best4..Enterocytes_Inflamed  RNA.Enteroendocrine_Inflamed  \\\n",
       "7SK                             -6.981502                           NaN   \n",
       "A1BG                                  NaN                           NaN   \n",
       "A1BG-AS1                              NaN                           NaN   \n",
       "A1CF                            -0.361069                       1.40798   \n",
       "A2M                                   NaN                           NaN   \n",
       "\n",
       "          RNA.M.cells_Inflamed  RNA.M.cells_Uninflamed  \n",
       "7SK                        NaN                     NaN  \n",
       "A1BG                       NaN                     NaN  \n",
       "A1BG-AS1                   NaN                     NaN  \n",
       "A1CF                  -1.12881                 -0.4392  \n",
       "A2M                        NaN                     NaN  \n",
       "\n",
       "[5 rows x 153 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_cell_log2.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will call every genee which is expressed below the treshold with \"NaN\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\modos\\.conda\\envs\\Single_cell_RNA_seq\\lib\\site-packages\\ipykernel_launcher.py:1: RuntimeWarning: invalid value encountered in less\n",
      "  \"\"\"Entry point for launching an IPython kernel.\n"
     ]
    }
   ],
   "source": [
    "df_cell_log2[df_cell_log2.values < (mean_cell-2*std)] = \"NaN\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's see what we hjave done and the avaialbe cell types."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>RNA.CD8..IELs_Healthy</th>\n",
       "      <th>RNA.CD8..LP_Healthy</th>\n",
       "      <th>RNA.CD4..Memory_Healthy</th>\n",
       "      <th>RNA.MT.hi_Healthy</th>\n",
       "      <th>RNA.Cycling.T_Healthy</th>\n",
       "      <th>RNA.NKs_Healthy</th>\n",
       "      <th>RNA.CD4..Activated.Fos.lo_Healthy</th>\n",
       "      <th>RNA.CD4..Activated.Fos.hi_Healthy</th>\n",
       "      <th>RNA.CD8..IELs_Uninflamed</th>\n",
       "      <th>RNA.MT.hi_Uninflamed</th>\n",
       "      <th>...</th>\n",
       "      <th>RNA.Immature.Goblet_Inflamed</th>\n",
       "      <th>RNA.Stem_Uninflamed</th>\n",
       "      <th>RNA.Immature.Enterocytes.2_Inflamed</th>\n",
       "      <th>RNA.Goblet_Inflamed</th>\n",
       "      <th>RNA.Tuft_Inflamed</th>\n",
       "      <th>RNA.Enterocytes_Inflamed</th>\n",
       "      <th>RNA.Best4..Enterocytes_Inflamed</th>\n",
       "      <th>RNA.Enteroendocrine_Inflamed</th>\n",
       "      <th>RNA.M.cells_Inflamed</th>\n",
       "      <th>RNA.M.cells_Uninflamed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7SK</th>\n",
       "      <td>-5.05017</td>\n",
       "      <td>-4.56806</td>\n",
       "      <td>-5.896044</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-2.126639</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-6.05009</td>\n",
       "      <td>-5.031619</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.808425</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-6.9815</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>-4.00176</td>\n",
       "      <td>-5.2738</td>\n",
       "      <td>-4.336499</td>\n",
       "      <td>-2.262358</td>\n",
       "      <td>-1.635142</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-4.46888</td>\n",
       "      <td>-5.75388</td>\n",
       "      <td>-4.236870</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-6.24071</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-3.48576</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>-1.85207</td>\n",
       "      <td>-2.33212</td>\n",
       "      <td>-2.853021</td>\n",
       "      <td>-1.761008</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.003621</td>\n",
       "      <td>-2.94135</td>\n",
       "      <td>-1.9649</td>\n",
       "      <td>-1.710843</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>NaN</td>\n",
       "      <td>-7.46276</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-5.274231</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-3.659258</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.596775</td>\n",
       "      <td>-2.24338</td>\n",
       "      <td>-0.236862</td>\n",
       "      <td>-2.4153</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.04031</td>\n",
       "      <td>-0.361069</td>\n",
       "      <td>1.40798</td>\n",
       "      <td>-1.12881</td>\n",
       "      <td>-0.4392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>-5.40843</td>\n",
       "      <td>-2.22134</td>\n",
       "      <td>-2.303227</td>\n",
       "      <td>-2.574148</td>\n",
       "      <td>-3.354513</td>\n",
       "      <td>-0.873881</td>\n",
       "      <td>-2.78693</td>\n",
       "      <td>-2.58246</td>\n",
       "      <td>-6.600697</td>\n",
       "      <td>-2.040528</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 153 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         RNA.CD8..IELs_Healthy RNA.CD8..LP_Healthy  RNA.CD4..Memory_Healthy  \\\n",
       "7SK                   -5.05017            -4.56806                -5.896044   \n",
       "A1BG                  -4.00176             -5.2738                -4.336499   \n",
       "A1BG-AS1              -1.85207            -2.33212                -2.853021   \n",
       "A1CF                       NaN            -7.46276                      NaN   \n",
       "A2M                   -5.40843            -2.22134                -2.303227   \n",
       "\n",
       "          RNA.MT.hi_Healthy  RNA.Cycling.T_Healthy  RNA.NKs_Healthy  \\\n",
       "7SK                     NaN                    NaN        -2.126639   \n",
       "A1BG              -2.262358              -1.635142              NaN   \n",
       "A1BG-AS1          -1.761008                    NaN        -1.003621   \n",
       "A1CF                    NaN              -5.274231              NaN   \n",
       "A2M               -2.574148              -3.354513        -0.873881   \n",
       "\n",
       "         RNA.CD4..Activated.Fos.lo_Healthy RNA.CD4..Activated.Fos.hi_Healthy  \\\n",
       "7SK                                    NaN                          -6.05009   \n",
       "A1BG                              -4.46888                          -5.75388   \n",
       "A1BG-AS1                          -2.94135                           -1.9649   \n",
       "A1CF                                   NaN                               NaN   \n",
       "A2M                               -2.78693                          -2.58246   \n",
       "\n",
       "          RNA.CD8..IELs_Uninflamed  RNA.MT.hi_Uninflamed  ...  \\\n",
       "7SK                      -5.031619                   NaN  ...   \n",
       "A1BG                     -4.236870                   NaN  ...   \n",
       "A1BG-AS1                 -1.710843                   NaN  ...   \n",
       "A1CF                     -3.659258                   NaN  ...   \n",
       "A2M                      -6.600697             -2.040528  ...   \n",
       "\n",
       "          RNA.Immature.Goblet_Inflamed  RNA.Stem_Uninflamed  \\\n",
       "7SK                                NaN                  NaN   \n",
       "A1BG                               NaN             -6.24071   \n",
       "A1BG-AS1                           NaN                  NaN   \n",
       "A1CF                         -0.596775             -2.24338   \n",
       "A2M                                NaN                  NaN   \n",
       "\n",
       "          RNA.Immature.Enterocytes.2_Inflamed  RNA.Goblet_Inflamed  \\\n",
       "7SK                                       NaN                  NaN   \n",
       "A1BG                                      NaN             -3.48576   \n",
       "A1BG-AS1                                  NaN                  NaN   \n",
       "A1CF                                -0.236862              -2.4153   \n",
       "A2M                                       NaN                  NaN   \n",
       "\n",
       "          RNA.Tuft_Inflamed  RNA.Enterocytes_Inflamed  \\\n",
       "7SK                2.808425                       NaN   \n",
       "A1BG                    NaN                       NaN   \n",
       "A1BG-AS1                NaN                       NaN   \n",
       "A1CF                    NaN                  -1.04031   \n",
       "A2M                     NaN                       NaN   \n",
       "\n",
       "          RNA.Best4..Enterocytes_Inflamed  RNA.Enteroendocrine_Inflamed  \\\n",
       "7SK                               -6.9815                           NaN   \n",
       "A1BG                                  NaN                           NaN   \n",
       "A1BG-AS1                              NaN                           NaN   \n",
       "A1CF                            -0.361069                       1.40798   \n",
       "A2M                                   NaN                           NaN   \n",
       "\n",
       "          RNA.M.cells_Inflamed  RNA.M.cells_Uninflamed  \n",
       "7SK                        NaN                     NaN  \n",
       "A1BG                       NaN                     NaN  \n",
       "A1BG-AS1                   NaN                     NaN  \n",
       "A1CF                  -1.12881                 -0.4392  \n",
       "A2M                        NaN                     NaN  \n",
       "\n",
       "[5 rows x 153 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_cell_log2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'RNA.Macrophages_Uninflamed', 'RNA.CD4..PD1._Inflamed', 'RNA.ILCs_Inflamed', 'RNA.Cycling.TA_Healthy', 'RNA.TA.1_Inflamed', 'RNA.WNT2B..Fos.lo.1_Uninflamed', 'RNA.Tregs_Healthy', 'RNA.Best4..Enterocytes_Inflamed', 'RNA.Enterocyte.Progenitors_Inflamed', 'RNA.Microvascular_Inflamed', 'RNA.Goblet_Inflamed', 'RNA.CD4..Activated.Fos.hi_Healthy', 'RNA.CD8..IELs_Inflamed', 'RNA.WNT5B..2_Uninflamed', 'RNA.Inflammatory.Fibroblasts_Inflamed', 'RNA.CD8..IELs_Healthy', 'RNA.DC2_Inflamed', 'RNA.Tregs_Uninflamed', 'RNA.Immature.Enterocytes.1_Inflamed', 'RNA.WNT5B..2_Healthy', 'RNA.Follicular_Uninflamed', 'RNA.Inflammatory.Monocytes_Uninflamed', 'RNA.Microvascular_Uninflamed', 'RNA.Plasma_Uninflamed', 'RNA.CD8..IL17._Inflamed', 'RNA.CD8..IELs_Uninflamed', 'RNA.DC1_Inflamed', 'RNA.M.cells_Uninflamed', 'RNA.WNT5B..1_Inflamed', 'RNA.WNT2B..Fos.lo.2_Uninflamed', 'RNA.CD4..Activated.Fos.lo_Uninflamed', 'RNA.Cycling.Monocytes_Inflamed', 'RNA.NKs_Healthy', 'RNA.Secretory.TA_Inflamed', 'RNA.GC_Healthy', 'RNA.Cycling.B_Inflamed', 'RNA.Inflammatory.Monocytes_Healthy', 'RNA.MT.hi_Inflamed', 'RNA.Post.capillary.Venules_Inflamed', 'RNA.Cycling.T_Healthy', 'RNA.WNT2B..Fos.hi_Inflamed', 'RNA.RSPO3._Inflamed', 'RNA.Cycling.TA_Uninflamed', 'RNA.Enterocyte.Progenitors_Healthy', 'RNA.CD69..Mast_Inflamed', 'RNA.TA.2_Inflamed', 'RNA.TA.2_Healthy', 'RNA.DC1_Uninflamed', 'RNA.Endothelial_Inflamed', 'RNA.WNT5B..2_Inflamed', 'RNA.Pericytes_Healthy', 'RNA.Inflammatory.Monocytes_Inflamed', 'RNA.Inflammatory.Fibroblasts_Uninflamed', 'RNA.Microvascular_Healthy', 'RNA.M.cells_Inflamed', 'RNA.Glia_Healthy', 'RNA.MT.hi_Uninflamed', 'RNA.DC2_Healthy', 'RNA.Macrophages_Healthy', 'RNA.Macrophages_Inflamed', 'RNA.CD4..PD1._Healthy', 'RNA.Cycling.T_Uninflamed', 'RNA.CD69..Mast_Healthy', 'RNA.Enteroendocrine_Inflamed', 'RNA.Tregs_Inflamed', 'RNA.Secretory.TA_Healthy', 'RNA.CD4..Activated.Fos.hi_Uninflamed', 'RNA.Follicular_Inflamed', 'RNA.Plasma_Healthy', 'RNA.Immature.Enterocytes.1_Uninflamed', 'RNA.Glia_Inflamed', 'RNA.Best4..Enterocytes_Healthy', 'RNA.Best4..Enterocytes_Uninflamed', 'RNA.Stem_Inflamed', 'RNA.Tuft_Inflamed', 'RNA.Enterocytes_Uninflamed', 'RNA.Goblet_Uninflamed', 'RNA.CD4..Memory_Healthy', 'RNA.RSPO3._Uninflamed', 'RNA.Immature.Goblet_Inflamed', 'RNA.Enterocytes_Healthy', 'RNA.CD4..Activated.Fos.hi_Inflamed', 'RNA.MT.hi_Healthy', 'RNA.CD8..IL17._Healthy', 'RNA.WNT2B..Fos.lo.2_Inflamed', 'RNA.Goblet_Healthy', 'RNA.Cycling.B_Uninflamed', 'RNA.Cycling.Monocytes_Healthy', 'RNA.Pericytes_Inflamed', 'RNA.Immature.Enterocytes.2_Uninflamed', 'RNA.ILCs_Uninflamed', 'RNA.WNT2B..Fos.lo.1_Inflamed', 'RNA.Myofibroblasts_Healthy', 'RNA.CD4..Activated.Fos.lo_Healthy', 'RNA.Immature.Enterocytes.2_Healthy', 'RNA.WNT2B..Fos.lo.2_Healthy', 'RNA.M.cells_Healthy', 'RNA.CD8..IL17._Uninflamed', 'RNA.WNT2B..Fos.hi_Uninflamed', 'RNA.DC2_Uninflamed', 'RNA.TA.1_Uninflamed', 'RNA.Cycling.TA_Inflamed', 'RNA.Immature.Goblet_Healthy', 'RNA.Secretory.TA_Uninflamed', 'RNA.CD4..Memory_Uninflamed', 'RNA.Enteroendocrine_Uninflamed', 'RNA.Tuft_Uninflamed', 'RNA.Plasma_Inflamed', 'RNA.CD8..LP_Healthy', 'RNA.Stem_Uninflamed', 'RNA.Glia_Uninflamed', 'RNA.Follicular_Healthy', 'RNA.NKs_Uninflamed', 'RNA.DC1_Healthy', 'RNA.WNT5B..1_Uninflamed', 'RNA.CD4..Activated.Fos.lo_Inflamed', 'RNA.CD69..Mast_Healthy.1', 'RNA.NKs_Inflamed', 'RNA.Stem_Healthy', 'RNA.Inflammatory.Fibroblasts_Healthy', 'RNA.CD8..LP_Inflamed', 'RNA.CD69..Mast_Uninflamed.1', 'RNA.GC_Inflamed', 'RNA.Enterocyte.Progenitors_Uninflamed', 'RNA.GC_Uninflamed', 'RNA.Enterocytes_Inflamed', 'RNA.CD69..Mast_Inflamed.1', 'RNA.WNT2B..Fos.lo.1_Healthy', 'RNA.ILCs_Healthy', 'RNA.Post.capillary.Venules_Healthy', 'RNA.CD4..Memory_Inflamed', 'RNA.Myofibroblasts_Inflamed', 'RNA.WNT2B..Fos.hi_Healthy', 'RNA.WNT5B..1_Healthy', 'RNA.RSPO3._Healthy', 'RNA.Cycling.B_Healthy', 'RNA.CD8..LP_Uninflamed', 'RNA.Enteroendocrine_Healthy', 'RNA.TA.1_Healthy', 'RNA.Tuft_Healthy', 'RNA.Post.capillary.Venules_Uninflamed', 'RNA.Myofibroblasts_Uninflamed', 'RNA.Immature.Enterocytes.2_Inflamed', 'RNA.TA.2_Uninflamed', 'RNA.Immature.Goblet_Uninflamed', 'RNA.Endothelial_Uninflamed', 'RNA.Cycling.Monocytes_Uninflamed', 'RNA.Pericytes_Uninflamed', 'RNA.Immature.Enterocytes.1_Healthy', 'RNA.Cycling.T_Inflamed', 'RNA.CD69..Mast_Uninflamed', 'RNA.CD4..PD1._Uninflamed', 'RNA.Endothelial_Healthy'}\n"
     ]
    }
   ],
   "source": [
    "print(set(df_cell_log2.columns))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we select healthy/uninflamed average expression data in the 5 selected cell types."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_macrophage_healthy = df_cell_log2['RNA.Macrophages_Healthy']\n",
    "df_macrophage_uninflamed = df_cell_log2['RNA.Macrophages_Uninflamed']\n",
    "\n",
    "df_DC1_healthy = df_cell_log2['RNA.DC1_Healthy']\n",
    "df_DC1_uninflamed = df_cell_log2['RNA.DC1_Uninflamed']\n",
    "\n",
    "df_Treg_healthy = df_cell_log2['RNA.Tregs_Healthy']\n",
    "df_Treg_uninflamed = df_cell_log2['RNA.Tregs_Uninflamed'] \n",
    "\n",
    "df_myofibroblast_healthy = df_cell_log2['RNA.Myofibroblasts_Healthy']\n",
    "df_myofibroblast_uninflamed = df_cell_log2['RNA.Myofibroblasts_Uninflamed']\n",
    "\n",
    "df_goblet_healthy = df_cell_log2['RNA.Goblet_Healthy']\n",
    "df_goblet_uninflamed = df_cell_log2['RNA.Goblet_Uninflamed']\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For calcualting the intercellular interactions we can store theese dataframes in a dictionarry."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "healthy_cell_types = {\"DC1\": df_DC1_healthy,\"Macrophage\": df_macrophage_healthy, \"Goblet\": df_goblet_healthy,\n",
    "                      \"Myofibroblast\": df_myofibroblast_healthy, 'Treg': df_Treg_healthy}\n",
    "\n",
    "uninflamed_cell_types = {\"DC1\": df_DC1_uninflamed,\"Macrophage\": df_macrophage_uninflamed, \"Goblet\": df_goblet_uninflamed,\n",
    "                         \"Myofibroblast\": df_myofibroblast_uninflamed, 'Treg': df_Treg_uninflamed} "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we can use the intercellular interactions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create dictionaries for source-target interactions and their annotations from OmniPath\n",
    "interactions = {}\n",
    "interaction_annotation = {}\n",
    "\n",
    "for i, interaction in filtered_intercell_network.iterrows():\n",
    "    if interaction[\"genesymbol_intercell_source\"] not in interactions:\n",
    "        interactions[interaction[\"genesymbol_intercell_source\"]] = []\n",
    "    if interaction[\"genesymbol_intercell_target\"] not in  interactions[interaction[\"genesymbol_intercell_source\"]]:\n",
    "        interactions[interaction[\"genesymbol_intercell_source\"]].append(interaction[\"genesymbol_intercell_target\"])\n",
    "    source_tupple = (interaction[\"genesymbol_intercell_source\"],interaction[\"category_intercell_source\"])\n",
    "    target_tupple = (interaction[\"genesymbol_intercell_target\"],interaction[\"category_intercell_target\"])\n",
    "    if source_tupple not in interaction_annotation:\n",
    "        interaction_annotation[source_tupple] = []\n",
    "    if target_tupple not in interaction_annotation[source_tupple]:\n",
    "        interaction_annotation[source_tupple].append(target_tupple)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create all of the possible interactions of cells (independently from the condition)\n",
    "Important: directionality (A-B and B-A are different)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_interactions(cells, healthy_cell_types,interactions):\n",
    "    healthy_interactions = set()\n",
    "    for source in healthy_cell_types[cells[0]].keys():\n",
    "            # selecting those ones which play role in intercellular communication as a transmitter\n",
    "            if source in interactions.keys():\n",
    "                # iterating through target cell type expressed genes\n",
    "                for target in healthy_cell_types[cells[1]].keys():\n",
    "\n",
    "                    # selecting those ones which play role in intercellular communication as a receiver\n",
    "                    if target in interactions[source]:\n",
    "                        if str(healthy_cell_types[cells[1]][target]) != 'nan' and str(healthy_cell_types[cells[0]][source]) != 'nan':\n",
    "                            healthy_interactions.add((source, target))\n",
    "    return healthy_interactions\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Creating dictinaries dataframe for healthy and UC interactions\n",
    "healthy_intercell_tupplellist = []\n",
    "uc_intercell_tupplelist = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DC1 Macrophage\n",
      "DC1 Goblet\n",
      "DC1 Myofibroblast\n",
      "DC1 Treg\n",
      "Macrophage DC1\n",
      "Macrophage Goblet\n",
      "Macrophage Myofibroblast\n",
      "Macrophage Treg\n",
      "Goblet DC1\n",
      "Goblet Macrophage\n",
      "Goblet Myofibroblast\n",
      "Goblet Treg\n",
      "Myofibroblast DC1\n",
      "Myofibroblast Macrophage\n",
      "Myofibroblast Goblet\n",
      "Myofibroblast Treg\n",
      "Treg DC1\n",
      "Treg Macrophage\n",
      "Treg Goblet\n",
      "Treg Myofibroblast\n"
     ]
    }
   ],
   "source": [
    "for i in healthy_cell_types.keys():\n",
    "    for j in healthy_cell_types.keys():\n",
    "        if i!=j:\n",
    "            print(i,j)\n",
    "            # storing cell-type specific connections in sets\n",
    "            healthy_interactions = create_interactions([i,j], healthy_cell_types, interactions)\n",
    "            UC_interactions = create_interactions([i,j], uninflamed_cell_types,interactions)\n",
    "            # writing out the condition specific interactions\n",
    "            healthy_only = healthy_interactions.difference(UC_interactions)\n",
    "            UC_only = UC_interactions.difference(healthy_interactions)\n",
    "            #Adding the intercellular interactions\n",
    "            healthy_intercell_tupplellist.append((i,j,len(healthy_only)))\n",
    "            uc_intercell_tupplelist.append((i,j,len(UC_only)))\n",
    "\n",
    "            with open(i + \"_\" + j + \"_healthy_only.txt\", 'w') as output_file_1:\n",
    "                #print(UC_only)\n",
    "                for inter in healthy_only:\n",
    "                    for source_annotation in interaction_annotation:\n",
    "                        if inter[0] == source_annotation[0]:\n",
    "                            for target_annotation in interaction_annotation[source_annotation]:\n",
    "                                if inter[1] == target_annotation[0]:\n",
    "                                    output_file_1.write(source_annotation[0] + \",\" + source_annotation[1] + \",\" \n",
    "                                                        + target_annotation[0] + \",\" + target_annotation[1] + \"\\n\")\n",
    "            with open(i + \"_\" + j + \"_UC_only.txt\", 'w') as output_file_2:\n",
    "                for inter in UC_only:\n",
    "                        for source_annotation in interaction_annotation:\n",
    "                            if inter[0] == source_annotation[0]:\n",
    "                                for target_annotation in interaction_annotation[source_annotation]:\n",
    "                                    if inter[1] == target_annotation[0]:\n",
    "                                        output_file_2.write(source_annotation[0] + \",\" + source_annotation[1] + \",\"\n",
    "                                                            + target_annotation[0] + \",\" + target_annotation[1] + \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we make the graphs of the interactions between the cells."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "g_healthy = ig.Graph.TupleList(healthy_intercell_tupplellist, weights=True, directed=True)\n",
    "g_uc = ig.Graph.TupleList(uc_intercell_tupplelist, weights=True, directed=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('DC1', 'Macrophage', 912),\n",
       " ('DC1', 'Goblet', 781),\n",
       " ('DC1', 'Myofibroblast', 731),\n",
       " ('DC1', 'Treg', 557),\n",
       " ('Macrophage', 'DC1', 568),\n",
       " ('Macrophage', 'Goblet', 689),\n",
       " ('Macrophage', 'Myofibroblast', 617),\n",
       " ('Macrophage', 'Treg', 501),\n",
       " ('Goblet', 'DC1', 457),\n",
       " ('Goblet', 'Macrophage', 652),\n",
       " ('Goblet', 'Myofibroblast', 483),\n",
       " ('Goblet', 'Treg', 438),\n",
       " ('Myofibroblast', 'DC1', 437),\n",
       " ('Myofibroblast', 'Macrophage', 558),\n",
       " ('Myofibroblast', 'Goblet', 483),\n",
       " ('Myofibroblast', 'Treg', 387),\n",
       " ('Treg', 'DC1', 491),\n",
       " ('Treg', 'Macrophage', 800),\n",
       " ('Treg', 'Goblet', 682),\n",
       " ('Treg', 'Myofibroblast', 616)]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "healthy_intercell_tupplellist"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can visualise the interactions.\n",
    "Igraph contain the visulasitation parameters as a dictionarry. You can also wirte out the edgelist to visualise it in cytoscape."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<igraph.EdgeSeq at 0x19ee53e1948>"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g_healthy.es"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "visual_style = {}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below are the iGraph visual parameters which you can play with for your intercellular network."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Curve the edges\n",
    "visual_style[\"edge_curved\"] = True\n",
    "# Set the layout\n",
    "my_layout = g_healthy.layout_circle(order=[\"Goblet\",\"Treg\",\"Myofibroblast\",\"Macrophage\",\"DC1\"])\n",
    "visual_style[\"layout\"] = my_layout\n",
    "#Add annoation\n",
    "visual_style[\"vertex_label\"] = g_healthy.vs[\"name\"]\n",
    "#Calcualte the edge wheight relative the number of edges\n",
    "visual_style[\"edge_width\"] = 0.01 * np.array(g_healthy.es[\"weight\"])\n",
    "\n",
    "#Setting the vertex visualistation parameters -colours and label position\n",
    "visual_style[\"vertex_label_size\"] = 30\n",
    "visual_style[\"vertex_color\"] = \"grey\"\n",
    "visual_style[\"vertex_label_dist\"] = 0.75\n",
    "visual_style[\"vertex_size\"] = 40\n",
    "\n",
    "#Setting the edge visualisation parameters\n",
    "visual_style[\"edge_color\"] = \"#004C66\" #hexa code for blue\n",
    "visual_style[\"edge_arrow_size\"] = 2\n",
    "\n",
    "#Setting the visualistation place. igraph need margin yo work\n",
    "visual_style[\"bbox\"] = (600, 600)\n",
    "visual_style[\"margin\"] = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n",
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"600pt\" height=\"600pt\" viewBox=\"0 0 600 600\" version=\"1.1\">\n",
       "<defs>\n",
       "<g>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-0\">\n",
       "<path style=\"stroke:none;\" d=\"M 3.75 0 L 3.75 -18.75 L 18.75 -18.75 L 18.75 0 Z M 4.21875 -0.46875 L 18.28125 -0.46875 L 18.28125 -18.28125 L 4.21875 -18.28125 Z M 4.21875 -0.46875 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-1\">\n",
       "<path style=\"stroke:none;\" d=\"M 2.3125 0 L 2.3125 -21.476563 L 9.710938 -21.476563 C 11.378906 -21.472656 12.65625 -21.371094 13.535156 -21.167969 C 14.765625 -20.882813 15.8125 -20.371094 16.683594 -19.628906 C 17.8125 -18.671875 18.660156 -17.449219 19.226563 -15.960938 C 19.785156 -14.46875 20.066406 -12.765625 20.070313 -10.855469 C 20.066406 -9.21875 19.875 -7.773438 19.496094 -6.515625 C 19.113281 -5.257813 18.625 -4.214844 18.03125 -3.390625 C 17.433594 -2.5625 16.785156 -1.914063 16.078125 -1.441406 C 15.367188 -0.96875 14.511719 -0.609375 13.511719 -0.367188 C 12.511719 -0.121094 11.359375 0 10.0625 0 Z M 5.15625 -2.535156 L 9.742188 -2.535156 C 11.15625 -2.53125 12.265625 -2.664063 13.074219 -2.929688 C 13.875 -3.191406 14.519531 -3.5625 15 -4.042969 C 15.671875 -4.714844 16.195313 -5.621094 16.574219 -6.761719 C 16.949219 -7.898438 17.136719 -9.277344 17.140625 -10.898438 C 17.136719 -13.140625 16.769531 -14.863281 16.03125 -16.074219 C 15.292969 -17.277344 14.394531 -18.085938 13.34375 -18.5 C 12.578125 -18.789063 11.355469 -18.9375 9.667969 -18.941406 L 5.15625 -18.941406 Z M 5.15625 -2.535156 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-2\">\n",
       "<path style=\"stroke:none;\" d=\"M 17.636719 -7.53125 L 20.476563 -6.8125 C 19.878906 -4.476563 18.808594 -2.695313 17.261719 -1.472656 C 15.714844 -0.246094 13.820313 0.363281 11.585938 0.367188 C 9.269531 0.363281 7.390625 -0.105469 5.941406 -1.046875 C 4.488281 -1.988281 3.382813 -3.351563 2.628906 -5.140625 C 1.867188 -6.921875 1.488281 -8.84375 1.492188 -10.898438 C 1.488281 -13.132813 1.914063 -15.082031 2.773438 -16.75 C 3.625 -18.414063 4.84375 -19.679688 6.421875 -20.542969 C 8 -21.40625 9.734375 -21.835938 11.632813 -21.839844 C 13.773438 -21.835938 15.582031 -21.289063 17.050781 -20.199219 C 18.515625 -19.101563 19.535156 -17.566406 20.113281 -15.585938 L 17.3125 -14.925781 C 16.8125 -16.484375 16.089844 -17.625 15.144531 -18.339844 C 14.195313 -19.050781 13.003906 -19.40625 11.570313 -19.410156 C 9.917969 -19.40625 8.539063 -19.011719 7.433594 -18.222656 C 6.324219 -17.429688 5.546875 -16.367188 5.097656 -15.035156 C 4.644531 -13.703125 4.417969 -12.328125 4.421875 -10.914063 C 4.417969 -9.082031 4.683594 -7.488281 5.21875 -6.128906 C 5.75 -4.765625 6.578125 -3.746094 7.703125 -3.074219 C 8.824219 -2.398438 10.039063 -2.0625 11.351563 -2.066406 C 12.941406 -2.0625 14.289063 -2.519531 15.394531 -3.441406 C 16.496094 -4.355469 17.246094 -5.71875 17.636719 -7.53125 Z M 17.636719 -7.53125 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-3\">\n",
       "<path style=\"stroke:none;\" d=\"M 11.175781 0 L 8.539063 0 L 8.539063 -16.800781 C 7.898438 -16.191406 7.066406 -15.585938 6.039063 -14.984375 C 5.007813 -14.375 4.082031 -13.921875 3.265625 -13.625 L 3.265625 -16.171875 C 4.734375 -16.863281 6.023438 -17.703125 7.132813 -18.691406 C 8.234375 -19.675781 9.015625 -20.632813 9.476563 -21.5625 L 11.175781 -21.5625 Z M 11.175781 0 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-4\">\n",
       "<path style=\"stroke:none;\" d=\"M 2.226563 0 L 2.226563 -21.476563 L 6.503906 -21.476563 L 11.585938 -6.269531 C 12.050781 -4.851563 12.394531 -3.789063 12.613281 -3.089844 C 12.851563 -3.867188 13.230469 -5.015625 13.753906 -6.53125 L 18.898438 -21.476563 L 22.71875 -21.476563 L 22.71875 0 L 19.980469 0 L 19.980469 -17.972656 L 13.742188 0 L 11.175781 0 L 4.964844 -18.28125 L 4.964844 0 Z M 2.226563 0 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-5\">\n",
       "<path style=\"stroke:none;\" d=\"M 12.128906 -1.917969 C 11.152344 -1.082031 10.210938 -0.496094 9.308594 -0.160156 C 8.402344 0.183594 7.433594 0.351563 6.402344 0.351563 C 4.6875 0.351563 3.375 -0.0625 2.460938 -0.898438 C 1.542969 -1.730469 1.082031 -2.800781 1.085938 -4.101563 C 1.082031 -4.859375 1.257813 -5.554688 1.605469 -6.1875 C 1.949219 -6.816406 2.402344 -7.320313 2.964844 -7.703125 C 3.527344 -8.082031 4.160156 -8.371094 4.863281 -8.570313 C 5.378906 -8.703125 6.160156 -8.835938 7.207031 -8.964844 C 9.332031 -9.214844 10.898438 -9.515625 11.910156 -9.875 C 11.914063 -10.230469 11.917969 -10.460938 11.921875 -10.5625 C 11.917969 -11.632813 11.667969 -12.390625 11.175781 -12.832031 C 10.496094 -13.425781 9.496094 -13.722656 8.171875 -13.726563 C 6.929688 -13.722656 6.015625 -13.503906 5.425781 -13.074219 C 4.832031 -12.636719 4.394531 -11.867188 4.117188 -10.765625 L 1.539063 -11.117188 C 1.769531 -12.21875 2.15625 -13.109375 2.695313 -13.789063 C 3.230469 -14.464844 4.003906 -14.988281 5.023438 -15.355469 C 6.035156 -15.71875 7.214844 -15.902344 8.554688 -15.90625 C 9.882813 -15.902344 10.960938 -15.746094 11.792969 -15.4375 C 12.621094 -15.121094 13.230469 -14.726563 13.625 -14.257813 C 14.011719 -13.78125 14.285156 -13.183594 14.445313 -12.464844 C 14.527344 -12.011719 14.570313 -11.203125 14.574219 -10.035156 L 14.574219 -6.519531 C 14.570313 -4.066406 14.628906 -2.515625 14.742188 -1.867188 C 14.855469 -1.214844 15.078125 -0.59375 15.410156 0 L 12.65625 0 C 12.382813 -0.546875 12.207031 -1.183594 12.128906 -1.917969 Z M 11.910156 -7.808594 C 10.949219 -7.414063 9.511719 -7.082031 7.601563 -6.8125 C 6.515625 -6.652344 5.75 -6.476563 5.300781 -6.285156 C 4.851563 -6.085938 4.503906 -5.800781 4.261719 -5.425781 C 4.015625 -5.050781 3.894531 -4.632813 3.898438 -4.175781 C 3.894531 -3.46875 4.160156 -2.882813 4.695313 -2.417969 C 5.222656 -1.945313 6.003906 -1.710938 7.03125 -1.714844 C 8.046875 -1.710938 8.949219 -1.933594 9.742188 -2.378906 C 10.53125 -2.824219 11.113281 -3.429688 11.484375 -4.203125 C 11.765625 -4.796875 11.90625 -5.675781 11.910156 -6.839844 Z M 11.910156 -7.808594 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-6\">\n",
       "<path style=\"stroke:none;\" d=\"M 12.128906 -5.699219 L 14.722656 -5.359375 C 14.4375 -3.570313 13.710938 -2.171875 12.546875 -1.164063 C 11.375 -0.152344 9.941406 0.351563 8.246094 0.351563 C 6.113281 0.351563 4.402344 -0.34375 3.113281 -1.734375 C 1.816406 -3.125 1.171875 -5.117188 1.171875 -7.71875 C 1.171875 -9.394531 1.449219 -10.867188 2.007813 -12.128906 C 2.5625 -13.386719 3.410156 -14.328125 4.546875 -14.960938 C 5.683594 -15.585938 6.921875 -15.902344 8.261719 -15.90625 C 9.949219 -15.902344 11.328125 -15.476563 12.40625 -14.625 C 13.476563 -13.769531 14.167969 -12.554688 14.472656 -10.984375 L 11.910156 -10.589844 C 11.664063 -11.628906 11.230469 -12.417969 10.613281 -12.949219 C 9.988281 -13.476563 9.238281 -13.738281 8.363281 -13.742188 C 7.03125 -13.738281 5.953125 -13.261719 5.125 -12.3125 C 4.292969 -11.355469 3.878906 -9.851563 3.882813 -7.792969 C 3.878906 -5.699219 4.277344 -4.179688 5.082031 -3.234375 C 5.878906 -2.289063 6.925781 -1.816406 8.21875 -1.816406 C 9.25 -1.816406 10.113281 -2.132813 10.808594 -2.765625 C 11.5 -3.398438 11.941406 -4.375 12.128906 -5.699219 Z M 12.128906 -5.699219 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-7\">\n",
       "<path style=\"stroke:none;\" d=\"M 1.949219 0 L 1.949219 -15.554688 L 4.320313 -15.554688 L 4.320313 -13.199219 C 4.921875 -14.296875 5.480469 -15.023438 5.996094 -15.378906 C 6.503906 -15.726563 7.070313 -15.902344 7.691406 -15.90625 C 8.578125 -15.902344 9.480469 -15.621094 10.398438 -15.058594 L 9.492188 -12.613281 C 8.84375 -12.992188 8.199219 -13.183594 7.558594 -13.183594 C 6.980469 -13.183594 6.460938 -13.007813 6.003906 -12.664063 C 5.542969 -12.3125 5.214844 -11.832031 5.023438 -11.21875 C 4.726563 -10.277344 4.582031 -9.253906 4.585938 -8.144531 L 4.585938 0 Z M 1.949219 0 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-8\">\n",
       "<path style=\"stroke:none;\" d=\"M 0.996094 -7.777344 C 0.996094 -10.65625 1.796875 -12.792969 3.398438 -14.179688 C 4.734375 -15.328125 6.363281 -15.902344 8.289063 -15.90625 C 10.425781 -15.902344 12.175781 -15.203125 13.535156 -13.804688 C 14.890625 -12.402344 15.566406 -10.464844 15.570313 -8 C 15.566406 -5.992188 15.265625 -4.417969 14.667969 -3.273438 C 14.066406 -2.125 13.195313 -1.234375 12.046875 -0.601563 C 10.898438 0.0351563 9.644531 0.351563 8.289063 0.351563 C 6.109375 0.351563 4.347656 -0.34375 3.007813 -1.742188 C 1.664063 -3.132813 0.996094 -5.144531 0.996094 -7.777344 Z M 3.707031 -7.777344 C 3.703125 -5.78125 4.136719 -4.289063 5.007813 -3.300781 C 5.875 -2.308594 6.96875 -1.816406 8.289063 -1.816406 C 9.597656 -1.816406 10.6875 -2.3125 11.554688 -3.308594 C 12.421875 -4.304688 12.855469 -5.824219 12.859375 -7.867188 C 12.855469 -9.785156 12.417969 -11.242188 11.546875 -12.238281 C 10.671875 -13.226563 9.585938 -13.722656 8.289063 -13.726563 C 6.96875 -13.722656 5.875 -13.230469 5.007813 -12.246094 C 4.136719 -11.257813 3.703125 -9.765625 3.707031 -7.777344 Z M 3.707031 -7.777344 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-9\">\n",
       "<path style=\"stroke:none;\" d=\"M 1.976563 5.960938 L 1.976563 -15.554688 L 4.378906 -15.554688 L 4.378906 -13.535156 C 4.941406 -14.324219 5.582031 -14.914063 6.296875 -15.3125 C 7.007813 -15.703125 7.871094 -15.902344 8.890625 -15.90625 C 10.214844 -15.902344 11.386719 -15.5625 12.40625 -14.882813 C 13.417969 -14.199219 14.1875 -13.234375 14.707031 -11.988281 C 15.222656 -10.742188 15.480469 -9.375 15.484375 -7.894531 C 15.480469 -6.300781 15.195313 -4.867188 14.625 -3.59375 C 14.054688 -2.316406 13.222656 -1.339844 12.136719 -0.664063 C 11.042969 0.015625 9.898438 0.351563 8.703125 0.351563 C 7.820313 0.351563 7.03125 0.167969 6.335938 -0.203125 C 5.632813 -0.574219 5.058594 -1.042969 4.613281 -1.609375 L 4.613281 5.960938 Z M 4.367188 -7.691406 C 4.363281 -5.6875 4.769531 -4.207031 5.582031 -3.25 C 6.390625 -2.292969 7.371094 -1.816406 8.523438 -1.816406 C 9.691406 -1.816406 10.695313 -2.308594 11.535156 -3.300781 C 12.367188 -4.289063 12.785156 -5.828125 12.789063 -7.910156 C 12.785156 -9.890625 12.378906 -11.375 11.566406 -12.363281 C 10.75 -13.347656 9.773438 -13.839844 8.640625 -13.84375 C 7.515625 -13.839844 6.523438 -13.316406 5.660156 -12.269531 C 4.796875 -11.21875 4.363281 -9.691406 4.367188 -7.691406 Z M 4.367188 -7.691406 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-10\">\n",
       "<path style=\"stroke:none;\" d=\"M 1.976563 0 L 1.976563 -21.476563 L 4.613281 -21.476563 L 4.613281 -13.769531 C 5.839844 -15.191406 7.394531 -15.902344 9.273438 -15.90625 C 10.421875 -15.902344 11.421875 -15.675781 12.273438 -15.226563 C 13.121094 -14.769531 13.730469 -14.140625 14.097656 -13.34375 C 14.464844 -12.539063 14.648438 -11.378906 14.648438 -9.859375 L 14.648438 0 L 12.011719 0 L 12.011719 -9.859375 C 12.011719 -11.175781 11.722656 -12.132813 11.152344 -12.734375 C 10.578125 -13.332031 9.773438 -13.632813 8.730469 -13.636719 C 7.949219 -13.632813 7.210938 -13.429688 6.523438 -13.027344 C 5.832031 -12.621094 5.34375 -12.074219 5.050781 -11.382813 C 4.757813 -10.6875 4.609375 -9.730469 4.613281 -8.511719 L 4.613281 0 Z M 1.976563 0 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-11\">\n",
       "<path style=\"stroke:none;\" d=\"M 1.492188 1.289063 L 4.058594 1.671875 C 4.164063 2.457031 4.460938 3.035156 4.953125 3.398438 C 5.601563 3.886719 6.496094 4.128906 7.632813 4.132813 C 8.847656 4.128906 9.789063 3.886719 10.457031 3.398438 C 11.117188 2.90625 11.566406 2.222656 11.804688 1.347656 C 11.941406 0.808594 12.003906 -0.316406 11.996094 -2.035156 C 10.839844 -0.675781 9.40625 0 7.691406 0 C 5.550781 0 3.894531 -0.769531 2.726563 -2.3125 C 1.550781 -3.855469 0.964844 -5.707031 0.96875 -7.867188 C 0.964844 -9.347656 1.234375 -10.71875 1.773438 -11.976563 C 2.308594 -13.230469 3.085938 -14.199219 4.109375 -14.882813 C 5.125 -15.5625 6.324219 -15.902344 7.703125 -15.90625 C 9.539063 -15.902344 11.054688 -15.160156 12.246094 -13.679688 L 12.246094 -15.554688 L 14.679688 -15.554688 L 14.679688 -2.109375 C 14.675781 0.3125 14.429688 2.027344 13.9375 3.039063 C 13.441406 4.046875 12.660156 4.84375 11.59375 5.433594 C 10.523438 6.015625 9.207031 6.308594 7.648438 6.3125 C 5.789063 6.308594 4.289063 5.890625 3.148438 5.058594 C 2.003906 4.222656 1.453125 2.96875 1.492188 1.289063 Z M 3.675781 -8.054688 C 3.671875 -6.011719 4.078125 -4.523438 4.890625 -3.589844 C 5.699219 -2.648438 6.714844 -2.179688 7.9375 -2.183594 C 9.148438 -2.179688 10.164063 -2.644531 10.988281 -3.582031 C 11.804688 -4.511719 12.214844 -5.976563 12.21875 -7.96875 C 12.214844 -9.871094 11.792969 -11.304688 10.949219 -12.273438 C 10.101563 -13.238281 9.082031 -13.722656 7.894531 -13.726563 C 6.71875 -13.722656 5.722656 -13.246094 4.90625 -12.296875 C 4.082031 -11.339844 3.671875 -9.925781 3.675781 -8.054688 Z M 3.675781 -8.054688 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-12\">\n",
       "<path style=\"stroke:none;\" d=\"M 12.625 -5.007813 L 15.351563 -4.671875 C 14.921875 -3.078125 14.125 -1.84375 12.960938 -0.964844 C 11.796875 -0.0859375 10.3125 0.351563 8.511719 0.351563 C 6.234375 0.351563 4.429688 -0.347656 3.097656 -1.75 C 1.761719 -3.148438 1.09375 -5.113281 1.097656 -7.648438 C 1.09375 -10.257813 1.769531 -12.289063 3.117188 -13.738281 C 4.464844 -15.179688 6.210938 -15.902344 8.363281 -15.90625 C 10.441406 -15.902344 12.140625 -15.195313 13.460938 -13.78125 C 14.773438 -12.367188 15.433594 -10.375 15.4375 -7.808594 C 15.433594 -7.648438 15.429688 -7.414063 15.425781 -7.105469 L 3.824219 -7.105469 C 3.917969 -5.390625 4.402344 -4.082031 5.273438 -3.175781 C 6.140625 -2.269531 7.222656 -1.816406 8.523438 -1.816406 C 9.488281 -1.816406 10.3125 -2.070313 11 -2.578125 C 11.679688 -3.085938 12.222656 -3.894531 12.625 -5.007813 Z M 3.96875 -9.273438 L 12.65625 -9.273438 C 12.539063 -10.578125 12.207031 -11.558594 11.660156 -12.21875 C 10.816406 -13.230469 9.726563 -13.738281 8.394531 -13.742188 C 7.179688 -13.738281 6.160156 -13.332031 5.339844 -12.523438 C 4.511719 -11.710938 4.054688 -10.628906 3.96875 -9.273438 Z M 3.96875 -9.273438 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-13\">\n",
       "<path style=\"stroke:none;\" d=\"M 12.363281 -8.421875 L 12.363281 -10.941406 L 21.460938 -10.957031 L 21.460938 -2.988281 C 20.0625 -1.875 18.621094 -1.035156 17.140625 -0.476563 C 15.652344 0.0859375 14.128906 0.363281 12.570313 0.367188 C 10.457031 0.363281 8.542969 -0.0859375 6.820313 -0.988281 C 5.09375 -1.890625 3.789063 -3.195313 2.914063 -4.90625 C 2.03125 -6.609375 1.59375 -8.519531 1.597656 -10.632813 C 1.59375 -12.722656 2.03125 -14.671875 2.90625 -16.484375 C 3.78125 -18.292969 5.039063 -19.640625 6.679688 -20.519531 C 8.320313 -21.398438 10.207031 -21.835938 12.347656 -21.839844 C 13.898438 -21.835938 15.304688 -21.585938 16.558594 -21.085938 C 17.8125 -20.582031 18.796875 -19.882813 19.511719 -18.984375 C 20.222656 -18.085938 20.761719 -16.914063 21.136719 -15.46875 L 18.574219 -14.765625 C 18.25 -15.859375 17.847656 -16.71875 17.371094 -17.34375 C 16.890625 -17.96875 16.207031 -18.46875 15.320313 -18.84375 C 14.429688 -19.21875 13.445313 -19.40625 12.363281 -19.410156 C 11.0625 -19.40625 9.9375 -19.207031 8.992188 -18.816406 C 8.042969 -18.417969 7.277344 -17.898438 6.699219 -17.253906 C 6.113281 -16.609375 5.664063 -15.902344 5.347656 -15.132813 C 4.796875 -13.800781 4.523438 -12.359375 4.527344 -10.8125 C 4.523438 -8.890625 4.851563 -7.289063 5.515625 -6.003906 C 6.171875 -4.710938 7.132813 -3.753906 8.390625 -3.132813 C 9.648438 -2.503906 10.984375 -2.191406 12.40625 -2.195313 C 13.632813 -2.191406 14.835938 -2.429688 16.011719 -2.90625 C 17.179688 -3.378906 18.066406 -3.882813 18.675781 -4.421875 L 18.675781 -8.421875 Z M 12.363281 -8.421875 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-14\">\n",
       "<path style=\"stroke:none;\" d=\"M 4.410156 0 L 1.960938 0 L 1.960938 -21.476563 L 4.601563 -21.476563 L 4.601563 -13.8125 C 5.710938 -15.203125 7.132813 -15.902344 8.863281 -15.90625 C 9.816406 -15.902344 10.722656 -15.710938 11.578125 -15.328125 C 12.429688 -14.941406 13.132813 -14.398438 13.6875 -13.703125 C 14.238281 -13.003906 14.667969 -12.160156 14.984375 -11.175781 C 15.292969 -10.183594 15.449219 -9.128906 15.453125 -8.011719 C 15.449219 -5.34375 14.792969 -3.285156 13.476563 -1.832031 C 12.15625 -0.375 10.574219 0.351563 8.730469 0.351563 C 6.894531 0.351563 5.453125 -0.414063 4.410156 -1.949219 Z M 4.378906 -7.894531 C 4.375 -6.027344 4.628906 -4.679688 5.140625 -3.851563 C 5.96875 -2.492188 7.09375 -1.816406 8.511719 -1.816406 C 9.660156 -1.816406 10.65625 -2.316406 11.496094 -3.316406 C 12.335938 -4.316406 12.753906 -5.808594 12.757813 -7.792969 C 12.753906 -9.824219 12.351563 -11.320313 11.550781 -12.289063 C 10.742188 -13.253906 9.769531 -13.738281 8.628906 -13.742188 C 7.472656 -13.738281 6.476563 -13.238281 5.636719 -12.238281 C 4.796875 -11.234375 4.375 -9.785156 4.378906 -7.894531 Z M 4.378906 -7.894531 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-15\">\n",
       "<path style=\"stroke:none;\" d=\"M 1.917969 0 L 1.917969 -21.476563 L 4.554688 -21.476563 L 4.554688 0 Z M 1.917969 0 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-16\">\n",
       "<path style=\"stroke:none;\" d=\"M 7.734375 -2.359375 L 8.117188 -0.03125 C 7.367188 0.125 6.703125 0.199219 6.125 0.203125 C 5.164063 0.199219 4.421875 0.0507813 3.894531 -0.25 C 3.367188 -0.550781 2.996094 -0.949219 2.78125 -1.441406 C 2.566406 -1.933594 2.460938 -2.96875 2.460938 -4.554688 L 2.460938 -13.507813 L 0.527344 -13.507813 L 0.527344 -15.554688 L 2.460938 -15.554688 L 2.460938 -19.410156 L 5.082031 -20.992188 L 5.082031 -15.554688 L 7.734375 -15.554688 L 7.734375 -13.507813 L 5.082031 -13.507813 L 5.082031 -4.410156 C 5.078125 -3.65625 5.125 -3.171875 5.21875 -2.957031 C 5.308594 -2.742188 5.460938 -2.570313 5.675781 -2.445313 C 5.882813 -2.316406 6.183594 -2.253906 6.578125 -2.257813 C 6.867188 -2.253906 7.253906 -2.289063 7.734375 -2.359375 Z M 7.734375 -2.359375 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-17\">\n",
       "<path style=\"stroke:none;\" d=\"M 1.859375 5.992188 L 1.566406 3.515625 C 2.136719 3.671875 2.640625 3.75 3.078125 3.75 C 3.65625 3.75 4.125 3.652344 4.480469 3.457031 C 4.828125 3.261719 5.117188 2.988281 5.347656 2.636719 C 5.507813 2.371094 5.777344 1.714844 6.152344 0.671875 C 6.199219 0.523438 6.277344 0.308594 6.386719 0.03125 L 0.484375 -15.554688 L 3.324219 -15.554688 L 6.5625 -6.546875 C 6.980469 -5.402344 7.355469 -4.203125 7.691406 -2.945313 C 7.992188 -4.152344 8.351563 -5.332031 8.773438 -6.488281 L 12.101563 -15.554688 L 14.734375 -15.554688 L 8.820313 0.265625 C 8.179688 1.96875 7.6875 3.144531 7.339844 3.792969 C 6.867188 4.660156 6.328125 5.296875 5.726563 5.703125 C 5.117188 6.105469 4.394531 6.308594 3.558594 6.3125 C 3.046875 6.308594 2.480469 6.203125 1.859375 5.992188 Z M 1.859375 5.992188 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-18\">\n",
       "<path style=\"stroke:none;\" d=\"M 2.609375 0 L 2.609375 -13.507813 L 0.277344 -13.507813 L 0.277344 -15.554688 L 2.609375 -15.554688 L 2.609375 -17.210938 C 2.605469 -18.253906 2.699219 -19.03125 2.886719 -19.539063 C 3.136719 -20.222656 3.582031 -20.777344 4.226563 -21.203125 C 4.863281 -21.625 5.761719 -21.835938 6.914063 -21.839844 C 7.65625 -21.835938 8.476563 -21.75 9.375 -21.578125 L 8.980469 -19.277344 C 8.429688 -19.371094 7.910156 -19.417969 7.425781 -19.421875 C 6.621094 -19.417969 6.054688 -19.246094 5.726563 -18.910156 C 5.390625 -18.566406 5.226563 -17.929688 5.230469 -16.992188 L 5.230469 -15.554688 L 8.261719 -15.554688 L 8.261719 -13.507813 L 5.230469 -13.507813 L 5.230469 0 Z M 2.609375 0 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-19\">\n",
       "<path style=\"stroke:none;\" d=\"M 1.992188 -18.441406 L 1.992188 -21.476563 L 4.628906 -21.476563 L 4.628906 -18.441406 Z M 1.992188 0 L 1.992188 -15.554688 L 4.628906 -15.554688 L 4.628906 0 Z M 1.992188 0 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-20\">\n",
       "<path style=\"stroke:none;\" d=\"M 0.921875 -4.644531 L 3.53125 -5.054688 C 3.675781 -4.003906 4.082031 -3.203125 4.753906 -2.648438 C 5.417969 -2.09375 6.355469 -1.816406 7.558594 -1.816406 C 8.769531 -1.816406 9.667969 -2.0625 10.253906 -2.554688 C 10.839844 -3.046875 11.132813 -3.625 11.132813 -4.292969 C 11.132813 -4.886719 10.871094 -5.355469 10.355469 -5.699219 C 9.988281 -5.929688 9.089844 -6.226563 7.660156 -6.59375 C 5.722656 -7.078125 4.382813 -7.5 3.640625 -7.859375 C 2.890625 -8.210938 2.324219 -8.703125 1.941406 -9.335938 C 1.550781 -9.960938 1.359375 -10.660156 1.363281 -11.425781 C 1.359375 -12.117188 1.515625 -12.757813 1.835938 -13.351563 C 2.148438 -13.941406 2.582031 -14.433594 3.132813 -14.824219 C 3.542969 -15.125 4.101563 -15.378906 4.8125 -15.589844 C 5.515625 -15.796875 6.277344 -15.902344 7.089844 -15.90625 C 8.308594 -15.902344 9.378906 -15.726563 10.304688 -15.378906 C 11.222656 -15.023438 11.902344 -14.546875 12.347656 -13.949219 C 12.785156 -13.347656 13.089844 -12.546875 13.257813 -11.542969 L 10.679688 -11.191406 C 10.558594 -11.988281 10.21875 -12.613281 9.660156 -13.066406 C 9.097656 -13.515625 8.304688 -13.738281 7.28125 -13.742188 C 6.066406 -13.738281 5.203125 -13.539063 4.6875 -13.140625 C 4.167969 -12.738281 3.90625 -12.269531 3.910156 -11.734375 C 3.90625 -11.386719 4.015625 -11.078125 4.234375 -10.8125 C 4.445313 -10.523438 4.78125 -10.289063 5.242188 -10.109375 C 5.503906 -10.007813 6.28125 -9.785156 7.574219 -9.433594 C 9.4375 -8.933594 10.738281 -8.523438 11.476563 -8.210938 C 12.210938 -7.890625 12.789063 -7.429688 13.210938 -6.824219 C 13.628906 -6.21875 13.839844 -5.46875 13.84375 -4.570313 C 13.839844 -3.6875 13.582031 -2.859375 13.074219 -2.085938 C 12.558594 -1.308594 11.820313 -0.707031 10.855469 -0.285156 C 9.886719 0.140625 8.792969 0.351563 7.574219 0.351563 C 5.546875 0.351563 4.003906 -0.0664063 2.949219 -0.90625 C 1.886719 -1.746094 1.210938 -2.992188 0.921875 -4.644531 Z M 0.921875 -4.644531 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-21\">\n",
       "<path style=\"stroke:none;\" d=\"M 7.777344 0 L 7.777344 -18.941406 L 0.703125 -18.941406 L 0.703125 -21.476563 L 17.726563 -21.476563 L 17.726563 -18.941406 L 10.621094 -18.941406 L 10.621094 0 Z M 7.777344 0 \"/>\n",
       "</symbol>\n",
       "</g>\n",
       "</defs>\n",
       "<g id=\"surface91\">\n",
       "<rect x=\"0\" y=\"0\" width=\"600\" height=\"600\" style=\"fill:rgb(100%,100%,100%);fill-opacity:1;stroke:none;\"/>\n",
       "<path style=\"fill:none;stroke-width:9.12;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(0%,29.803922%,40%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 347.214844 100 C 265.152344 58.304688 185.675781 65.640625 129.609375 138.113281 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(0%,29.803922%,40%);fill-opacity:1;\" d=\"M 112.152344 160.683594 L 122.277344 132.441406 L 136.945313 143.789063 L 112.152344 160.683594 \"/>\n",
       "<path style=\"fill:none;stroke-width:7.81;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(0%,29.803922%,40%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 347.214844 100 C 336.417969 188.464844 369.324219 258.621094 453.875 285.394531 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(0%,29.803922%,40%);fill-opacity:1;\" d=\"M 481.074219 294.007813 L 451.078125 294.234375 L 456.671875 276.558594 L 481.074219 294.007813 \"/>\n",
       "<path style=\"fill:none;stroke-width:7.31;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(0%,29.803922%,40%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 347.214844 100 C 196.5 130.230469 105.714844 221.667969 101.375 375.070313 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(0%,29.803922%,40%);fill-opacity:1;\" d=\"M 100.566406 403.589844 L 92.105469 374.808594 L 110.640625 375.332031 L 100.566406 403.589844 \"/>\n",
       "<path style=\"fill:none;stroke-width:5.57;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(0%,29.803922%,40%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 347.214844 100 C 247.613281 213.109375 230.984375 338.46875 318.855469 460.589844 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(0%,29.803922%,40%);fill-opacity:1;\" d=\"M 335.519531 483.75 L 311.332031 466.003906 L 326.382813 455.175781 L 335.519531 483.75 \"/>\n",
       "<path style=\"fill:none;stroke-width:5.68;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(0%,29.803922%,40%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 100 176.394531 C 182.0625 218.089844 261.539063 210.753906 317.601563 138.277344 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(0%,29.803922%,40%);fill-opacity:1;\" d=\"M 335.058594 115.710938 L 324.933594 143.949219 L 310.269531 132.605469 L 335.058594 115.710938 \"/>\n",
       "<path style=\"fill:none;stroke-width:6.89;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(0%,29.803922%,40%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 100 176.394531 C 183.214844 311.257813 304.03125 366.007813 453.976563 315.503906 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(0%,29.803922%,40%);fill-opacity:1;\" d=\"M 481.015625 306.394531 L 456.9375 324.289063 L 451.015625 306.714844 L 481.015625 306.394531 \"/>\n",
       "<path style=\"fill:none;stroke-width:6.17;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(0%,29.803922%,40%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 100 176.394531 C 38.847656 238.652344 22.414063 312.882813 72.242188 383.992188 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(0%,29.803922%,40%);fill-opacity:1;\" d=\"M 88.613281 407.355469 L 64.648438 389.308594 L 79.832031 378.671875 L 88.613281 407.355469 \"/>\n",
       "<path style=\"fill:none;stroke-width:5.01;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(0%,29.803922%,40%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 100 176.394531 C 89.535156 329.753906 153.886719 441.386719 300.746094 485.914063 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(0%,29.803922%,40%);fill-opacity:1;\" d=\"M 328.050781 494.191406 L 298.054688 494.785156 L 303.4375 477.039063 L 328.050781 494.191406 \"/>\n",
       "<path style=\"fill:none;stroke-width:4.57;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(0%,29.803922%,40%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 500 300 C 510.792969 211.535156 477.886719 141.378906 393.339844 114.605469 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(0%,29.803922%,40%);fill-opacity:1;\" d=\"M 366.136719 105.992188 L 396.136719 105.765625 L 390.539063 123.441406 L 366.136719 105.992188 \"/>\n",
       "<path style=\"fill:none;stroke-width:6.52;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(0%,29.803922%,40%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 500 300 C 416.785156 165.136719 295.96875 110.382813 146.023438 160.890625 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(0%,29.803922%,40%);fill-opacity:1;\" d=\"M 118.984375 170 L 143.0625 152.105469 L 148.984375 169.675781 L 118.984375 170 \"/>\n",
       "<path style=\"fill:none;stroke-width:4.83;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(0%,29.803922%,40%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 500 300 C 355.207031 235.59375 224.566406 258.546875 129.253906 384.84375 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(0%,29.803922%,40%);fill-opacity:1;\" d=\"M 112.066406 407.617188 L 121.855469 379.257813 L 136.652344 390.425781 L 112.066406 407.617188 \"/>\n",
       "<path style=\"fill:none;stroke-width:4.38;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(0%,29.803922%,40%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 500 300 C 411.8125 312.847656 352.773438 363.042969 349.175781 451.65625 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(0%,29.803922%,40%);fill-opacity:1;\" d=\"M 348.019531 480.164063 L 339.914063 451.28125 L 358.4375 452.035156 L 348.019531 480.164063 \"/>\n",
       "<path style=\"fill:none;stroke-width:4.37;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(0%,29.803922%,40%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 100 423.605469 C 250.714844 393.375 341.5 301.9375 345.839844 148.535156 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(0%,29.803922%,40%);fill-opacity:1;\" d=\"M 346.648438 120.015625 L 355.105469 148.800781 L 336.574219 148.273438 L 346.648438 120.015625 \"/>\n",
       "<path style=\"fill:none;stroke-width:5.58;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(0%,29.803922%,40%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 100 423.605469 C 161.152344 361.347656 177.585938 287.117188 127.757813 216.007813 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(0%,29.803922%,40%);fill-opacity:1;\" d=\"M 111.386719 192.644531 L 135.351563 210.691406 L 120.167969 221.328125 L 111.386719 192.644531 \"/>\n",
       "<path style=\"fill:none;stroke-width:4.83;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(0%,29.803922%,40%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 100 423.605469 C 244.792969 488.011719 375.433594 465.058594 470.746094 338.765625 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(0%,29.803922%,40%);fill-opacity:1;\" d=\"M 487.933594 315.988281 L 478.144531 344.347656 L 463.347656 333.179688 L 487.933594 315.988281 \"/>\n",
       "<path style=\"fill:none;stroke-width:3.87;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(0%,29.803922%,40%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 100 423.605469 C 144.234375 504.332031 213.992188 543.109375 301.167969 514.898438 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(0%,29.803922%,40%);fill-opacity:1;\" d=\"M 328.316406 506.117188 L 304.023438 523.71875 L 298.316406 506.078125 L 328.316406 506.117188 \"/>\n",
       "<path style=\"fill:none;stroke-width:4.91;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(0%,29.803922%,40%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 347.214844 500 C 446.8125 386.890625 463.441406 261.53125 375.570313 139.410156 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(0%,29.803922%,40%);fill-opacity:1;\" d=\"M 358.90625 116.25 L 383.09375 133.996094 L 368.046875 144.824219 L 358.90625 116.25 \"/>\n",
       "<path style=\"fill:none;stroke-width:8;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(0%,29.803922%,40%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 347.214844 500 C 357.679688 346.640625 293.328125 235.003906 146.46875 190.480469 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(0%,29.803922%,40%);fill-opacity:1;\" d=\"M 119.164063 182.203125 L 149.15625 181.609375 L 143.777344 199.351563 L 119.164063 182.203125 \"/>\n",
       "<path style=\"fill:none;stroke-width:6.82;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(0%,29.803922%,40%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 347.214844 500 C 435.402344 487.152344 494.441406 436.957031 498.039063 348.34375 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(0%,29.803922%,40%);fill-opacity:1;\" d=\"M 499.195313 319.835938 L 507.300781 348.71875 L 488.773438 347.964844 L 499.195313 319.835938 \"/>\n",
       "<path style=\"fill:none;stroke-width:6.16;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(0%,29.803922%,40%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 347.214844 500 C 302.980469 419.277344 233.222656 380.496094 146.042969 408.707031 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(0%,29.803922%,40%);fill-opacity:1;\" d=\"M 118.898438 417.492188 L 143.191406 399.886719 L 148.898438 417.527344 L 118.898438 417.492188 \"/>\n",
       "<path style=\"fill-rule:nonzero;fill:rgb(74.509804%,74.509804%,74.509804%);fill-opacity:1;stroke-width:1;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(0%,0%,0%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 367.214844 100 C 367.214844 111.046875 358.257813 120 347.214844 120 C 336.167969 120 327.214844 111.046875 327.214844 100 C 327.214844 88.953125 336.167969 80 347.214844 80 C 358.257813 80 367.214844 88.953125 367.214844 100 \"/>\n",
       "<path style=\"fill-rule:nonzero;fill:rgb(74.509804%,74.509804%,74.509804%);fill-opacity:1;stroke-width:1;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(0%,0%,0%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 120 176.394531 C 120 187.4375 111.046875 196.394531 100 196.394531 C 88.953125 196.394531 80 187.4375 80 176.394531 C 80 165.347656 88.953125 156.394531 100 156.394531 C 111.046875 156.394531 120 165.347656 120 176.394531 \"/>\n",
       "<path style=\"fill-rule:nonzero;fill:rgb(74.509804%,74.509804%,74.509804%);fill-opacity:1;stroke-width:1;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(0%,0%,0%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 520 300 C 520 311.046875 511.046875 320 500 320 C 488.953125 320 480 311.046875 480 300 C 480 288.953125 488.953125 280 500 280 C 511.046875 280 520 288.953125 520 300 \"/>\n",
       "<path style=\"fill-rule:nonzero;fill:rgb(74.509804%,74.509804%,74.509804%);fill-opacity:1;stroke-width:1;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(0%,0%,0%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 120 423.605469 C 120 434.652344 111.046875 443.605469 100 443.605469 C 88.953125 443.605469 80 434.652344 80 423.605469 C 80 412.5625 88.953125 403.605469 100 403.605469 C 111.046875 403.605469 120 412.5625 120 423.605469 \"/>\n",
       "<path style=\"fill-rule:nonzero;fill:rgb(74.509804%,74.509804%,74.509804%);fill-opacity:1;stroke-width:1;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(0%,0%,0%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 367.214844 500 C 367.214844 511.046875 358.257813 520 347.214844 520 C 336.167969 520 327.214844 511.046875 327.214844 500 C 327.214844 488.953125 336.167969 480 347.214844 480 C 358.257813 480 367.214844 488.953125 367.214844 500 \"/>\n",
       "<g style=\"fill:rgb(0%,0%,0%);fill-opacity:1;\">\n",
       "  <use xlink:href=\"#glyph0-1\" x=\"318.804688\" y=\"136.839844\"/>\n",
       "  <use xlink:href=\"#glyph0-2\" x=\"340.469727\" y=\"136.839844\"/>\n",
       "  <use xlink:href=\"#glyph0-3\" x=\"362.134766\" y=\"136.839844\"/>\n",
       "</g>\n",
       "<g style=\"fill:rgb(0%,0%,0%);fill-opacity:1;\">\n",
       "  <use xlink:href=\"#glyph0-4\" x=\"16.125\" y=\"212.867188\"/>\n",
       "  <use xlink:href=\"#glyph0-5\" x=\"41.115234\" y=\"212.867188\"/>\n",
       "  <use xlink:href=\"#glyph0-6\" x=\"57.799805\" y=\"212.867188\"/>\n",
       "  <use xlink:href=\"#glyph0-7\" x=\"72.799805\" y=\"212.867188\"/>\n",
       "  <use xlink:href=\"#glyph0-8\" x=\"82.790039\" y=\"212.867188\"/>\n",
       "  <use xlink:href=\"#glyph0-9\" x=\"99.474609\" y=\"212.867188\"/>\n",
       "  <use xlink:href=\"#glyph0-10\" x=\"116.15918\" y=\"212.867188\"/>\n",
       "  <use xlink:href=\"#glyph0-5\" x=\"132.84375\" y=\"212.867188\"/>\n",
       "  <use xlink:href=\"#glyph0-11\" x=\"149.52832\" y=\"212.867188\"/>\n",
       "  <use xlink:href=\"#glyph0-12\" x=\"166.212891\" y=\"212.867188\"/>\n",
       "</g>\n",
       "<g style=\"fill:rgb(0%,0%,0%);fill-opacity:1;\">\n",
       "  <use xlink:href=\"#glyph0-13\" x=\"455.117188\" y=\"336.839844\"/>\n",
       "  <use xlink:href=\"#glyph0-8\" x=\"478.452148\" y=\"336.839844\"/>\n",
       "  <use xlink:href=\"#glyph0-14\" x=\"495.136719\" y=\"336.839844\"/>\n",
       "  <use xlink:href=\"#glyph0-15\" x=\"511.821289\" y=\"336.839844\"/>\n",
       "  <use xlink:href=\"#glyph0-12\" x=\"518.486328\" y=\"336.839844\"/>\n",
       "  <use xlink:href=\"#glyph0-16\" x=\"535.170898\" y=\"336.839844\"/>\n",
       "</g>\n",
       "<g style=\"fill:rgb(0%,0%,0%);fill-opacity:1;\">\n",
       "  <use xlink:href=\"#glyph0-4\" x=\"9.796875\" y=\"460.449219\"/>\n",
       "  <use xlink:href=\"#glyph0-17\" x=\"34.787109\" y=\"460.449219\"/>\n",
       "  <use xlink:href=\"#glyph0-8\" x=\"49.787109\" y=\"460.449219\"/>\n",
       "  <use xlink:href=\"#glyph0-18\" x=\"66.47168\" y=\"460.449219\"/>\n",
       "  <use xlink:href=\"#glyph0-19\" x=\"74.806641\" y=\"460.449219\"/>\n",
       "  <use xlink:href=\"#glyph0-14\" x=\"81.47168\" y=\"460.449219\"/>\n",
       "  <use xlink:href=\"#glyph0-7\" x=\"98.15625\" y=\"460.449219\"/>\n",
       "  <use xlink:href=\"#glyph0-8\" x=\"108.146484\" y=\"460.449219\"/>\n",
       "  <use xlink:href=\"#glyph0-14\" x=\"124.831055\" y=\"460.449219\"/>\n",
       "  <use xlink:href=\"#glyph0-15\" x=\"141.515625\" y=\"460.449219\"/>\n",
       "  <use xlink:href=\"#glyph0-5\" x=\"148.180664\" y=\"460.449219\"/>\n",
       "  <use xlink:href=\"#glyph0-20\" x=\"164.865234\" y=\"460.449219\"/>\n",
       "  <use xlink:href=\"#glyph0-16\" x=\"179.865234\" y=\"460.449219\"/>\n",
       "</g>\n",
       "<g style=\"fill:rgb(0%,0%,0%);fill-opacity:1;\">\n",
       "  <use xlink:href=\"#glyph0-21\" x=\"317.023438\" y=\"536.476563\"/>\n",
       "  <use xlink:href=\"#glyph0-7\" x=\"335.348633\" y=\"536.476563\"/>\n",
       "  <use xlink:href=\"#glyph0-12\" x=\"345.338867\" y=\"536.476563\"/>\n",
       "  <use xlink:href=\"#glyph0-11\" x=\"362.023438\" y=\"536.476563\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<igraph.drawing.Plot at 0x19e8e8c5108>"
      ]
     },
     "execution_count": 38,
     "metadata": {
      "image/svg+xml": {
       "isolated": true
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ig.plot(g_healthy, \"healthy.png\", **visual_style)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can do the same for the uninflmaed UC cells."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Curve the edges\n",
    "visual_style[\"edge_curved\"] = True\n",
    "# Set the layout - we keep the same layout for comapriosn\n",
    "visual_style[\"layout\"] = my_layout\n",
    "#Add annoation\n",
    "visual_style[\"vertex_label\"] = g_uc.vs[\"name\"]\n",
    "#Calcualte the edge wheight relative the number of edges\n",
    "visual_style[\"edge_width\"] = 0.01 * np.array(g_uc.es[\"weight\"])\n",
    "\n",
    "#Setting the vertex visualistation parameters -colours and label position\n",
    "visual_style[\"vertex_label_size\"] = 30\n",
    "visual_style[\"vertex_color\"] = \"grey\"\n",
    "visual_style[\"vertex_label_dist\"] = 0.75\n",
    "visual_style[\"vertex_size\"] = 40\n",
    "\n",
    "#Setting the edge visualisation parameters\n",
    "visual_style[\"edge_color\"] = \"#CE1612\" #hexa code for red\n",
    "visual_style[\"edge_arrow_size\"] = 2\n",
    "\n",
    "#Setting the visualistation place. igraph need margin yo work\n",
    "visual_style[\"bbox\"] = (600, 600)\n",
    "visual_style[\"margin\"] = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n",
       "<svg xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"600pt\" height=\"600pt\" viewBox=\"0 0 600 600\" version=\"1.1\">\n",
       "<defs>\n",
       "<g>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-0\">\n",
       "<path style=\"stroke:none;\" d=\"M 3.75 0 L 3.75 -18.75 L 18.75 -18.75 L 18.75 0 Z M 4.21875 -0.46875 L 18.28125 -0.46875 L 18.28125 -18.28125 L 4.21875 -18.28125 Z M 4.21875 -0.46875 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-1\">\n",
       "<path style=\"stroke:none;\" d=\"M 2.3125 0 L 2.3125 -21.476563 L 9.710938 -21.476563 C 11.378906 -21.472656 12.65625 -21.371094 13.535156 -21.167969 C 14.765625 -20.882813 15.8125 -20.371094 16.683594 -19.628906 C 17.8125 -18.671875 18.660156 -17.449219 19.226563 -15.960938 C 19.785156 -14.46875 20.066406 -12.765625 20.070313 -10.855469 C 20.066406 -9.21875 19.875 -7.773438 19.496094 -6.515625 C 19.113281 -5.257813 18.625 -4.214844 18.03125 -3.390625 C 17.433594 -2.5625 16.785156 -1.914063 16.078125 -1.441406 C 15.367188 -0.96875 14.511719 -0.609375 13.511719 -0.367188 C 12.511719 -0.121094 11.359375 0 10.0625 0 Z M 5.15625 -2.535156 L 9.742188 -2.535156 C 11.15625 -2.53125 12.265625 -2.664063 13.074219 -2.929688 C 13.875 -3.191406 14.519531 -3.5625 15 -4.042969 C 15.671875 -4.714844 16.195313 -5.621094 16.574219 -6.761719 C 16.949219 -7.898438 17.136719 -9.277344 17.140625 -10.898438 C 17.136719 -13.140625 16.769531 -14.863281 16.03125 -16.074219 C 15.292969 -17.277344 14.394531 -18.085938 13.34375 -18.5 C 12.578125 -18.789063 11.355469 -18.9375 9.667969 -18.941406 L 5.15625 -18.941406 Z M 5.15625 -2.535156 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-2\">\n",
       "<path style=\"stroke:none;\" d=\"M 17.636719 -7.53125 L 20.476563 -6.8125 C 19.878906 -4.476563 18.808594 -2.695313 17.261719 -1.472656 C 15.714844 -0.246094 13.820313 0.363281 11.585938 0.367188 C 9.269531 0.363281 7.390625 -0.105469 5.941406 -1.046875 C 4.488281 -1.988281 3.382813 -3.351563 2.628906 -5.140625 C 1.867188 -6.921875 1.488281 -8.84375 1.492188 -10.898438 C 1.488281 -13.132813 1.914063 -15.082031 2.773438 -16.75 C 3.625 -18.414063 4.84375 -19.679688 6.421875 -20.542969 C 8 -21.40625 9.734375 -21.835938 11.632813 -21.839844 C 13.773438 -21.835938 15.582031 -21.289063 17.050781 -20.199219 C 18.515625 -19.101563 19.535156 -17.566406 20.113281 -15.585938 L 17.3125 -14.925781 C 16.8125 -16.484375 16.089844 -17.625 15.144531 -18.339844 C 14.195313 -19.050781 13.003906 -19.40625 11.570313 -19.410156 C 9.917969 -19.40625 8.539063 -19.011719 7.433594 -18.222656 C 6.324219 -17.429688 5.546875 -16.367188 5.097656 -15.035156 C 4.644531 -13.703125 4.417969 -12.328125 4.421875 -10.914063 C 4.417969 -9.082031 4.683594 -7.488281 5.21875 -6.128906 C 5.75 -4.765625 6.578125 -3.746094 7.703125 -3.074219 C 8.824219 -2.398438 10.039063 -2.0625 11.351563 -2.066406 C 12.941406 -2.0625 14.289063 -2.519531 15.394531 -3.441406 C 16.496094 -4.355469 17.246094 -5.71875 17.636719 -7.53125 Z M 17.636719 -7.53125 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-3\">\n",
       "<path style=\"stroke:none;\" d=\"M 11.175781 0 L 8.539063 0 L 8.539063 -16.800781 C 7.898438 -16.191406 7.066406 -15.585938 6.039063 -14.984375 C 5.007813 -14.375 4.082031 -13.921875 3.265625 -13.625 L 3.265625 -16.171875 C 4.734375 -16.863281 6.023438 -17.703125 7.132813 -18.691406 C 8.234375 -19.675781 9.015625 -20.632813 9.476563 -21.5625 L 11.175781 -21.5625 Z M 11.175781 0 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-4\">\n",
       "<path style=\"stroke:none;\" d=\"M 2.226563 0 L 2.226563 -21.476563 L 6.503906 -21.476563 L 11.585938 -6.269531 C 12.050781 -4.851563 12.394531 -3.789063 12.613281 -3.089844 C 12.851563 -3.867188 13.230469 -5.015625 13.753906 -6.53125 L 18.898438 -21.476563 L 22.71875 -21.476563 L 22.71875 0 L 19.980469 0 L 19.980469 -17.972656 L 13.742188 0 L 11.175781 0 L 4.964844 -18.28125 L 4.964844 0 Z M 2.226563 0 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-5\">\n",
       "<path style=\"stroke:none;\" d=\"M 12.128906 -1.917969 C 11.152344 -1.082031 10.210938 -0.496094 9.308594 -0.160156 C 8.402344 0.183594 7.433594 0.351563 6.402344 0.351563 C 4.6875 0.351563 3.375 -0.0625 2.460938 -0.898438 C 1.542969 -1.730469 1.082031 -2.800781 1.085938 -4.101563 C 1.082031 -4.859375 1.257813 -5.554688 1.605469 -6.1875 C 1.949219 -6.816406 2.402344 -7.320313 2.964844 -7.703125 C 3.527344 -8.082031 4.160156 -8.371094 4.863281 -8.570313 C 5.378906 -8.703125 6.160156 -8.835938 7.207031 -8.964844 C 9.332031 -9.214844 10.898438 -9.515625 11.910156 -9.875 C 11.914063 -10.230469 11.917969 -10.460938 11.921875 -10.5625 C 11.917969 -11.632813 11.667969 -12.390625 11.175781 -12.832031 C 10.496094 -13.425781 9.496094 -13.722656 8.171875 -13.726563 C 6.929688 -13.722656 6.015625 -13.503906 5.425781 -13.074219 C 4.832031 -12.636719 4.394531 -11.867188 4.117188 -10.765625 L 1.539063 -11.117188 C 1.769531 -12.21875 2.15625 -13.109375 2.695313 -13.789063 C 3.230469 -14.464844 4.003906 -14.988281 5.023438 -15.355469 C 6.035156 -15.71875 7.214844 -15.902344 8.554688 -15.90625 C 9.882813 -15.902344 10.960938 -15.746094 11.792969 -15.4375 C 12.621094 -15.121094 13.230469 -14.726563 13.625 -14.257813 C 14.011719 -13.78125 14.285156 -13.183594 14.445313 -12.464844 C 14.527344 -12.011719 14.570313 -11.203125 14.574219 -10.035156 L 14.574219 -6.519531 C 14.570313 -4.066406 14.628906 -2.515625 14.742188 -1.867188 C 14.855469 -1.214844 15.078125 -0.59375 15.410156 0 L 12.65625 0 C 12.382813 -0.546875 12.207031 -1.183594 12.128906 -1.917969 Z M 11.910156 -7.808594 C 10.949219 -7.414063 9.511719 -7.082031 7.601563 -6.8125 C 6.515625 -6.652344 5.75 -6.476563 5.300781 -6.285156 C 4.851563 -6.085938 4.503906 -5.800781 4.261719 -5.425781 C 4.015625 -5.050781 3.894531 -4.632813 3.898438 -4.175781 C 3.894531 -3.46875 4.160156 -2.882813 4.695313 -2.417969 C 5.222656 -1.945313 6.003906 -1.710938 7.03125 -1.714844 C 8.046875 -1.710938 8.949219 -1.933594 9.742188 -2.378906 C 10.53125 -2.824219 11.113281 -3.429688 11.484375 -4.203125 C 11.765625 -4.796875 11.90625 -5.675781 11.910156 -6.839844 Z M 11.910156 -7.808594 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-6\">\n",
       "<path style=\"stroke:none;\" d=\"M 12.128906 -5.699219 L 14.722656 -5.359375 C 14.4375 -3.570313 13.710938 -2.171875 12.546875 -1.164063 C 11.375 -0.152344 9.941406 0.351563 8.246094 0.351563 C 6.113281 0.351563 4.402344 -0.34375 3.113281 -1.734375 C 1.816406 -3.125 1.171875 -5.117188 1.171875 -7.71875 C 1.171875 -9.394531 1.449219 -10.867188 2.007813 -12.128906 C 2.5625 -13.386719 3.410156 -14.328125 4.546875 -14.960938 C 5.683594 -15.585938 6.921875 -15.902344 8.261719 -15.90625 C 9.949219 -15.902344 11.328125 -15.476563 12.40625 -14.625 C 13.476563 -13.769531 14.167969 -12.554688 14.472656 -10.984375 L 11.910156 -10.589844 C 11.664063 -11.628906 11.230469 -12.417969 10.613281 -12.949219 C 9.988281 -13.476563 9.238281 -13.738281 8.363281 -13.742188 C 7.03125 -13.738281 5.953125 -13.261719 5.125 -12.3125 C 4.292969 -11.355469 3.878906 -9.851563 3.882813 -7.792969 C 3.878906 -5.699219 4.277344 -4.179688 5.082031 -3.234375 C 5.878906 -2.289063 6.925781 -1.816406 8.21875 -1.816406 C 9.25 -1.816406 10.113281 -2.132813 10.808594 -2.765625 C 11.5 -3.398438 11.941406 -4.375 12.128906 -5.699219 Z M 12.128906 -5.699219 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-7\">\n",
       "<path style=\"stroke:none;\" d=\"M 1.949219 0 L 1.949219 -15.554688 L 4.320313 -15.554688 L 4.320313 -13.199219 C 4.921875 -14.296875 5.480469 -15.023438 5.996094 -15.378906 C 6.503906 -15.726563 7.070313 -15.902344 7.691406 -15.90625 C 8.578125 -15.902344 9.480469 -15.621094 10.398438 -15.058594 L 9.492188 -12.613281 C 8.84375 -12.992188 8.199219 -13.183594 7.558594 -13.183594 C 6.980469 -13.183594 6.460938 -13.007813 6.003906 -12.664063 C 5.542969 -12.3125 5.214844 -11.832031 5.023438 -11.21875 C 4.726563 -10.277344 4.582031 -9.253906 4.585938 -8.144531 L 4.585938 0 Z M 1.949219 0 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-8\">\n",
       "<path style=\"stroke:none;\" d=\"M 0.996094 -7.777344 C 0.996094 -10.65625 1.796875 -12.792969 3.398438 -14.179688 C 4.734375 -15.328125 6.363281 -15.902344 8.289063 -15.90625 C 10.425781 -15.902344 12.175781 -15.203125 13.535156 -13.804688 C 14.890625 -12.402344 15.566406 -10.464844 15.570313 -8 C 15.566406 -5.992188 15.265625 -4.417969 14.667969 -3.273438 C 14.066406 -2.125 13.195313 -1.234375 12.046875 -0.601563 C 10.898438 0.0351563 9.644531 0.351563 8.289063 0.351563 C 6.109375 0.351563 4.347656 -0.34375 3.007813 -1.742188 C 1.664063 -3.132813 0.996094 -5.144531 0.996094 -7.777344 Z M 3.707031 -7.777344 C 3.703125 -5.78125 4.136719 -4.289063 5.007813 -3.300781 C 5.875 -2.308594 6.96875 -1.816406 8.289063 -1.816406 C 9.597656 -1.816406 10.6875 -2.3125 11.554688 -3.308594 C 12.421875 -4.304688 12.855469 -5.824219 12.859375 -7.867188 C 12.855469 -9.785156 12.417969 -11.242188 11.546875 -12.238281 C 10.671875 -13.226563 9.585938 -13.722656 8.289063 -13.726563 C 6.96875 -13.722656 5.875 -13.230469 5.007813 -12.246094 C 4.136719 -11.257813 3.703125 -9.765625 3.707031 -7.777344 Z M 3.707031 -7.777344 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-9\">\n",
       "<path style=\"stroke:none;\" d=\"M 1.976563 5.960938 L 1.976563 -15.554688 L 4.378906 -15.554688 L 4.378906 -13.535156 C 4.941406 -14.324219 5.582031 -14.914063 6.296875 -15.3125 C 7.007813 -15.703125 7.871094 -15.902344 8.890625 -15.90625 C 10.214844 -15.902344 11.386719 -15.5625 12.40625 -14.882813 C 13.417969 -14.199219 14.1875 -13.234375 14.707031 -11.988281 C 15.222656 -10.742188 15.480469 -9.375 15.484375 -7.894531 C 15.480469 -6.300781 15.195313 -4.867188 14.625 -3.59375 C 14.054688 -2.316406 13.222656 -1.339844 12.136719 -0.664063 C 11.042969 0.015625 9.898438 0.351563 8.703125 0.351563 C 7.820313 0.351563 7.03125 0.167969 6.335938 -0.203125 C 5.632813 -0.574219 5.058594 -1.042969 4.613281 -1.609375 L 4.613281 5.960938 Z M 4.367188 -7.691406 C 4.363281 -5.6875 4.769531 -4.207031 5.582031 -3.25 C 6.390625 -2.292969 7.371094 -1.816406 8.523438 -1.816406 C 9.691406 -1.816406 10.695313 -2.308594 11.535156 -3.300781 C 12.367188 -4.289063 12.785156 -5.828125 12.789063 -7.910156 C 12.785156 -9.890625 12.378906 -11.375 11.566406 -12.363281 C 10.75 -13.347656 9.773438 -13.839844 8.640625 -13.84375 C 7.515625 -13.839844 6.523438 -13.316406 5.660156 -12.269531 C 4.796875 -11.21875 4.363281 -9.691406 4.367188 -7.691406 Z M 4.367188 -7.691406 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-10\">\n",
       "<path style=\"stroke:none;\" d=\"M 1.976563 0 L 1.976563 -21.476563 L 4.613281 -21.476563 L 4.613281 -13.769531 C 5.839844 -15.191406 7.394531 -15.902344 9.273438 -15.90625 C 10.421875 -15.902344 11.421875 -15.675781 12.273438 -15.226563 C 13.121094 -14.769531 13.730469 -14.140625 14.097656 -13.34375 C 14.464844 -12.539063 14.648438 -11.378906 14.648438 -9.859375 L 14.648438 0 L 12.011719 0 L 12.011719 -9.859375 C 12.011719 -11.175781 11.722656 -12.132813 11.152344 -12.734375 C 10.578125 -13.332031 9.773438 -13.632813 8.730469 -13.636719 C 7.949219 -13.632813 7.210938 -13.429688 6.523438 -13.027344 C 5.832031 -12.621094 5.34375 -12.074219 5.050781 -11.382813 C 4.757813 -10.6875 4.609375 -9.730469 4.613281 -8.511719 L 4.613281 0 Z M 1.976563 0 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-11\">\n",
       "<path style=\"stroke:none;\" d=\"M 1.492188 1.289063 L 4.058594 1.671875 C 4.164063 2.457031 4.460938 3.035156 4.953125 3.398438 C 5.601563 3.886719 6.496094 4.128906 7.632813 4.132813 C 8.847656 4.128906 9.789063 3.886719 10.457031 3.398438 C 11.117188 2.90625 11.566406 2.222656 11.804688 1.347656 C 11.941406 0.808594 12.003906 -0.316406 11.996094 -2.035156 C 10.839844 -0.675781 9.40625 0 7.691406 0 C 5.550781 0 3.894531 -0.769531 2.726563 -2.3125 C 1.550781 -3.855469 0.964844 -5.707031 0.96875 -7.867188 C 0.964844 -9.347656 1.234375 -10.71875 1.773438 -11.976563 C 2.308594 -13.230469 3.085938 -14.199219 4.109375 -14.882813 C 5.125 -15.5625 6.324219 -15.902344 7.703125 -15.90625 C 9.539063 -15.902344 11.054688 -15.160156 12.246094 -13.679688 L 12.246094 -15.554688 L 14.679688 -15.554688 L 14.679688 -2.109375 C 14.675781 0.3125 14.429688 2.027344 13.9375 3.039063 C 13.441406 4.046875 12.660156 4.84375 11.59375 5.433594 C 10.523438 6.015625 9.207031 6.308594 7.648438 6.3125 C 5.789063 6.308594 4.289063 5.890625 3.148438 5.058594 C 2.003906 4.222656 1.453125 2.96875 1.492188 1.289063 Z M 3.675781 -8.054688 C 3.671875 -6.011719 4.078125 -4.523438 4.890625 -3.589844 C 5.699219 -2.648438 6.714844 -2.179688 7.9375 -2.183594 C 9.148438 -2.179688 10.164063 -2.644531 10.988281 -3.582031 C 11.804688 -4.511719 12.214844 -5.976563 12.21875 -7.96875 C 12.214844 -9.871094 11.792969 -11.304688 10.949219 -12.273438 C 10.101563 -13.238281 9.082031 -13.722656 7.894531 -13.726563 C 6.71875 -13.722656 5.722656 -13.246094 4.90625 -12.296875 C 4.082031 -11.339844 3.671875 -9.925781 3.675781 -8.054688 Z M 3.675781 -8.054688 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-12\">\n",
       "<path style=\"stroke:none;\" d=\"M 12.625 -5.007813 L 15.351563 -4.671875 C 14.921875 -3.078125 14.125 -1.84375 12.960938 -0.964844 C 11.796875 -0.0859375 10.3125 0.351563 8.511719 0.351563 C 6.234375 0.351563 4.429688 -0.347656 3.097656 -1.75 C 1.761719 -3.148438 1.09375 -5.113281 1.097656 -7.648438 C 1.09375 -10.257813 1.769531 -12.289063 3.117188 -13.738281 C 4.464844 -15.179688 6.210938 -15.902344 8.363281 -15.90625 C 10.441406 -15.902344 12.140625 -15.195313 13.460938 -13.78125 C 14.773438 -12.367188 15.433594 -10.375 15.4375 -7.808594 C 15.433594 -7.648438 15.429688 -7.414063 15.425781 -7.105469 L 3.824219 -7.105469 C 3.917969 -5.390625 4.402344 -4.082031 5.273438 -3.175781 C 6.140625 -2.269531 7.222656 -1.816406 8.523438 -1.816406 C 9.488281 -1.816406 10.3125 -2.070313 11 -2.578125 C 11.679688 -3.085938 12.222656 -3.894531 12.625 -5.007813 Z M 3.96875 -9.273438 L 12.65625 -9.273438 C 12.539063 -10.578125 12.207031 -11.558594 11.660156 -12.21875 C 10.816406 -13.230469 9.726563 -13.738281 8.394531 -13.742188 C 7.179688 -13.738281 6.160156 -13.332031 5.339844 -12.523438 C 4.511719 -11.710938 4.054688 -10.628906 3.96875 -9.273438 Z M 3.96875 -9.273438 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-13\">\n",
       "<path style=\"stroke:none;\" d=\"M 12.363281 -8.421875 L 12.363281 -10.941406 L 21.460938 -10.957031 L 21.460938 -2.988281 C 20.0625 -1.875 18.621094 -1.035156 17.140625 -0.476563 C 15.652344 0.0859375 14.128906 0.363281 12.570313 0.367188 C 10.457031 0.363281 8.542969 -0.0859375 6.820313 -0.988281 C 5.09375 -1.890625 3.789063 -3.195313 2.914063 -4.90625 C 2.03125 -6.609375 1.59375 -8.519531 1.597656 -10.632813 C 1.59375 -12.722656 2.03125 -14.671875 2.90625 -16.484375 C 3.78125 -18.292969 5.039063 -19.640625 6.679688 -20.519531 C 8.320313 -21.398438 10.207031 -21.835938 12.347656 -21.839844 C 13.898438 -21.835938 15.304688 -21.585938 16.558594 -21.085938 C 17.8125 -20.582031 18.796875 -19.882813 19.511719 -18.984375 C 20.222656 -18.085938 20.761719 -16.914063 21.136719 -15.46875 L 18.574219 -14.765625 C 18.25 -15.859375 17.847656 -16.71875 17.371094 -17.34375 C 16.890625 -17.96875 16.207031 -18.46875 15.320313 -18.84375 C 14.429688 -19.21875 13.445313 -19.40625 12.363281 -19.410156 C 11.0625 -19.40625 9.9375 -19.207031 8.992188 -18.816406 C 8.042969 -18.417969 7.277344 -17.898438 6.699219 -17.253906 C 6.113281 -16.609375 5.664063 -15.902344 5.347656 -15.132813 C 4.796875 -13.800781 4.523438 -12.359375 4.527344 -10.8125 C 4.523438 -8.890625 4.851563 -7.289063 5.515625 -6.003906 C 6.171875 -4.710938 7.132813 -3.753906 8.390625 -3.132813 C 9.648438 -2.503906 10.984375 -2.191406 12.40625 -2.195313 C 13.632813 -2.191406 14.835938 -2.429688 16.011719 -2.90625 C 17.179688 -3.378906 18.066406 -3.882813 18.675781 -4.421875 L 18.675781 -8.421875 Z M 12.363281 -8.421875 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-14\">\n",
       "<path style=\"stroke:none;\" d=\"M 4.410156 0 L 1.960938 0 L 1.960938 -21.476563 L 4.601563 -21.476563 L 4.601563 -13.8125 C 5.710938 -15.203125 7.132813 -15.902344 8.863281 -15.90625 C 9.816406 -15.902344 10.722656 -15.710938 11.578125 -15.328125 C 12.429688 -14.941406 13.132813 -14.398438 13.6875 -13.703125 C 14.238281 -13.003906 14.667969 -12.160156 14.984375 -11.175781 C 15.292969 -10.183594 15.449219 -9.128906 15.453125 -8.011719 C 15.449219 -5.34375 14.792969 -3.285156 13.476563 -1.832031 C 12.15625 -0.375 10.574219 0.351563 8.730469 0.351563 C 6.894531 0.351563 5.453125 -0.414063 4.410156 -1.949219 Z M 4.378906 -7.894531 C 4.375 -6.027344 4.628906 -4.679688 5.140625 -3.851563 C 5.96875 -2.492188 7.09375 -1.816406 8.511719 -1.816406 C 9.660156 -1.816406 10.65625 -2.316406 11.496094 -3.316406 C 12.335938 -4.316406 12.753906 -5.808594 12.757813 -7.792969 C 12.753906 -9.824219 12.351563 -11.320313 11.550781 -12.289063 C 10.742188 -13.253906 9.769531 -13.738281 8.628906 -13.742188 C 7.472656 -13.738281 6.476563 -13.238281 5.636719 -12.238281 C 4.796875 -11.234375 4.375 -9.785156 4.378906 -7.894531 Z M 4.378906 -7.894531 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-15\">\n",
       "<path style=\"stroke:none;\" d=\"M 1.917969 0 L 1.917969 -21.476563 L 4.554688 -21.476563 L 4.554688 0 Z M 1.917969 0 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-16\">\n",
       "<path style=\"stroke:none;\" d=\"M 7.734375 -2.359375 L 8.117188 -0.03125 C 7.367188 0.125 6.703125 0.199219 6.125 0.203125 C 5.164063 0.199219 4.421875 0.0507813 3.894531 -0.25 C 3.367188 -0.550781 2.996094 -0.949219 2.78125 -1.441406 C 2.566406 -1.933594 2.460938 -2.96875 2.460938 -4.554688 L 2.460938 -13.507813 L 0.527344 -13.507813 L 0.527344 -15.554688 L 2.460938 -15.554688 L 2.460938 -19.410156 L 5.082031 -20.992188 L 5.082031 -15.554688 L 7.734375 -15.554688 L 7.734375 -13.507813 L 5.082031 -13.507813 L 5.082031 -4.410156 C 5.078125 -3.65625 5.125 -3.171875 5.21875 -2.957031 C 5.308594 -2.742188 5.460938 -2.570313 5.675781 -2.445313 C 5.882813 -2.316406 6.183594 -2.253906 6.578125 -2.257813 C 6.867188 -2.253906 7.253906 -2.289063 7.734375 -2.359375 Z M 7.734375 -2.359375 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-17\">\n",
       "<path style=\"stroke:none;\" d=\"M 1.859375 5.992188 L 1.566406 3.515625 C 2.136719 3.671875 2.640625 3.75 3.078125 3.75 C 3.65625 3.75 4.125 3.652344 4.480469 3.457031 C 4.828125 3.261719 5.117188 2.988281 5.347656 2.636719 C 5.507813 2.371094 5.777344 1.714844 6.152344 0.671875 C 6.199219 0.523438 6.277344 0.308594 6.386719 0.03125 L 0.484375 -15.554688 L 3.324219 -15.554688 L 6.5625 -6.546875 C 6.980469 -5.402344 7.355469 -4.203125 7.691406 -2.945313 C 7.992188 -4.152344 8.351563 -5.332031 8.773438 -6.488281 L 12.101563 -15.554688 L 14.734375 -15.554688 L 8.820313 0.265625 C 8.179688 1.96875 7.6875 3.144531 7.339844 3.792969 C 6.867188 4.660156 6.328125 5.296875 5.726563 5.703125 C 5.117188 6.105469 4.394531 6.308594 3.558594 6.3125 C 3.046875 6.308594 2.480469 6.203125 1.859375 5.992188 Z M 1.859375 5.992188 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-18\">\n",
       "<path style=\"stroke:none;\" d=\"M 2.609375 0 L 2.609375 -13.507813 L 0.277344 -13.507813 L 0.277344 -15.554688 L 2.609375 -15.554688 L 2.609375 -17.210938 C 2.605469 -18.253906 2.699219 -19.03125 2.886719 -19.539063 C 3.136719 -20.222656 3.582031 -20.777344 4.226563 -21.203125 C 4.863281 -21.625 5.761719 -21.835938 6.914063 -21.839844 C 7.65625 -21.835938 8.476563 -21.75 9.375 -21.578125 L 8.980469 -19.277344 C 8.429688 -19.371094 7.910156 -19.417969 7.425781 -19.421875 C 6.621094 -19.417969 6.054688 -19.246094 5.726563 -18.910156 C 5.390625 -18.566406 5.226563 -17.929688 5.230469 -16.992188 L 5.230469 -15.554688 L 8.261719 -15.554688 L 8.261719 -13.507813 L 5.230469 -13.507813 L 5.230469 0 Z M 2.609375 0 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-19\">\n",
       "<path style=\"stroke:none;\" d=\"M 1.992188 -18.441406 L 1.992188 -21.476563 L 4.628906 -21.476563 L 4.628906 -18.441406 Z M 1.992188 0 L 1.992188 -15.554688 L 4.628906 -15.554688 L 4.628906 0 Z M 1.992188 0 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-20\">\n",
       "<path style=\"stroke:none;\" d=\"M 0.921875 -4.644531 L 3.53125 -5.054688 C 3.675781 -4.003906 4.082031 -3.203125 4.753906 -2.648438 C 5.417969 -2.09375 6.355469 -1.816406 7.558594 -1.816406 C 8.769531 -1.816406 9.667969 -2.0625 10.253906 -2.554688 C 10.839844 -3.046875 11.132813 -3.625 11.132813 -4.292969 C 11.132813 -4.886719 10.871094 -5.355469 10.355469 -5.699219 C 9.988281 -5.929688 9.089844 -6.226563 7.660156 -6.59375 C 5.722656 -7.078125 4.382813 -7.5 3.640625 -7.859375 C 2.890625 -8.210938 2.324219 -8.703125 1.941406 -9.335938 C 1.550781 -9.960938 1.359375 -10.660156 1.363281 -11.425781 C 1.359375 -12.117188 1.515625 -12.757813 1.835938 -13.351563 C 2.148438 -13.941406 2.582031 -14.433594 3.132813 -14.824219 C 3.542969 -15.125 4.101563 -15.378906 4.8125 -15.589844 C 5.515625 -15.796875 6.277344 -15.902344 7.089844 -15.90625 C 8.308594 -15.902344 9.378906 -15.726563 10.304688 -15.378906 C 11.222656 -15.023438 11.902344 -14.546875 12.347656 -13.949219 C 12.785156 -13.347656 13.089844 -12.546875 13.257813 -11.542969 L 10.679688 -11.191406 C 10.558594 -11.988281 10.21875 -12.613281 9.660156 -13.066406 C 9.097656 -13.515625 8.304688 -13.738281 7.28125 -13.742188 C 6.066406 -13.738281 5.203125 -13.539063 4.6875 -13.140625 C 4.167969 -12.738281 3.90625 -12.269531 3.910156 -11.734375 C 3.90625 -11.386719 4.015625 -11.078125 4.234375 -10.8125 C 4.445313 -10.523438 4.78125 -10.289063 5.242188 -10.109375 C 5.503906 -10.007813 6.28125 -9.785156 7.574219 -9.433594 C 9.4375 -8.933594 10.738281 -8.523438 11.476563 -8.210938 C 12.210938 -7.890625 12.789063 -7.429688 13.210938 -6.824219 C 13.628906 -6.21875 13.839844 -5.46875 13.84375 -4.570313 C 13.839844 -3.6875 13.582031 -2.859375 13.074219 -2.085938 C 12.558594 -1.308594 11.820313 -0.707031 10.855469 -0.285156 C 9.886719 0.140625 8.792969 0.351563 7.574219 0.351563 C 5.546875 0.351563 4.003906 -0.0664063 2.949219 -0.90625 C 1.886719 -1.746094 1.210938 -2.992188 0.921875 -4.644531 Z M 0.921875 -4.644531 \"/>\n",
       "</symbol>\n",
       "<symbol overflow=\"visible\" id=\"glyph0-21\">\n",
       "<path style=\"stroke:none;\" d=\"M 7.777344 0 L 7.777344 -18.941406 L 0.703125 -18.941406 L 0.703125 -21.476563 L 17.726563 -21.476563 L 17.726563 -18.941406 L 10.621094 -18.941406 L 10.621094 0 Z M 7.777344 0 \"/>\n",
       "</symbol>\n",
       "</g>\n",
       "</defs>\n",
       "<g id=\"surface102\">\n",
       "<rect x=\"0\" y=\"0\" width=\"600\" height=\"600\" style=\"fill:rgb(100%,100%,100%);fill-opacity:1;stroke:none;\"/>\n",
       "<path style=\"fill:none;stroke-width:2.13;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(80.784314%,8.627451%,7.058824%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 347.214844 100 C 265.152344 58.304688 185.675781 65.640625 129.609375 138.113281 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(80.784314%,8.627451%,7.058824%);fill-opacity:1;\" d=\"M 112.152344 160.683594 L 122.277344 132.441406 L 136.945313 143.789063 L 112.152344 160.683594 \"/>\n",
       "<path style=\"fill:none;stroke-width:1.61;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(80.784314%,8.627451%,7.058824%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 347.214844 100 C 336.417969 188.464844 369.324219 258.621094 453.875 285.394531 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(80.784314%,8.627451%,7.058824%);fill-opacity:1;\" d=\"M 481.074219 294.007813 L 451.078125 294.234375 L 456.671875 276.558594 L 481.074219 294.007813 \"/>\n",
       "<path style=\"fill:none;stroke-width:1.91;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(80.784314%,8.627451%,7.058824%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 347.214844 100 C 196.5 130.230469 105.714844 221.667969 101.375 375.070313 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(80.784314%,8.627451%,7.058824%);fill-opacity:1;\" d=\"M 100.566406 403.589844 L 92.105469 374.808594 L 110.640625 375.332031 L 100.566406 403.589844 \"/>\n",
       "<path style=\"fill:none;stroke-width:2.35;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(80.784314%,8.627451%,7.058824%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 347.214844 100 C 247.613281 213.109375 230.984375 338.46875 318.855469 460.589844 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(80.784314%,8.627451%,7.058824%);fill-opacity:1;\" d=\"M 335.519531 483.75 L 311.332031 466.003906 L 326.382813 455.175781 L 335.519531 483.75 \"/>\n",
       "<path style=\"fill:none;stroke-width:2.26;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(80.784314%,8.627451%,7.058824%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 100 176.394531 C 182.0625 218.089844 261.539063 210.753906 317.601563 138.277344 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(80.784314%,8.627451%,7.058824%);fill-opacity:1;\" d=\"M 335.058594 115.710938 L 324.933594 143.949219 L 310.269531 132.605469 L 335.058594 115.710938 \"/>\n",
       "<path style=\"fill:none;stroke-width:3.02;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(80.784314%,8.627451%,7.058824%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 100 176.394531 C 183.214844 311.257813 304.03125 366.007813 453.976563 315.503906 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(80.784314%,8.627451%,7.058824%);fill-opacity:1;\" d=\"M 481.015625 306.394531 L 456.9375 324.289063 L 451.015625 306.714844 L 481.015625 306.394531 \"/>\n",
       "<path style=\"fill:none;stroke-width:3.67;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(80.784314%,8.627451%,7.058824%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 100 176.394531 C 38.847656 238.652344 22.414063 312.882813 72.242188 383.992188 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(80.784314%,8.627451%,7.058824%);fill-opacity:1;\" d=\"M 88.613281 407.355469 L 64.648438 389.308594 L 79.832031 378.671875 L 88.613281 407.355469 \"/>\n",
       "<path style=\"fill:none;stroke-width:5.86;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(80.784314%,8.627451%,7.058824%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 100 176.394531 C 89.535156 329.753906 153.886719 441.386719 300.746094 485.914063 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(80.784314%,8.627451%,7.058824%);fill-opacity:1;\" d=\"M 328.050781 494.191406 L 298.054688 494.785156 L 303.4375 477.039063 L 328.050781 494.191406 \"/>\n",
       "<path style=\"fill:none;stroke-width:1.44;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(80.784314%,8.627451%,7.058824%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 500 300 C 510.792969 211.535156 477.886719 141.378906 393.339844 114.605469 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(80.784314%,8.627451%,7.058824%);fill-opacity:1;\" d=\"M 366.136719 105.992188 L 396.136719 105.765625 L 390.539063 123.441406 L 366.136719 105.992188 \"/>\n",
       "<path style=\"fill:none;stroke-width:2.8;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(80.784314%,8.627451%,7.058824%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 500 300 C 416.785156 165.136719 295.96875 110.382813 146.023438 160.890625 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(80.784314%,8.627451%,7.058824%);fill-opacity:1;\" d=\"M 118.984375 170 L 143.0625 152.105469 L 148.984375 169.675781 L 118.984375 170 \"/>\n",
       "<path style=\"fill:none;stroke-width:2.36;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(80.784314%,8.627451%,7.058824%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 500 300 C 355.207031 235.59375 224.566406 258.546875 129.253906 384.84375 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(80.784314%,8.627451%,7.058824%);fill-opacity:1;\" d=\"M 112.066406 407.617188 L 121.855469 379.257813 L 136.652344 390.425781 L 112.066406 407.617188 \"/>\n",
       "<path style=\"fill:none;stroke-width:3.91;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(80.784314%,8.627451%,7.058824%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 500 300 C 411.8125 312.847656 352.773438 363.042969 349.175781 451.65625 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(80.784314%,8.627451%,7.058824%);fill-opacity:1;\" d=\"M 348.019531 480.164063 L 339.914063 451.28125 L 358.4375 452.035156 L 348.019531 480.164063 \"/>\n",
       "<path style=\"fill:none;stroke-width:2.68;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(80.784314%,8.627451%,7.058824%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 100 423.605469 C 250.714844 393.375 341.5 301.9375 345.839844 148.535156 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(80.784314%,8.627451%,7.058824%);fill-opacity:1;\" d=\"M 346.648438 120.015625 L 355.105469 148.800781 L 336.574219 148.273438 L 346.648438 120.015625 \"/>\n",
       "<path style=\"fill:none;stroke-width:6.04;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(80.784314%,8.627451%,7.058824%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 100 423.605469 C 161.152344 361.347656 177.585938 287.117188 127.757813 216.007813 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(80.784314%,8.627451%,7.058824%);fill-opacity:1;\" d=\"M 111.386719 192.644531 L 135.351563 210.691406 L 120.167969 221.328125 L 111.386719 192.644531 \"/>\n",
       "<path style=\"fill:none;stroke-width:4.57;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(80.784314%,8.627451%,7.058824%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 100 423.605469 C 244.792969 488.011719 375.433594 465.058594 470.746094 338.765625 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(80.784314%,8.627451%,7.058824%);fill-opacity:1;\" d=\"M 487.933594 315.988281 L 478.144531 344.347656 L 463.347656 333.179688 L 487.933594 315.988281 \"/>\n",
       "<path style=\"fill:none;stroke-width:6.39;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(80.784314%,8.627451%,7.058824%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 100 423.605469 C 144.234375 504.332031 213.992188 543.109375 301.167969 514.898438 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(80.784314%,8.627451%,7.058824%);fill-opacity:1;\" d=\"M 328.316406 506.117188 L 304.023438 523.71875 L 298.316406 506.078125 L 328.316406 506.117188 \"/>\n",
       "<path style=\"fill:none;stroke-width:2.38;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(80.784314%,8.627451%,7.058824%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 347.214844 500 C 446.8125 386.890625 463.441406 261.53125 375.570313 139.410156 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(80.784314%,8.627451%,7.058824%);fill-opacity:1;\" d=\"M 358.90625 116.25 L 383.09375 133.996094 L 368.046875 144.824219 L 358.90625 116.25 \"/>\n",
       "<path style=\"fill:none;stroke-width:5.72;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(80.784314%,8.627451%,7.058824%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 347.214844 500 C 357.679688 346.640625 293.328125 235.003906 146.46875 190.480469 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(80.784314%,8.627451%,7.058824%);fill-opacity:1;\" d=\"M 119.164063 182.203125 L 149.15625 181.609375 L 143.777344 199.351563 L 119.164063 182.203125 \"/>\n",
       "<path style=\"fill:none;stroke-width:4.29;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(80.784314%,8.627451%,7.058824%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 347.214844 500 C 435.402344 487.152344 494.441406 436.957031 498.039063 348.34375 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(80.784314%,8.627451%,7.058824%);fill-opacity:1;\" d=\"M 499.195313 319.835938 L 507.300781 348.71875 L 488.773438 347.964844 L 499.195313 319.835938 \"/>\n",
       "<path style=\"fill:none;stroke-width:4.75;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(80.784314%,8.627451%,7.058824%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 347.214844 500 C 302.980469 419.277344 233.222656 380.496094 146.042969 408.707031 \"/>\n",
       "<path style=\" stroke:none;fill-rule:nonzero;fill:rgb(80.784314%,8.627451%,7.058824%);fill-opacity:1;\" d=\"M 118.898438 417.492188 L 143.191406 399.886719 L 148.898438 417.527344 L 118.898438 417.492188 \"/>\n",
       "<path style=\"fill-rule:nonzero;fill:rgb(74.509804%,74.509804%,74.509804%);fill-opacity:1;stroke-width:1;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(0%,0%,0%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 367.214844 100 C 367.214844 111.046875 358.257813 120 347.214844 120 C 336.167969 120 327.214844 111.046875 327.214844 100 C 327.214844 88.953125 336.167969 80 347.214844 80 C 358.257813 80 367.214844 88.953125 367.214844 100 \"/>\n",
       "<path style=\"fill-rule:nonzero;fill:rgb(74.509804%,74.509804%,74.509804%);fill-opacity:1;stroke-width:1;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(0%,0%,0%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 120 176.394531 C 120 187.4375 111.046875 196.394531 100 196.394531 C 88.953125 196.394531 80 187.4375 80 176.394531 C 80 165.347656 88.953125 156.394531 100 156.394531 C 111.046875 156.394531 120 165.347656 120 176.394531 \"/>\n",
       "<path style=\"fill-rule:nonzero;fill:rgb(74.509804%,74.509804%,74.509804%);fill-opacity:1;stroke-width:1;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(0%,0%,0%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 520 300 C 520 311.046875 511.046875 320 500 320 C 488.953125 320 480 311.046875 480 300 C 480 288.953125 488.953125 280 500 280 C 511.046875 280 520 288.953125 520 300 \"/>\n",
       "<path style=\"fill-rule:nonzero;fill:rgb(74.509804%,74.509804%,74.509804%);fill-opacity:1;stroke-width:1;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(0%,0%,0%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 120 423.605469 C 120 434.652344 111.046875 443.605469 100 443.605469 C 88.953125 443.605469 80 434.652344 80 423.605469 C 80 412.5625 88.953125 403.605469 100 403.605469 C 111.046875 403.605469 120 412.5625 120 423.605469 \"/>\n",
       "<path style=\"fill-rule:nonzero;fill:rgb(74.509804%,74.509804%,74.509804%);fill-opacity:1;stroke-width:1;stroke-linecap:butt;stroke-linejoin:miter;stroke:rgb(0%,0%,0%);stroke-opacity:1;stroke-miterlimit:10;\" d=\"M 367.214844 500 C 367.214844 511.046875 358.257813 520 347.214844 520 C 336.167969 520 327.214844 511.046875 327.214844 500 C 327.214844 488.953125 336.167969 480 347.214844 480 C 358.257813 480 367.214844 488.953125 367.214844 500 \"/>\n",
       "<g style=\"fill:rgb(0%,0%,0%);fill-opacity:1;\">\n",
       "  <use xlink:href=\"#glyph0-1\" x=\"318.804688\" y=\"136.839844\"/>\n",
       "  <use xlink:href=\"#glyph0-2\" x=\"340.469727\" y=\"136.839844\"/>\n",
       "  <use xlink:href=\"#glyph0-3\" x=\"362.134766\" y=\"136.839844\"/>\n",
       "</g>\n",
       "<g style=\"fill:rgb(0%,0%,0%);fill-opacity:1;\">\n",
       "  <use xlink:href=\"#glyph0-4\" x=\"16.125\" y=\"212.867188\"/>\n",
       "  <use xlink:href=\"#glyph0-5\" x=\"41.115234\" y=\"212.867188\"/>\n",
       "  <use xlink:href=\"#glyph0-6\" x=\"57.799805\" y=\"212.867188\"/>\n",
       "  <use xlink:href=\"#glyph0-7\" x=\"72.799805\" y=\"212.867188\"/>\n",
       "  <use xlink:href=\"#glyph0-8\" x=\"82.790039\" y=\"212.867188\"/>\n",
       "  <use xlink:href=\"#glyph0-9\" x=\"99.474609\" y=\"212.867188\"/>\n",
       "  <use xlink:href=\"#glyph0-10\" x=\"116.15918\" y=\"212.867188\"/>\n",
       "  <use xlink:href=\"#glyph0-5\" x=\"132.84375\" y=\"212.867188\"/>\n",
       "  <use xlink:href=\"#glyph0-11\" x=\"149.52832\" y=\"212.867188\"/>\n",
       "  <use xlink:href=\"#glyph0-12\" x=\"166.212891\" y=\"212.867188\"/>\n",
       "</g>\n",
       "<g style=\"fill:rgb(0%,0%,0%);fill-opacity:1;\">\n",
       "  <use xlink:href=\"#glyph0-13\" x=\"455.117188\" y=\"336.839844\"/>\n",
       "  <use xlink:href=\"#glyph0-8\" x=\"478.452148\" y=\"336.839844\"/>\n",
       "  <use xlink:href=\"#glyph0-14\" x=\"495.136719\" y=\"336.839844\"/>\n",
       "  <use xlink:href=\"#glyph0-15\" x=\"511.821289\" y=\"336.839844\"/>\n",
       "  <use xlink:href=\"#glyph0-12\" x=\"518.486328\" y=\"336.839844\"/>\n",
       "  <use xlink:href=\"#glyph0-16\" x=\"535.170898\" y=\"336.839844\"/>\n",
       "</g>\n",
       "<g style=\"fill:rgb(0%,0%,0%);fill-opacity:1;\">\n",
       "  <use xlink:href=\"#glyph0-4\" x=\"9.796875\" y=\"460.449219\"/>\n",
       "  <use xlink:href=\"#glyph0-17\" x=\"34.787109\" y=\"460.449219\"/>\n",
       "  <use xlink:href=\"#glyph0-8\" x=\"49.787109\" y=\"460.449219\"/>\n",
       "  <use xlink:href=\"#glyph0-18\" x=\"66.47168\" y=\"460.449219\"/>\n",
       "  <use xlink:href=\"#glyph0-19\" x=\"74.806641\" y=\"460.449219\"/>\n",
       "  <use xlink:href=\"#glyph0-14\" x=\"81.47168\" y=\"460.449219\"/>\n",
       "  <use xlink:href=\"#glyph0-7\" x=\"98.15625\" y=\"460.449219\"/>\n",
       "  <use xlink:href=\"#glyph0-8\" x=\"108.146484\" y=\"460.449219\"/>\n",
       "  <use xlink:href=\"#glyph0-14\" x=\"124.831055\" y=\"460.449219\"/>\n",
       "  <use xlink:href=\"#glyph0-15\" x=\"141.515625\" y=\"460.449219\"/>\n",
       "  <use xlink:href=\"#glyph0-5\" x=\"148.180664\" y=\"460.449219\"/>\n",
       "  <use xlink:href=\"#glyph0-20\" x=\"164.865234\" y=\"460.449219\"/>\n",
       "  <use xlink:href=\"#glyph0-16\" x=\"179.865234\" y=\"460.449219\"/>\n",
       "</g>\n",
       "<g style=\"fill:rgb(0%,0%,0%);fill-opacity:1;\">\n",
       "  <use xlink:href=\"#glyph0-21\" x=\"317.023438\" y=\"536.476563\"/>\n",
       "  <use xlink:href=\"#glyph0-7\" x=\"335.348633\" y=\"536.476563\"/>\n",
       "  <use xlink:href=\"#glyph0-12\" x=\"345.338867\" y=\"536.476563\"/>\n",
       "  <use xlink:href=\"#glyph0-11\" x=\"362.023438\" y=\"536.476563\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<igraph.drawing.Plot at 0x19e8e92aec8>"
      ]
     },
     "execution_count": 40,
     "metadata": {
      "image/svg+xml": {
       "isolated": true
      }
     },
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ig.plot(g_uc, \"UC_uninflamed.png\", **visual_style)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below you can see what kind of data we have used."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['builtins',\n",
       " 'builtins',\n",
       " 'pandas',\n",
       " 'igraph',\n",
       " 'os',\n",
       " 'numpy',\n",
       " 'omnipath',\n",
       " 'seaborn',\n",
       " 'types']"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import types\n",
    "def imports():\n",
    "    for name, val in globals().items():\n",
    "        if isinstance(val, types.ModuleType):\n",
    "            yield val.__name__\n",
    "list(imports())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pandas 1.1.4\n",
      "igraph 0.8.3\n",
      "numpy 1.18.5\n",
      "omnipath 1.0.0\n",
      "seaborn 0.11.0\n"
     ]
    }
   ],
   "source": [
    "print('\\n'.join(f'{m.__name__} {m.__version__}' for m in globals().values() if getattr(m, '__version__', None)))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
