{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Network propagation development\n",
    "\n",
    "The notebook currently covers how results from a AnnData/MuData object can be added to a cpr_graph to setup network-based inference. The general strategy is to:\n",
    "1. pull out a pd.DataFrame containing feature-level measures of interest along with feature metadata.\n",
    "2. These are then mapped on the species ids in an sbml_dfs model by shared on ontology, disambiguated (to handle mapping of multiple features to the same s_id), and s_id-indexed results are embedded in the sbml_dfs as a table in species_data\n",
    "3. attributes of interrest are then passed from the sbml_dfs model into the graph.\n",
    "\n",
    "This example uses real MuData results but only a small sbml_dfs object which has uniprot but not ENSG identifiers. This makes things easy to work with but a genome-scale graph will need to be used for a real analysis.\n",
    "\n",
    "Reflecting on the current functionality,\n",
    "\n",
    "(1) is not too hard but the interface can probably be cleaned up as we should have a function which applies 1-3 in a single call.\n",
    "(2) is in pretty good shape following a LOT of new functionality being added to napistu-py for handling many-to-one mappings and wide/nested formats for identifiers.\n",
    "(3) will need some better functionality since the reaction_attrs syntax is pretty cryptic but the core functionality is all there.\n",
    "\n",
    "Next, steps will be develop basic PPR functionality."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "import mudata as md\n",
    "import pandas as pd\n",
    "\n",
    "from napistu.ingestion import sbml\n",
    "from napistu import sbml_dfs_core\n",
    "from napistu import mechanism_matching\n",
    "from napistu.network import net_propagation\n",
    "from napistu import utils as napistu_utils\n",
    "from napistu.network import net_utils\n",
    "from napistu.gcs import downloads\n",
    "\n",
    "# local library\n",
    "import regulation\n",
    "\n",
    "# setup logging\n",
    "import logging\n",
    "logging.basicConfig(level=logging.INFO)\n",
    "logger = logging.getLogger(__name__)\n",
    "\n",
    "# paths\n",
    "PROJECT_DIR =  os.path.expanduser(\"~/Desktop/DATA/Forny2023\")\n",
    "SUPPLEMENTAL_DATA_DIR = os.path.join(PROJECT_DIR, \"input\")\n",
    "CACHE_DIR = os.path.join(PROJECT_DIR, \"cache\")\n",
    "NAPISTU_DATA_DIR = os.path.expanduser(\"~/Desktop/DATA/napistu_data\")\n",
    "\n",
    "# Define the path to save hyperparameter scan results\n",
    "MOFA_PARAM_SCAN_MODELS_PATH = os.path.join(CACHE_DIR, \"mofa_param_scan_h5mu\")\n",
    "# Final results \n",
    "OPTIMAL_MODEL_H5MU_PATH = os.path.join(CACHE_DIR, \"mofa_optimal_model.h5mu\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "sbml_dfs_path = downloads.load_public_napistu_asset(\n",
    "    asset = \"human_consensus\",\n",
    "    data_dir = NAPISTU_DATA_DIR,\n",
    "    subasset = \"sbml_dfs\"\n",
    ")\n",
    "\n",
    "cpr_graph_path  = downloads.load_public_napistu_asset(\n",
    "    asset = \"human_consensus\",\n",
    "    data_dir = NAPISTU_DATA_DIR,\n",
    "    subasset = \"regulatory_graph\"\n",
    ")\n",
    "\n",
    "identifiers_path = downloads.load_public_napistu_asset(\n",
    "    asset = \"human_consensus\",\n",
    "    data_dir = NAPISTU_DATA_DIR,\n",
    "    subasset = \"identifiers\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ~2 min load\n",
    "sbml_dfs = napistu_utils.load_pickle(sbml_dfs_path)\n",
    "cpr_graph = napistu_utils.load_pickle(cpr_graph_path)\n",
    "identifiers = pd.read_csv(identifiers_path, delimiter = \"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "#net_utils.validate_assets(\n",
    "#    sbml_dfs = sbml_dfs,\n",
    "#    cpr_graph = cpr_graph,\n",
    "#    # TODO - it should really possible for this to be optional\n",
    "#    precomputed_distances = None,\n",
    "#    identifiers = identifiers\n",
    "#    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/sean/Desktop/GITHUB/napistu/lib/napistu-scrapyard/applications/forny_2023/.venv/lib/python3.11/site-packages/mudata/_core/mudata.py:1531: FutureWarning: From 0.4 .update() will not pull obs/var columns from individual modalities by default anymore. Set mudata.set_options(pull_on_update=False) to adopt the new behaviour, which will become the default. Use new pull_obs/pull_var and push_obs/push_var methods for more flexibility.\n",
      "  self._update_attr(\"var\", axis=0, join_common=join_common)\n",
      "/Users/sean/Desktop/GITHUB/napistu/lib/napistu-scrapyard/applications/forny_2023/.venv/lib/python3.11/site-packages/mudata/_core/mudata.py:1429: FutureWarning: From 0.4 .update() will not pull obs/var columns from individual modalities by default anymore. Set mudata.set_options(pull_on_update=False) to adopt the new behaviour, which will become the default. Use new pull_obs/pull_var and push_obs/push_var methods for more flexibility.\n",
      "  self._update_attr(\"obs\", axis=1, join_common=join_common)\n"
     ]
    }
   ],
   "source": [
    "# lets load the Forny results so we trying adding a few different types of tables to the sbml_dfs\n",
    "mdata = md.read_h5mu(OPTIMAL_MODEL_H5MU_PATH)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Adding genome-scale datasets\n",
    "\n",
    "To use an 'omic dataset in Napistu, we want to:\n",
    "1. mount the dataset on the pathway `sbml_dfs`. This entails:\n",
    "    - matching systematic identifiers between the dataset and pathway to connect 'omic features to Napistu `species`.\n",
    "    - resolve many-to-1 mappings (e.g., where 2+ features match the same species).\n",
    "    - create a table with unique species ids as the index with variable from the dataset.\n",
    "    - add this to the `species_data` attriute of the `sbml_dfs`. Multiple tables and/or datasets can be added to `species_data`.\n",
    "2. pass variables from one or more `species_data` tables to a `cpr_graph`'s vertices with `net_create._add_graph_species_attribute`. Variables can be transformed (e.g., to make them non-negative for personalized pagerank) at this point (or this could be done before step (1)).\n",
    "3. use these verterx attributes for downstream analysis (e.g., using it in the reset_proportional_to parameters of PPR).\n",
    "\n",
    "Step (1) needs to be adapted depending on how datasets are organized. The currently, supported inputs are:\n",
    "- `pd.DataFrame` objects which including 1+ systematic identifiers\n",
    "- `anndata.AnnData` objects where the `var` table provided identifiers, and feature-level summaries come from either the `var`, `varm` or `X` tables.\n",
    "- `mudata.MuData` objects containing multiple `AnnData` objects where `var` and `varm` attributes can be defined across multiple datasets.\n",
    "\n",
    "We'll provide an examples using each of these inputs\n",
    "\n",
    "### Loading results from a pd.DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "SUPPLEMENTAL_DATA_DIR = os.path.join(PROJECT_DIR, \"input\")\n",
    "VZ_LMM_RESULTS = {\n",
    "    \"transcriptomics\": \"diff_exp_lmm_rnaseq_pathwayact_all_annotout.txt\",\n",
    "    \"proteomics\": \"diff_exp_lmm_prot_pathwayact_all_annotout.txt\"\n",
    "}\n",
    "\n",
    "sideloaded_data_path = {x : os.path.join(SUPPLEMENTAL_DATA_DIR, y) for x, y in VZ_LMM_RESULTS.items()}\n",
    "\n",
    "assert all([os.path.isfile(x) for x in sideloaded_data_path.values()])\n",
    "\n",
    "sideloaded_data = {\n",
    "    x : pd.read_csv(y, delimiter= \"\\t\") for x, y in sideloaded_data_path.items()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:napistu.sbml_dfs_utils:Running in non-dogmatic mode - genes, transcripts, and proteins will be merged if possible.\n",
      "DEBUG:napistu.mechanism_matching:Validated ontology columns: {'ensembl_gene'}\n",
      "WARNING:napistu.mechanism_matching:No feature_id column found in DataFrame, creating one\n",
      "INFO:napistu.mechanism_matching:Using columns as results: ['fdr', 'feature_id', 'chi_sq', 'pval']\n",
      "DEBUG:napistu.mechanism_matching:Final long format shape: (14749, 6)\n",
      "DEBUG:napistu.mechanism_matching:Matching 14749 features to 42421 species for ontology ensembl_gene\n",
      "INFO:napistu.mechanism_matching:Found 28526 total matches across 1 ontologies\n",
      "INFO:napistu.mechanism_matching:98.6% of feature_ids are present one or more times in the output (14544/14749)\n",
      "INFO:napistu.mechanism_matching:2617 s_id(s) map to more than one feature_id.\n",
      "INFO:napistu.mechanism_matching:Examples of s_id mapping to multiple feature_ids (showing up to 3):\n",
      "s_id       s_name                        \n",
      "S00000054  UBE2L3                                     [3752, 9356]\n",
      "S00000105  Ferritin Complex                           [2387, 7544]\n",
      "S00000106  Golgi-associated Vesicle Cargo    [12302, 13418, 14573]\n",
      "Name: feature_id, dtype: object\n",
      "INFO:napistu.mechanism_matching:4848 feature_id(s) map to more than one s_id.\n",
      "INFO:napistu.mechanism_matching:Examples of feature_id mapping to multiple s_ids (showing up to 3):\n",
      "feature_id\n",
      "1    [(S00014190, NOC2L), (S00014141, NOC2L:p-S15,S...\n",
      "7    [(S00012093, agrin), (S00039569, XylS-AGRN(30-...\n",
      "9    [(S00003584, TNFRSF18 gene), (S00003613, TNFSF...\n",
      "dtype: object\n",
      "INFO:napistu.sbml_dfs_utils:Running in non-dogmatic mode - genes, transcripts, and proteins will be merged if possible.\n",
      "DEBUG:napistu.mechanism_matching:Validated ontology columns: {'ensembl_gene'}\n",
      "WARNING:napistu.mechanism_matching:No feature_id column found in DataFrame, creating one\n",
      "INFO:napistu.mechanism_matching:Using columns as results: ['fdr', 'feature_id', 'chi_sq', 'pval']\n",
      "DEBUG:napistu.mechanism_matching:Final long format shape: (4395, 6)\n",
      "DEBUG:napistu.mechanism_matching:Matching 4395 features to 42421 species for ontology ensembl_gene\n",
      "INFO:napistu.mechanism_matching:Found 10209 total matches across 1 ontologies\n",
      "INFO:napistu.mechanism_matching:100.0% of feature_ids are present one or more times in the output (4395/4395)\n",
      "INFO:napistu.mechanism_matching:708 s_id(s) map to more than one feature_id.\n",
      "INFO:napistu.mechanism_matching:Examples of s_id mapping to multiple feature_ids (showing up to 3):\n",
      "s_id       s_name                        \n",
      "S00000105  Ferritin Complex                    [593, 594]\n",
      "S00000106  Golgi-associated Vesicle Cargo    [1979, 2381]\n",
      "S00000111  ACBD3:GOLGB1 dimer                [2209, 3664]\n",
      "Name: feature_id, dtype: object\n",
      "INFO:napistu.mechanism_matching:1901 feature_id(s) map to more than one s_id.\n",
      "INFO:napistu.mechanism_matching:Examples of feature_id mapping to multiple s_ids (showing up to 3):\n",
      "feature_id\n",
      "2    [(S00030519, extended synaptotagmin 2), (S0001...\n",
      "5    [(S00035227, NBAS subunit of NRZ tethering com...\n",
      "9    [(S00023465, phosphoglycolate phosphatase), (S...\n",
      "dtype: object\n"
     ]
    }
   ],
   "source": [
    "for k in sideloaded_data.keys():\n",
    "    x = sideloaded_data[k][[\"ensembl\", \"chi_sq\", \"pval\", \"fdr\"]]\n",
    "\n",
    "    mechanism_matching.bind_wide_results(\n",
    "        sbml_dfs,\n",
    "        x,\n",
    "        f\"{k}_loose_data\",\n",
    "        # map columns to controlled Napistu's vocabulary\n",
    "        ontologies = {\"ensembl\" : \"ensembl_gene\"},\n",
    "        species_identifiers = identifiers,\n",
    "        dogmatic = False,\n",
    "        verbose = True\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading Results from an AnnData object\n",
    "\n",
    "Since the Forny dataset is a multiomics experiment many of the variablges we are interested in will hold a common interpretation across all modalities. For example, the effect size of a term in a regression holds a common meaning as do the loadings from a multi-omic factor analysis (MOFA) decomposition.\n",
    "\n",
    "But, many datasets will just be a single modality, and even for multiomic datasets we may be interested in exploring the biology of datamodality-specific attributes. An example in this study is the data-modality specific principal component loadings. Since PCA was performed separately on each data modality the principal components will likely be relatively uncorrelated hence it doesn't make much sense to treat the loadings of PCX to one another across modalities. This is definitely the case for this dataset - PC1 of the proteomics data largely reflects a chromatography-driven technical batch effect which is not seen in the transcriptomics data. To more directly explore this proteomics batch effect we can pull PC1 out of its `AnnData` table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(mdata[\"proteomics\"].obsm.keys())\n",
    "\n",
    "#mdata[\"proteomics\"].var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "cannot concatenate object of type '<class 'numpy.ndarray'>'; only Series and DataFrame objs are valid",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mTypeError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[108]\u001b[39m\u001b[32m, line 165\u001b[39m\n\u001b[32m    161\u001b[39m         raw_results_table = pd.concat([var_table, raw_results_table], axis=\u001b[32m1\u001b[39m)\n\u001b[32m    163\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m raw_results_table\n\u001b[32m--> \u001b[39m\u001b[32m165\u001b[39m \u001b[43mprepare_anndata_results_df\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    166\u001b[39m \u001b[43m    \u001b[49m\u001b[43mmdata\u001b[49m\u001b[43m[\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mproteomics\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    167\u001b[39m \u001b[43m    \u001b[49m\u001b[43mtable_type\u001b[49m\u001b[43m \u001b[49m\u001b[43m=\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mX\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\n\u001b[32m    168\u001b[39m \u001b[43m)\u001b[49m\n",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[108]\u001b[39m\u001b[32m, line 161\u001b[39m, in \u001b[36mprepare_anndata_results_df\u001b[39m\u001b[34m(adata, table_type, table_name, results_attrs, ontologies, index_which_ontology, verbose)\u001b[39m\n\u001b[32m    159\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m (table_type \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m ADATA.VAR) \u001b[38;5;129;01mor\u001b[39;00m (ontologies == \u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[32m    160\u001b[39m     var_table = raw_results_table[table_name]\n\u001b[32m--> \u001b[39m\u001b[32m161\u001b[39m     raw_results_table = \u001b[43mpd\u001b[49m\u001b[43m.\u001b[49m\u001b[43mconcat\u001b[49m\u001b[43m(\u001b[49m\u001b[43m[\u001b[49m\u001b[43mvar_table\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mraw_results_table\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[43m=\u001b[49m\u001b[32;43m1\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[32m    163\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m raw_results_table\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Desktop/GITHUB/napistu/lib/napistu-scrapyard/applications/forny_2023/.venv/lib/python3.11/site-packages/pandas/util/_decorators.py:331\u001b[39m, in \u001b[36mdeprecate_nonkeyword_arguments.<locals>.decorate.<locals>.wrapper\u001b[39m\u001b[34m(*args, **kwargs)\u001b[39m\n\u001b[32m    325\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(args) > num_allow_args:\n\u001b[32m    326\u001b[39m     warnings.warn(\n\u001b[32m    327\u001b[39m         msg.format(arguments=_format_argument_list(allow_args)),\n\u001b[32m    328\u001b[39m         \u001b[38;5;167;01mFutureWarning\u001b[39;00m,\n\u001b[32m    329\u001b[39m         stacklevel=find_stack_level(),\n\u001b[32m    330\u001b[39m     )\n\u001b[32m--> \u001b[39m\u001b[32m331\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[43m*\u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Desktop/GITHUB/napistu/lib/napistu-scrapyard/applications/forny_2023/.venv/lib/python3.11/site-packages/pandas/core/reshape/concat.py:368\u001b[39m, in \u001b[36mconcat\u001b[39m\u001b[34m(objs, axis, join, ignore_index, keys, levels, names, verify_integrity, sort, copy)\u001b[39m\n\u001b[32m    146\u001b[39m \u001b[38;5;129m@deprecate_nonkeyword_arguments\u001b[39m(version=\u001b[38;5;28;01mNone\u001b[39;00m, allowed_args=[\u001b[33m\"\u001b[39m\u001b[33mobjs\u001b[39m\u001b[33m\"\u001b[39m])\n\u001b[32m    147\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34mconcat\u001b[39m(\n\u001b[32m    148\u001b[39m     objs: Iterable[NDFrame] | Mapping[HashableT, NDFrame],\n\u001b[32m   (...)\u001b[39m\u001b[32m    157\u001b[39m     copy: \u001b[38;5;28mbool\u001b[39m = \u001b[38;5;28;01mTrue\u001b[39;00m,\n\u001b[32m    158\u001b[39m ) -> DataFrame | Series:\n\u001b[32m    159\u001b[39m \u001b[38;5;250m    \u001b[39m\u001b[33;03m\"\"\"\u001b[39;00m\n\u001b[32m    160\u001b[39m \u001b[33;03m    Concatenate pandas objects along a particular axis.\u001b[39;00m\n\u001b[32m    161\u001b[39m \n\u001b[32m   (...)\u001b[39m\u001b[32m    366\u001b[39m \u001b[33;03m    1   3   4\u001b[39;00m\n\u001b[32m    367\u001b[39m \u001b[33;03m    \"\"\"\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m368\u001b[39m     op = \u001b[43m_Concatenator\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    369\u001b[39m \u001b[43m        \u001b[49m\u001b[43mobjs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    370\u001b[39m \u001b[43m        \u001b[49m\u001b[43maxis\u001b[49m\u001b[43m=\u001b[49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    371\u001b[39m \u001b[43m        \u001b[49m\u001b[43mignore_index\u001b[49m\u001b[43m=\u001b[49m\u001b[43mignore_index\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    372\u001b[39m \u001b[43m        \u001b[49m\u001b[43mjoin\u001b[49m\u001b[43m=\u001b[49m\u001b[43mjoin\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    373\u001b[39m \u001b[43m        \u001b[49m\u001b[43mkeys\u001b[49m\u001b[43m=\u001b[49m\u001b[43mkeys\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    374\u001b[39m \u001b[43m        \u001b[49m\u001b[43mlevels\u001b[49m\u001b[43m=\u001b[49m\u001b[43mlevels\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    375\u001b[39m \u001b[43m        \u001b[49m\u001b[43mnames\u001b[49m\u001b[43m=\u001b[49m\u001b[43mnames\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    376\u001b[39m \u001b[43m        \u001b[49m\u001b[43mverify_integrity\u001b[49m\u001b[43m=\u001b[49m\u001b[43mverify_integrity\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    377\u001b[39m \u001b[43m        \u001b[49m\u001b[43mcopy\u001b[49m\u001b[43m=\u001b[49m\u001b[43mcopy\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    378\u001b[39m \u001b[43m        \u001b[49m\u001b[43msort\u001b[49m\u001b[43m=\u001b[49m\u001b[43msort\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    379\u001b[39m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    381\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m op.get_result()\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Desktop/GITHUB/napistu/lib/napistu-scrapyard/applications/forny_2023/.venv/lib/python3.11/site-packages/pandas/core/reshape/concat.py:458\u001b[39m, in \u001b[36m_Concatenator.__init__\u001b[39m\u001b[34m(self, objs, axis, join, keys, levels, names, ignore_index, verify_integrity, copy, sort)\u001b[39m\n\u001b[32m    453\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(obj, (ABCSeries, ABCDataFrame)):\n\u001b[32m    454\u001b[39m         msg = (\n\u001b[32m    455\u001b[39m             \u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mcannot concatenate object of type \u001b[39m\u001b[33m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mtype\u001b[39m(obj)\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m'\u001b[39m\u001b[33m; \u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m    456\u001b[39m             \u001b[33m\"\u001b[39m\u001b[33monly Series and DataFrame objs are valid\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m    457\u001b[39m         )\n\u001b[32m--> \u001b[39m\u001b[32m458\u001b[39m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(msg)\n\u001b[32m    460\u001b[39m     ndims.add(obj.ndim)\n\u001b[32m    462\u001b[39m \u001b[38;5;66;03m# get the sample\u001b[39;00m\n\u001b[32m    463\u001b[39m \u001b[38;5;66;03m# want the highest ndim that we have, and must be non-empty\u001b[39;00m\n\u001b[32m    464\u001b[39m \u001b[38;5;66;03m# unless all objs are empty\u001b[39;00m\n",
      "\u001b[31mTypeError\u001b[39m: cannot concatenate object of type '<class 'numpy.ndarray'>'; only Series and DataFrame objs are valid"
     ]
    }
   ],
   "source": [
    "from types import SimpleNamespace\n",
    "import copy\n",
    "\n",
    "ADATA = SimpleNamespace(\n",
    "    LAYERS=\"layers\",\n",
    "    OBS=\"obs\",\n",
    "    OBSM=\"obsm\",\n",
    "    OBSP=\"obsp\",\n",
    "    VAR=\"var\",\n",
    "    VARM=\"varm\",\n",
    "    VARP=\"varp\",\n",
    "    X=\"X\",   \n",
    ")\n",
    "\n",
    "ADATA_DICTLIKE_ATTRS = [ADATA.LAYERS, ADATA.OBSM, ADATA.OBSP, ADATA.VARM, ADATA.VARP]\n",
    "ADATA_IDENTITY_ATTRS = [ADATA.OBS, ADATA.VAR, ADATA.X]\n",
    "\n",
    "\n",
    "import anndata\n",
    "from typing import Literal, Optional, List, Union, Set, Dict\n",
    "\n",
    "def _load_raw_table(\n",
    "    adata: anndata.AnnData,\n",
    "    table_type: str,\n",
    "    table_name: Optional[str] = None\n",
    "):\n",
    "    \n",
    "    \"\"\"\n",
    "    Load an AnnData table.\n",
    "    \n",
    "    This function loads an AnnData table and returns it as a pd.DataFrame.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    adata : anndata.AnnData\n",
    "        The AnnData object to load the table from.\n",
    "    table_type : str\n",
    "        The type of table to load.\n",
    "    table_name : str, optional\n",
    "        The name of the table to load.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    pd.DataFrame\n",
    "        The loaded table.\n",
    "    \"\"\"\n",
    "    \n",
    "    if table_type not in [*ADATA_DICTLIKE_ATTRS, *ADATA_IDENTITY_ATTRS]:\n",
    "        raise ValueError(f\"table_type {table_type} is not a valid AnnData attribute. Valid attributes are: {ADATA_DICTLIKE_ATTRS + ADATA_IDENTITY_ATTRS}\")\n",
    "\n",
    "    if table_type in ADATA_IDENTITY_ATTRS:\n",
    "        if table_name is not None:\n",
    "            logger.debug(f\"table_name {table_name} is not None, but table_type is in IDENTITY_TABLES. \"\n",
    "                        f\"table_name will be ignored.\")\n",
    "        return getattr(adata, table_type)\n",
    "\n",
    "    # pull out a dict-like attribute\n",
    "    return _get_table_from_dict_attr(\n",
    "        adata,\n",
    "        table_type,\n",
    "        table_name\n",
    "    )\n",
    "    \n",
    "    \n",
    "def _get_table_from_dict_attr(\n",
    "    adata: anndata.AnnData,\n",
    "    attr_name: str,\n",
    "    table_name: Optional[str] = None\n",
    "):\n",
    "    \"\"\"\n",
    "    Generic function to get a table from a dict-like AnnData attribute (varm, layers, etc.)\n",
    "    \n",
    "    Args:\n",
    "        adata: AnnData object\n",
    "        attr_name: Name of the attribute ('varm', 'layers', etc.)\n",
    "        table_name: Specific table name to retrieve, or None for auto-selection\n",
    "    \"\"\"\n",
    "\n",
    "    if attr_name not in ADATA_DICTLIKE_ATTRS:\n",
    "        raise ValueError(f\"attr_name {attr_name} is not a dict-like AnnData attribute. Valid attributes are: {VALID_ATTRS}\")\n",
    "\n",
    "    attr_dict = getattr(adata, attr_name)\n",
    "    available_tables = list(attr_dict.keys())\n",
    "    \n",
    "    if len(available_tables) == 0:\n",
    "        raise ValueError(f\"No tables found in adata.{attr_name}\")\n",
    "    elif (len(available_tables) > 1) and (table_name is None):\n",
    "        raise ValueError(f\"Multiple tables found in adata.{attr_name} and table_name is not specified. \"\n",
    "                        f\"Available: {available_tables}\")\n",
    "    elif (len(available_tables) == 1) and (table_name is None):\n",
    "        return attr_dict[available_tables[0]]\n",
    "    elif table_name not in available_tables:\n",
    "        raise ValueError(f\"table_name '{table_name}' not found in adata.{attr_name}. \"\n",
    "                        f\"Available: {available_tables}\")\n",
    "    else:\n",
    "        return attr_dict[table_name]\n",
    "    \n",
    "#_load_raw_table(adata, \"X\")\n",
    "#_load_raw_table(adata, \"var\", \"ignored\")\n",
    "# _load_raw_table(adata, \"layers\")\n",
    "# _load_raw_table(adata, \"foo\")\n",
    "\n",
    "def _select_results_attrs(\n",
    "    adata: anndata.AnnData,\n",
    "    raw_results_table: Union[pd.DataFrame, np.ndarray],\n",
    "    results_attrs: Optional[List[str]] = None\n",
    ") -> pd.DataFrame:\n",
    "\n",
    "    \"\"\"\n",
    "    Select results attributes from an AnnData object.\n",
    "\n",
    "    This function selects results attributes from raw_results_table derived\n",
    "    from an AnnData object and converts them if needed to a pd.DataFrame\n",
    "    with appropriate indicies.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    adata : anndata.AnnData\n",
    "        The AnnData object containing the results to be formatted.\n",
    "    raw_results_table : pd.DataFrame or np.ndarray\n",
    "        The raw results table to be formatted.\n",
    "    results_attrs : list of str, optional\n",
    "        The attributes to extract from the raw_results_table.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    pd.DataFrame\n",
    "        A DataFrame containing the formatted results.\n",
    "    \"\"\"\n",
    "    if isinstance(raw_results_table, pd.DataFrame):\n",
    "        if results_attrs is not None:\n",
    "            results_table_data = raw_results_table.loc[results_attrs]\n",
    "        else:\n",
    "            results_table_data = raw_results_table\n",
    "    else:\n",
    "        if results_attrs is not None:\n",
    "            # Check that results_attrs exist in adata.obs.index\n",
    "            valid_obs = adata.obs.index.tolist()\n",
    "            \n",
    "            invalid_results_attrs = [x for x in results_attrs if x not in valid_obs]\n",
    "            if len(invalid_results_attrs) > 0:\n",
    "                raise ValueError(f\"The following results attributes are not present in the AnnData object's obs index: {invalid_results_attrs}\")\n",
    "\n",
    "            # Find positions of desired rows in adata.obs.index\n",
    "            row_positions = [adata.obs.index.get_loc(attr) for attr in results_attrs]\n",
    "            \n",
    "            # Select ROWS from numpy array using positions\n",
    "            selected_array = raw_results_table[row_positions, :]\n",
    "            \n",
    "            # Convert to DataFrame and set row names to results_attrs\n",
    "            results_table_data = pd.DataFrame(\n",
    "                selected_array,\n",
    "                index = results_attrs,\n",
    "                columns = adata.var.index\n",
    "                ).T\n",
    "        else:\n",
    "            # Convert entire array to DataFrame\n",
    "            results_table_data = pd.DataFrame(\n",
    "                raw_results_table,\n",
    "                index = adata.obs.index,\n",
    "                columns = adata.var.index\n",
    "            ).T\n",
    "\n",
    "    return results_table_data\n",
    "\n",
    "# _select_results_attrs(adata, adata.X, [\"MMA001\", \"MMA004\", \"MMA005\"])\n",
    "\n",
    "def prepare_anndata_results_df(\n",
    "    adata: anndata.AnnData,\n",
    "    table_type: Literal[ADATA.VAR, ADATA.VARM, ADATA.X] = ADATA.VAR,\n",
    "    table_name: Optional[str] = None,\n",
    "    results_attrs: Optional[List[str]] = None,\n",
    "    ontologies: Optional[Union[Set[str], Dict[str, str]]] = None,\n",
    "    index_which_ontology: Optional[str] = None,\n",
    "    verbose: bool = True\n",
    "):\n",
    "\n",
    "    \"\"\"\n",
    "    Prepare a results table from an AnnData object for use in Napistu.\n",
    "\n",
    "    This function extracts a table from an AnnData object and formats it for use in Napistu.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    adata : anndata.AnnData\n",
    "        The AnnData object containing the results to be formatted.\n",
    "    table_type : Literal[\"var\", \"varm\", \"X\"], optional\n",
    "        The type of table to extract from the AnnData object.\n",
    "    table_name : str, optional\n",
    "        The name of the table to extract from the AnnData object.\n",
    "    results_attrs : list of str, optional\n",
    "        The attributes to extract from the table.\n",
    "    index_which_ontology : str, optional\n",
    "        The ontology to use for the systematic identifiers. This column will be pulled out of the\n",
    "        index renamed to the ontology name, and added to the results table as a new column with\n",
    "        the same name.\n",
    "    ontologies : Optional[Union[Set[str], Dict[str, str]]], default=None\n",
    "        Either:\n",
    "        - Set of columns to treat as ontologies (these should be entries in ONTOLOGIES_LIST )\n",
    "        - Dict mapping wide column names to ontology names in the ONTOLOGIES_LIST controlled vocabulary\n",
    "        - None to automatically detect valid ontology columns based on ONTOLOGIES_LIST\n",
    "\n",
    "        If index_which_ontology is defined, it should be represented in these ontologies. \n",
    "    verbose : bool, optional\n",
    "        Whether to print verbose output.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    pd.DataFrame\n",
    "        A DataFrame containing the formatted results.\n",
    "    \"\"\"\n",
    "\n",
    "    # pull out the table containing results\n",
    "    raw_results_table = _load_raw_table(adata, table_type, table_name)\n",
    "\n",
    "    if table_type == ADATA.VAR:\n",
    "        var_table = copy.deepcopy(raw_results_table)\n",
    "    else:\n",
    "        var_table = adata.var\n",
    "\n",
    "\n",
    "    return raw_results_table\n",
    "\n",
    "prepare_anndata_results_df(\n",
    "    mdata[\"proteomics\"],\n",
    "    table_type = \"X\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'table_type' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mNameError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[109]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[43mtable_type\u001b[49m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m [*ADATA_DICTLIKE_ATTRS, *ADATA_IDENTITY_ATTRS]:\n\u001b[32m      2\u001b[39m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mtable_type \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mtable_type\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m is not a valid AnnData attribute. Valid attributes are: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mADATA_DICTLIKE_ATTRS\u001b[38;5;250m \u001b[39m+\u001b[38;5;250m \u001b[39mADATA_IDENTITY_ATTRS\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m\"\u001b[39m)\n\u001b[32m      4\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m table_type \u001b[38;5;129;01min\u001b[39;00m ADATA_IDENTITY_ATTRS:\n",
      "\u001b[31mNameError\u001b[39m: name 'table_type' is not defined"
     ]
    }
   ],
   "source": [
    "table_type = \"var\"\n",
    "\n",
    "if table_type not in [*ADATA_DICTLIKE_ATTRS, *ADATA_IDENTITY_ATTRS]:\n",
    "        raise ValueError(f\"table_type {table_type} is not a valid AnnData attribute. Valid attributes are: {ADATA_DICTLIKE_ATTRS + ADATA_IDENTITY_ATTRS}\")\n",
    "\n",
    "if table_type in ADATA_IDENTITY_ATTRS:\n",
    "    if table_name is not None:\n",
    "        logger.debug(f\"table_name {table_name} is not None, but table_type is in IDENTITY_TABLES. \"\n",
    "                    f\"table_name will be ignored.\")\n",
    "    return getattr(adata, table_type)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Layers with keys: log2_centered"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.layers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "numpy.ndarray"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# if table_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MMA001</th>\n",
       "      <th>MMA004</th>\n",
       "      <th>MMA005</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>uniprot</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A0AVF1</th>\n",
       "      <td>83727.578125</td>\n",
       "      <td>351463.812500</td>\n",
       "      <td>473821.687500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0AVT1</th>\n",
       "      <td>37115.953125</td>\n",
       "      <td>71438.554688</td>\n",
       "      <td>42789.007812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0FGR8</th>\n",
       "      <td>40237.117188</td>\n",
       "      <td>122243.023438</td>\n",
       "      <td>152602.468750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1AG_BOVINAlpha-1-acidglycoproteinOS=BostaurusGN=ORM1PE=2SV=1;CONT_Q3SZR3</th>\n",
       "      <td>46269.468750</td>\n",
       "      <td>70375.281250</td>\n",
       "      <td>71832.406250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1L0T0</th>\n",
       "      <td>55125.847656</td>\n",
       "      <td>40272.636719</td>\n",
       "      <td>49587.910156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9Y6R0</th>\n",
       "      <td>28617.105469</td>\n",
       "      <td>75366.789062</td>\n",
       "      <td>20555.519531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9Y6R4</th>\n",
       "      <td>75033.437500</td>\n",
       "      <td>28757.421875</td>\n",
       "      <td>44098.183594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9Y6U3</th>\n",
       "      <td>325.786499</td>\n",
       "      <td>41348.601562</td>\n",
       "      <td>17314.455078</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9Y6W5</th>\n",
       "      <td>84969.218750</td>\n",
       "      <td>137218.062500</td>\n",
       "      <td>134523.421875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9Y6Y8</th>\n",
       "      <td>74944.687500</td>\n",
       "      <td>131124.015625</td>\n",
       "      <td>90398.140625</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4788 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                          MMA001  \\\n",
       "uniprot                                                            \n",
       "A0AVF1                                              83727.578125   \n",
       "A0AVT1                                              37115.953125   \n",
       "A0FGR8                                              40237.117188   \n",
       "A1AG_BOVINAlpha-1-acidglycoproteinOS=BostaurusG...  46269.468750   \n",
       "A1L0T0                                              55125.847656   \n",
       "...                                                          ...   \n",
       "Q9Y6R0                                              28617.105469   \n",
       "Q9Y6R4                                              75033.437500   \n",
       "Q9Y6U3                                                325.786499   \n",
       "Q9Y6W5                                              84969.218750   \n",
       "Q9Y6Y8                                              74944.687500   \n",
       "\n",
       "                                                           MMA004  \\\n",
       "uniprot                                                             \n",
       "A0AVF1                                              351463.812500   \n",
       "A0AVT1                                               71438.554688   \n",
       "A0FGR8                                              122243.023438   \n",
       "A1AG_BOVINAlpha-1-acidglycoproteinOS=BostaurusG...   70375.281250   \n",
       "A1L0T0                                               40272.636719   \n",
       "...                                                           ...   \n",
       "Q9Y6R0                                               75366.789062   \n",
       "Q9Y6R4                                               28757.421875   \n",
       "Q9Y6U3                                               41348.601562   \n",
       "Q9Y6W5                                              137218.062500   \n",
       "Q9Y6Y8                                              131124.015625   \n",
       "\n",
       "                                                           MMA005  \n",
       "uniprot                                                            \n",
       "A0AVF1                                              473821.687500  \n",
       "A0AVT1                                               42789.007812  \n",
       "A0FGR8                                              152602.468750  \n",
       "A1AG_BOVINAlpha-1-acidglycoproteinOS=BostaurusG...   71832.406250  \n",
       "A1L0T0                                               49587.910156  \n",
       "...                                                           ...  \n",
       "Q9Y6R0                                               20555.519531  \n",
       "Q9Y6R4                                               44098.183594  \n",
       "Q9Y6U3                                               17314.455078  \n",
       "Q9Y6W5                                              134523.421875  \n",
       "Q9Y6Y8                                               90398.140625  \n",
       "\n",
       "[4788 rows x 3 columns]"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "adata = mdata[\"proteomics\"]\n",
    "table_type = \"X\"\n",
    "table_name = None\n",
    "results_attrs = [\"MMA001\", \"MMA004\", \"MMA005\"]\n",
    "\n",
    "raw_results_table = _load_raw_table(adata, table_type, table_name)\n",
    "\n",
    "if table_type == ADATA.VAR:\n",
    "    var_table = copy.deepcopy(raw_results_table)\n",
    "else:\n",
    "    var_table = adata.var\n",
    "\n",
    "# select relevant attributes returning a pd.DataFrame\n",
    "# if raw_results_table is a np.ndarray select observations\n",
    "# based on their primary key if results_attrs is not None\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# selecting from array by name\n",
    "_select_results_attrs(adata, adata.X, [\"MMA001\", \"MMA004\", \"MMA005\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(221, 4788)"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def _select_results_attrs(\n",
    "    adata: anndata.AnnData,\n",
    "    raw_results_table: Union[pd.DataFrame, np.ndarray],\n",
    "    results_attrs: Optional[List[str]] = None\n",
    ") -> pd.DataFrame:\n",
    "\n",
    "    if isinstance(raw_results_table, pd.DataFrame):\n",
    "        if results_attrs is not None:\n",
    "            results_table_data = raw_results_table.loc[results_attrs]\n",
    "        else:\n",
    "            results_table_data = raw_results_table\n",
    "    else:\n",
    "        if results_attrs is not None:\n",
    "            # Check that results_attrs exist in adata.obs.index\n",
    "            valid_obs = adata.obs.index.tolist()\n",
    "            \n",
    "            invalid_results_attrs = [x for x in results_attrs if x not in valid_obs]\n",
    "            if len(invalid_results_attrs) > 0:\n",
    "                raise ValueError(f\"The following results attributes are not present in the AnnData object's obs index: {invalid_results_attrs}\")\n",
    "\n",
    "            # Find positions of desired rows in adata.obs.index\n",
    "            row_positions = [adata.obs.index.get_loc(attr) for attr in results_attrs]\n",
    "            \n",
    "            # Select ROWS from numpy array using positions\n",
    "            selected_array = raw_results_table[row_positions, :]\n",
    "            \n",
    "            # Convert to DataFrame and set row names to results_attrs\n",
    "            results_table_data = pd.DataFrame(\n",
    "                selected_array,\n",
    "                index = results_attrs,\n",
    "                columns = adata.var.index\n",
    "                ).T\n",
    "        else:\n",
    "            # Convert entire array to DataFrame\n",
    "            results_table_data = pd.DataFrame(\n",
    "                raw_results_table,\n",
    "                index = adata.obs.index,\n",
    "                columns = adata.var.index\n",
    "            ).T\n",
    "\n",
    "    return results_table_data\n",
    "adata.X.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "valid_obs = adata.obs.index.tolist()\n",
    "            \n",
    "invalid_results_attrs = [x for x in results_attrs if x not in valid_obs]\n",
    "if len(invalid_results_attrs) > 0:\n",
    "    raise ValueError(f\"The following results attributes are not present in the AnnData object's obs index: {invalid_results_attrs}\")\n",
    "\n",
    "# Find positions of desired rows in adata.obs.index\n",
    "row_positions = [adata.obs.index.get_loc(attr) for attr in results_attrs]\n",
    "\n",
    "# Select ROWS from numpy array using positions\n",
    "selected_array = raw_results_table[row_positions, :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MMA001</th>\n",
       "      <th>MMA004</th>\n",
       "      <th>MMA005</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>uniprot</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A0AVF1</th>\n",
       "      <td>83727.578125</td>\n",
       "      <td>351463.812500</td>\n",
       "      <td>473821.687500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0AVT1</th>\n",
       "      <td>37115.953125</td>\n",
       "      <td>71438.554688</td>\n",
       "      <td>42789.007812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0FGR8</th>\n",
       "      <td>40237.117188</td>\n",
       "      <td>122243.023438</td>\n",
       "      <td>152602.468750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1AG_BOVINAlpha-1-acidglycoproteinOS=BostaurusGN=ORM1PE=2SV=1;CONT_Q3SZR3</th>\n",
       "      <td>46269.468750</td>\n",
       "      <td>70375.281250</td>\n",
       "      <td>71832.406250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1L0T0</th>\n",
       "      <td>55125.847656</td>\n",
       "      <td>40272.636719</td>\n",
       "      <td>49587.910156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9Y6R0</th>\n",
       "      <td>28617.105469</td>\n",
       "      <td>75366.789062</td>\n",
       "      <td>20555.519531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9Y6R4</th>\n",
       "      <td>75033.437500</td>\n",
       "      <td>28757.421875</td>\n",
       "      <td>44098.183594</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9Y6U3</th>\n",
       "      <td>325.786499</td>\n",
       "      <td>41348.601562</td>\n",
       "      <td>17314.455078</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9Y6W5</th>\n",
       "      <td>84969.218750</td>\n",
       "      <td>137218.062500</td>\n",
       "      <td>134523.421875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9Y6Y8</th>\n",
       "      <td>74944.687500</td>\n",
       "      <td>131124.015625</td>\n",
       "      <td>90398.140625</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4788 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                          MMA001  \\\n",
       "uniprot                                                            \n",
       "A0AVF1                                              83727.578125   \n",
       "A0AVT1                                              37115.953125   \n",
       "A0FGR8                                              40237.117188   \n",
       "A1AG_BOVINAlpha-1-acidglycoproteinOS=BostaurusG...  46269.468750   \n",
       "A1L0T0                                              55125.847656   \n",
       "...                                                          ...   \n",
       "Q9Y6R0                                              28617.105469   \n",
       "Q9Y6R4                                              75033.437500   \n",
       "Q9Y6U3                                                325.786499   \n",
       "Q9Y6W5                                              84969.218750   \n",
       "Q9Y6Y8                                              74944.687500   \n",
       "\n",
       "                                                           MMA004  \\\n",
       "uniprot                                                             \n",
       "A0AVF1                                              351463.812500   \n",
       "A0AVT1                                               71438.554688   \n",
       "A0FGR8                                              122243.023438   \n",
       "A1AG_BOVINAlpha-1-acidglycoproteinOS=BostaurusG...   70375.281250   \n",
       "A1L0T0                                               40272.636719   \n",
       "...                                                           ...   \n",
       "Q9Y6R0                                               75366.789062   \n",
       "Q9Y6R4                                               28757.421875   \n",
       "Q9Y6U3                                               41348.601562   \n",
       "Q9Y6W5                                              137218.062500   \n",
       "Q9Y6Y8                                              131124.015625   \n",
       "\n",
       "                                                           MMA005  \n",
       "uniprot                                                            \n",
       "A0AVF1                                              473821.687500  \n",
       "A0AVT1                                               42789.007812  \n",
       "A0FGR8                                              152602.468750  \n",
       "A1AG_BOVINAlpha-1-acidglycoproteinOS=BostaurusG...   71832.406250  \n",
       "A1L0T0                                               49587.910156  \n",
       "...                                                           ...  \n",
       "Q9Y6R0                                               20555.519531  \n",
       "Q9Y6R4                                               44098.183594  \n",
       "Q9Y6U3                                               17314.455078  \n",
       "Q9Y6W5                                              134523.421875  \n",
       "Q9Y6Y8                                               90398.140625  \n",
       "\n",
       "[4788 rows x 3 columns]"
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(\n",
    "    selected_array,\n",
    "    index = results_attrs,\n",
    "    columns = adata.var.index\n",
    "    ).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case</th>\n",
       "      <th>gender</th>\n",
       "      <th>consanguinity</th>\n",
       "      <th>mut_category</th>\n",
       "      <th>wgs_zygosity</th>\n",
       "      <th>acidosis</th>\n",
       "      <th>metabolic_acidosis</th>\n",
       "      <th>metabolic_ketoacidosis</th>\n",
       "      <th>ketosis</th>\n",
       "      <th>hyperammonemia</th>\n",
       "      <th>...</th>\n",
       "      <th>ammonia_umolL</th>\n",
       "      <th>pH</th>\n",
       "      <th>base_excess</th>\n",
       "      <th>MMA_urine_after_treat</th>\n",
       "      <th>carnitine_dose</th>\n",
       "      <th>natural_protein_amount</th>\n",
       "      <th>total_protein_amount</th>\n",
       "      <th>weight_centile_quant</th>\n",
       "      <th>length_centile_quant</th>\n",
       "      <th>head_circumfernce_quant</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>patient_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MMA001</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.762349</td>\n",
       "      <td>0.521214</td>\n",
       "      <td>-15.36</td>\n",
       "      <td>0.533427</td>\n",
       "      <td>5.643856</td>\n",
       "      <td>-0.316992</td>\n",
       "      <td>1.222182</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMA002</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>9.732642</td>\n",
       "      <td>-0.077624</td>\n",
       "      <td>-21.00</td>\n",
       "      <td>0.105576</td>\n",
       "      <td>6.643856</td>\n",
       "      <td>-0.104195</td>\n",
       "      <td>1.397998</td>\n",
       "      <td>1.732051</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMA003</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>8.422174</td>\n",
       "      <td>-1.664230</td>\n",
       "      <td>-24.00</td>\n",
       "      <td>-0.514988</td>\n",
       "      <td>5.862947</td>\n",
       "      <td>-1.442128</td>\n",
       "      <td>1.411885</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMA004</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>8.666115</td>\n",
       "      <td>-0.641109</td>\n",
       "      <td>-16.94</td>\n",
       "      <td>-0.037999</td>\n",
       "      <td>6.715085</td>\n",
       "      <td>0.752554</td>\n",
       "      <td>1.096611</td>\n",
       "      <td>1.732051</td>\n",
       "      <td>1.732051</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMA005</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.108113</td>\n",
       "      <td>0.833357</td>\n",
       "      <td>-1.96</td>\n",
       "      <td>0.249966</td>\n",
       "      <td>6.513956</td>\n",
       "      <td>-0.103818</td>\n",
       "      <td>1.660840</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMA226</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.267153</td>\n",
       "      <td>0.414281</td>\n",
       "      <td>-3.76</td>\n",
       "      <td>-0.354647</td>\n",
       "      <td>6.360849</td>\n",
       "      <td>0.127875</td>\n",
       "      <td>0.886244</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMA227</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>5.746581</td>\n",
       "      <td>1.230584</td>\n",
       "      <td>-2.30</td>\n",
       "      <td>-0.824862</td>\n",
       "      <td>6.974927</td>\n",
       "      <td>0.081066</td>\n",
       "      <td>1.818446</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMA228</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>7.519280</td>\n",
       "      <td>0.982586</td>\n",
       "      <td>-6.10</td>\n",
       "      <td>0.060576</td>\n",
       "      <td>6.643856</td>\n",
       "      <td>0.127875</td>\n",
       "      <td>0.907549</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMA229</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>6.209349</td>\n",
       "      <td>1.144137</td>\n",
       "      <td>-4.36</td>\n",
       "      <td>-0.379918</td>\n",
       "      <td>6.643856</td>\n",
       "      <td>0.261158</td>\n",
       "      <td>1.271238</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MMA230</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>6.271202</td>\n",
       "      <td>1.093729</td>\n",
       "      <td>-4.86</td>\n",
       "      <td>0.091434</td>\n",
       "      <td>7.205327</td>\n",
       "      <td>0.166777</td>\n",
       "      <td>1.035001</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>221 rows × 89 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            case  gender  consanguinity  mut_category  wgs_zygosity  acidosis  \\\n",
       "patient_id                                                                      \n",
       "MMA001         1       1              0           1.0           1.0         1   \n",
       "MMA002         1       1              1           0.0           0.0         0   \n",
       "MMA003         1       1              1           0.0           1.0         0   \n",
       "MMA004         1       0              0           0.0           1.0         1   \n",
       "MMA005         1       0              0           0.0           0.0         0   \n",
       "...          ...     ...            ...           ...           ...       ...   \n",
       "MMA226         0       0              0           NaN           NaN         0   \n",
       "MMA227         0       0              1           NaN           NaN         0   \n",
       "MMA228         0       0              0           NaN           NaN         0   \n",
       "MMA229         0       1              0           NaN           NaN         0   \n",
       "MMA230         0       0              0           NaN           NaN         0   \n",
       "\n",
       "            metabolic_acidosis  metabolic_ketoacidosis  ketosis  \\\n",
       "patient_id                                                        \n",
       "MMA001                       1                       0        0   \n",
       "MMA002                       0                       0        0   \n",
       "MMA003                       0                       0        0   \n",
       "MMA004                       1                       0        0   \n",
       "MMA005                       0                       0        0   \n",
       "...                        ...                     ...      ...   \n",
       "MMA226                       0                       0        0   \n",
       "MMA227                       0                       0        0   \n",
       "MMA228                       0                       0        0   \n",
       "MMA229                       0                       0        0   \n",
       "MMA230                       0                       0        0   \n",
       "\n",
       "            hyperammonemia  ...  ammonia_umolL        pH  base_excess  \\\n",
       "patient_id                  ...                                         \n",
       "MMA001                   0  ...       3.762349  0.521214       -15.36   \n",
       "MMA002                   0  ...       9.732642 -0.077624       -21.00   \n",
       "MMA003                   0  ...       8.422174 -1.664230       -24.00   \n",
       "MMA004                   0  ...       8.666115 -0.641109       -16.94   \n",
       "MMA005                   0  ...       7.108113  0.833357        -1.96   \n",
       "...                    ...  ...            ...       ...          ...   \n",
       "MMA226                   0  ...       7.267153  0.414281        -3.76   \n",
       "MMA227                   0  ...       5.746581  1.230584        -2.30   \n",
       "MMA228                   0  ...       7.519280  0.982586        -6.10   \n",
       "MMA229                   0  ...       6.209349  1.144137        -4.36   \n",
       "MMA230                   0  ...       6.271202  1.093729        -4.86   \n",
       "\n",
       "            MMA_urine_after_treat  carnitine_dose  natural_protein_amount  \\\n",
       "patient_id                                                                  \n",
       "MMA001                   0.533427        5.643856               -0.316992   \n",
       "MMA002                   0.105576        6.643856               -0.104195   \n",
       "MMA003                  -0.514988        5.862947               -1.442128   \n",
       "MMA004                  -0.037999        6.715085                0.752554   \n",
       "MMA005                   0.249966        6.513956               -0.103818   \n",
       "...                           ...             ...                     ...   \n",
       "MMA226                  -0.354647        6.360849                0.127875   \n",
       "MMA227                  -0.824862        6.974927                0.081066   \n",
       "MMA228                   0.060576        6.643856                0.127875   \n",
       "MMA229                  -0.379918        6.643856                0.261158   \n",
       "MMA230                   0.091434        7.205327                0.166777   \n",
       "\n",
       "            total_protein_amount  weight_centile_quant  length_centile_quant  \\\n",
       "patient_id                                                                     \n",
       "MMA001                  1.222182              0.000000              0.000000   \n",
       "MMA002                  1.397998              1.732051              1.000000   \n",
       "MMA003                  1.411885              1.000000              1.000000   \n",
       "MMA004                  1.096611              1.732051              1.732051   \n",
       "MMA005                  1.660840              0.000000              0.000000   \n",
       "...                          ...                   ...                   ...   \n",
       "MMA226                  0.886244              0.000000              0.000000   \n",
       "MMA227                  1.818446              0.000000              0.000000   \n",
       "MMA228                  0.907549              0.000000              0.000000   \n",
       "MMA229                  1.271238              0.000000              0.000000   \n",
       "MMA230                  1.035001              0.000000              0.000000   \n",
       "\n",
       "            head_circumfernce_quant  \n",
       "patient_id                           \n",
       "MMA001                          0.0  \n",
       "MMA002                          3.0  \n",
       "MMA003                          3.0  \n",
       "MMA004                          0.0  \n",
       "MMA005                          0.0  \n",
       "...                             ...  \n",
       "MMA226                          0.0  \n",
       "MMA227                          0.0  \n",
       "MMA228                          0.0  \n",
       "MMA229                          0.0  \n",
       "MMA230                          0.0  \n",
       "\n",
       "[221 rows x 89 columns]"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['MMA001', 'MMA002', 'MMA003', 'MMA004', 'MMA005', 'MMA006', 'MMA007',\n",
       "       'MMA008', 'MMA009', 'MMA010',\n",
       "       ...\n",
       "       'MMA220', 'MMA222', 'MMA223', 'MMA224', 'MMA225', 'MMA226', 'MMA227',\n",
       "       'MMA228', 'MMA229', 'MMA230'],\n",
       "      dtype='object', name='patient_id', length=221)"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:napistu.sbml_dfs_utils:Running in non-dogmatic mode - genes, transcripts, and proteins will be merged if possible.\n",
      "DEBUG:napistu.mechanism_matching:Validated ontology columns: {'uniprot'}\n",
      "WARNING:napistu.mechanism_matching:No feature_id column found in DataFrame, creating one\n",
      "INFO:napistu.mechanism_matching:Using columns as results: ['feature_id', 'effect_case', 'qval_case']\n",
      "DEBUG:napistu.mechanism_matching:Final long format shape: (4788, 5)\n",
      "DEBUG:napistu.mechanism_matching:Matching 4788 features to 124813 species for ontology uniprot\n",
      "INFO:napistu.mechanism_matching:Found 7062 total matches across 1 ontologies\n",
      "INFO:napistu.mechanism_matching:96.2% of feature_ids are present one or more times in the output (4607/4788)\n",
      "INFO:napistu.mechanism_matching:64 s_id(s) map to more than one feature_id.\n",
      "INFO:napistu.mechanism_matching:Examples of s_id mapping to multiple feature_ids (showing up to 3):\n",
      "s_id       s_name     \n",
      "S00000766  KMT2B          [1046, 1058, 1538]\n",
      "S00001060  PML                   [836, 1335]\n",
      "S00001352  UGT1A9 gene          [4321, 4322]\n",
      "Name: feature_id, dtype: object\n",
      "INFO:napistu.mechanism_matching:903 feature_id(s) map to more than one s_id.\n",
      "INFO:napistu.mechanism_matching:Examples of feature_id mapping to multiple s_ids (showing up to 3):\n",
      "feature_id\n",
      "151        [(S00015206, AGPS), (S00046307, AGPS(1-658))]\n",
      "152    [(S00031276, kinesin family member 2A), (S0004...\n",
      "154    [(S00024094, DExD-box helicase 39A), (S0004133...\n",
      "dtype: object\n",
      "INFO:napistu.sbml_dfs_utils:Running in non-dogmatic mode - genes, transcripts, and proteins will be merged if possible.\n",
      "DEBUG:napistu.mechanism_matching:Validated ontology columns: {'ensembl_gene'}\n",
      "WARNING:napistu.mechanism_matching:No feature_id column found in DataFrame, creating one\n",
      "INFO:napistu.mechanism_matching:Using columns as results: ['feature_id', 'effect_case', 'qval_case']\n",
      "DEBUG:napistu.mechanism_matching:Final long format shape: (9134, 5)\n",
      "DEBUG:napistu.mechanism_matching:Matching 9134 features to 42421 species for ontology ensembl_gene\n",
      "INFO:napistu.mechanism_matching:Found 18475 total matches across 1 ontologies\n",
      "INFO:napistu.mechanism_matching:98.6% of feature_ids are present one or more times in the output (9010/9134)\n",
      "INFO:napistu.mechanism_matching:1493 s_id(s) map to more than one feature_id.\n",
      "INFO:napistu.mechanism_matching:Examples of s_id mapping to multiple feature_ids (showing up to 3):\n",
      "s_id       s_name                                 \n",
      "S00000105  Ferritin Complex                           [1507, 4696]\n",
      "S00000106  Golgi-associated Vesicle Cargo             [7615, 9013]\n",
      "S00000122  Golgi-associated Vesicle Destined Cargo    [7615, 9013]\n",
      "Name: feature_id, dtype: object\n",
      "INFO:napistu.mechanism_matching:3207 feature_id(s) map to more than one s_id.\n",
      "INFO:napistu.mechanism_matching:Examples of feature_id mapping to multiple s_ids (showing up to 3):\n",
      "feature_id\n",
      "1    [(S00014190, NOC2L), (S00014141, NOC2L:p-S15,S...\n",
      "3    [(S00012093, agrin), (S00039569, XylS-AGRN(30-...\n",
      "5    [(S00007733, B3GALT6), (S00039676, B3GALT6 P67...\n",
      "dtype: object\n"
     ]
    }
   ],
   "source": [
    "MODALITY_TO_ONTOLOGY = {\n",
    "    \"proteomics\": \"uniprot\",\n",
    "    \"transcriptomics\": \"ensembl_gene\",\n",
    "}\n",
    "\n",
    "for k, v in MODALITY_TO_ONTOLOGY.items():\n",
    "    # results from var\n",
    "    var_level_results = mdata[k].var[[\"effect_case\", \"qval_case\"]].copy()\n",
    "    var_level_results.index.name = \"feature_id\"\n",
    "    var_level_results[v] = var_level_results.index.to_series()\n",
    "\n",
    "    mechanism_matching.bind_wide_results(\n",
    "        sbml_dfs,\n",
    "        var_level_results,\n",
    "        f\"{k}_var_level_results\",\n",
    "        ontologies = {v},\n",
    "        dogmatic = False,\n",
    "        verbose = True\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[31mSignature:\u001b[39m\n",
      "mechanism_matching.bind_wide_results(\n",
      "    sbml_dfs: \u001b[33m'sbml_dfs_core.SBML_dfs'\u001b[39m,\n",
      "    results_df: \u001b[33m'pd.DataFrame'\u001b[39m,\n",
      "    results_name: \u001b[33m'str'\u001b[39m,\n",
      "    ontologies: \u001b[33m'Optional[Union[Set[str], Dict[str, str]]]'\u001b[39m = \u001b[38;5;28;01mNone\u001b[39;00m,\n",
      "    dogmatic: \u001b[33m'bool'\u001b[39m = \u001b[38;5;28;01mFalse\u001b[39;00m,\n",
      "    species_identifiers: \u001b[33m'Optional[pd.DataFrame]'\u001b[39m = \u001b[38;5;28;01mNone\u001b[39;00m,\n",
      "    feature_id_var: \u001b[33m'str'\u001b[39m = \u001b[33m'feature_id'\u001b[39m,\n",
      "    numeric_agg: \u001b[33m'str'\u001b[39m = \u001b[33m'weighted_mean'\u001b[39m,\n",
      "    keep_id_col: \u001b[33m'bool'\u001b[39m = \u001b[38;5;28;01mTrue\u001b[39;00m,\n",
      "    verbose: \u001b[33m'bool'\u001b[39m = \u001b[38;5;28;01mFalse\u001b[39;00m,\n",
      ") -> \u001b[33m'sbml_dfs_core.SBML_dfs'\u001b[39m\n",
      "\u001b[31mSource:\u001b[39m   \n",
      "\u001b[38;5;28;01mdef\u001b[39;00m bind_wide_results(\n",
      "    sbml_dfs : sbml_dfs_core.SBML_dfs,\n",
      "    results_df : pd.DataFrame,\n",
      "    results_name : str,\n",
      "    ontologies : Optional[Union[Set[str], Dict[str, str]]] = \u001b[38;5;28;01mNone\u001b[39;00m,\n",
      "    dogmatic : bool = \u001b[38;5;28;01mFalse\u001b[39;00m,\n",
      "    species_identifiers : Optional[pd.DataFrame] = \u001b[38;5;28;01mNone\u001b[39;00m,\n",
      "    feature_id_var : str = FEATURE_ID_VAR_DEFAULT,\n",
      "    numeric_agg : str = RESOLVE_MATCHES_AGGREGATORS.WEIGHTED_MEAN,\n",
      "    keep_id_col : bool = \u001b[38;5;28;01mTrue\u001b[39;00m,\n",
      "    verbose : bool = \u001b[38;5;28;01mFalse\u001b[39;00m\n",
      ") -> sbml_dfs_core.SBML_dfs:\n",
      "    \u001b[33m\"\"\"\u001b[39m\n",
      "\u001b[33m    Binds wide results to a sbml_dfs object.\u001b[39m\n",
      "\n",
      "\u001b[33m    Take a table with molecular species-level attributes tied to systematic identifiers and match them to an sbml_dfs_model transferring these attributes to species_data\u001b[39m\n",
      "\n",
      "\u001b[33m    Parameters\u001b[39m\n",
      "\u001b[33m    ----------\u001b[39m\n",
      "\u001b[33m    sbml_dfs : sbml_dfs_core.SBML_dfs\u001b[39m\n",
      "\u001b[33m        The sbml_dfs object to bind the results to.\u001b[39m\n",
      "\u001b[33m    results_df : pd.DataFrame\u001b[39m\n",
      "\u001b[33m        The table containing the results to bind.\u001b[39m\n",
      "\u001b[33m    results_name : str\u001b[39m\n",
      "\u001b[33m        The name of the results to bind.\u001b[39m\n",
      "\u001b[33m    ontologies : Optional[Union[Set[str], Dict[str, str]]]\u001b[39m\n",
      "\u001b[33m        The ontologies to use for matching.\u001b[39m\n",
      "\u001b[33m    dogmatic : bool\u001b[39m\n",
      "\u001b[33m        Whether to respect differences between genes, transcripts, and proteins (True) or ignore them (False).\u001b[39m\n",
      "\u001b[33m    species_identifiers : Optional[pd.DataFrame]\u001b[39m\n",
      "\u001b[33m        Systematic identifiers for the molecular species \"sbml_dfs\". If None this will be generate on-the-fly.\u001b[39m\n",
      "\u001b[33m    feature_id_var : str\u001b[39m\n",
      "\u001b[33m        The name of the column in the results_df that contains the feature identifiers. If this does not exist it will be created.\u001b[39m\n",
      "\u001b[33m    numeric_agg : str\u001b[39m\n",
      "\u001b[33m        The aggregation method to use for resolving degeneracy.\u001b[39m\n",
      "\u001b[33m    keep_id_col : bool\u001b[39m\n",
      "\u001b[33m        Whether to keep the identifier column in the results_df.\u001b[39m\n",
      "\u001b[33m    verbose : bool\u001b[39m\n",
      "\u001b[33m        Whether to log cases of 1-to-many and many-to-one mapping and to indicate the behavior for resolving degeneracy\u001b[39m\n",
      "\u001b[33m    \u001b[39m\n",
      "\u001b[33m    Returns\u001b[39m\n",
      "\u001b[33m    -------\u001b[39m\n",
      "\u001b[33m    sbml_dfs : sbml_dfs_core.SBML_dfs\u001b[39m\n",
      "\u001b[33m        The sbml_dfs object with the results bound.\u001b[39m\n",
      "\u001b[33m    \"\"\"\u001b[39m\n",
      "\n",
      "    species_identifiers = identifiers._prepare_species_identifiers(\n",
      "        sbml_dfs,\n",
      "        dogmatic = dogmatic,\n",
      "        species_identifiers = species_identifiers\n",
      "        )\n",
      "    \n",
      "    \u001b[38;5;66;03m# match\u001b[39;00m\n",
      "    matched_s_ids_from_wide = match_features_to_wide_pathway_species(\n",
      "        results_df,\n",
      "        species_identifiers,\n",
      "        ontologies = ontologies,\n",
      "        feature_id_var = feature_id_var,\n",
      "        verbose = verbose\n",
      "    )\n",
      "\n",
      "    disambiguated_matches = resolve_matches(\n",
      "        matched_data = matched_s_ids_from_wide,\n",
      "        feature_id_var = feature_id_var,\n",
      "        numeric_agg = numeric_agg,\n",
      "        keep_id_col = keep_id_col\n",
      "        )\n",
      "\n",
      "    clean_species_data = utils.drop_extra_cols(\n",
      "        results_df,\n",
      "        disambiguated_matches,\n",
      "        always_include = [feature_id_var]\n",
      "    )\n",
      "\n",
      "    sbml_dfs.add_species_data(\n",
      "        results_name,\n",
      "        clean_species_data\n",
      "        )\n",
      "\n",
      "    \u001b[38;5;28;01mreturn\u001b[39;00m sbml_dfs\n",
      "\u001b[31mFile:\u001b[39m      ~/Desktop/GITHUB/napistu/lib/napistu-scrapyard/applications/forny_2023/.venv/lib/python3.11/site-packages/napistu/mechanism_matching.py\n",
      "\u001b[31mType:\u001b[39m      function"
     ]
    }
   ],
   "source": [
    "??mechanism_matching.bind_wide_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'protoemics_var_level_results'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mKeyError\u001b[39m                                  Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[29]\u001b[39m\u001b[32m, line 3\u001b[39m\n\u001b[32m      1\u001b[39m sbml_dfs.species_data[\u001b[33m\"\u001b[39m\u001b[33mtranscriptomics_var_level_results\u001b[39m\u001b[33m\"\u001b[39m]\n\u001b[32m----> \u001b[39m\u001b[32m3\u001b[39m \u001b[43msbml_dfs\u001b[49m\u001b[43m.\u001b[49m\u001b[43mspecies_data\u001b[49m\u001b[43m[\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mprotoemics_var_level_results\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\n",
      "\u001b[31mKeyError\u001b[39m: 'protoemics_var_level_results'"
     ]
    }
   ],
   "source": [
    "sbml_dfs.species_data[\"transcriptomics_var_level_results\"]\n",
    "\n",
    "sbml_dfs.species_data[\"protoemics_var_level_results\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LFs1</th>\n",
       "      <th>LFs2</th>\n",
       "      <th>LFs3</th>\n",
       "      <th>LFs4</th>\n",
       "      <th>LFs5</th>\n",
       "      <th>LFs6</th>\n",
       "      <th>LFs7</th>\n",
       "      <th>LFs8</th>\n",
       "      <th>LFs9</th>\n",
       "      <th>LFs10</th>\n",
       "      <th>...</th>\n",
       "      <th>LFs24</th>\n",
       "      <th>LFs25</th>\n",
       "      <th>LFs26</th>\n",
       "      <th>LFs27</th>\n",
       "      <th>LFs28</th>\n",
       "      <th>LFs29</th>\n",
       "      <th>LFs30</th>\n",
       "      <th>ensembl_gene</th>\n",
       "      <th>modality</th>\n",
       "      <th>uniprot</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feature_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Q9H792</th>\n",
       "      <td>-0.027243</td>\n",
       "      <td>-0.119244</td>\n",
       "      <td>0.144007</td>\n",
       "      <td>0.001119</td>\n",
       "      <td>0.026899</td>\n",
       "      <td>-0.030268</td>\n",
       "      <td>0.033949</td>\n",
       "      <td>0.063107</td>\n",
       "      <td>0.009831</td>\n",
       "      <td>0.016116</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.009719</td>\n",
       "      <td>-0.003239</td>\n",
       "      <td>-0.001262</td>\n",
       "      <td>-0.214363</td>\n",
       "      <td>0.008808</td>\n",
       "      <td>0.021000</td>\n",
       "      <td>-0.003012</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>Q9H792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P0C0L4</th>\n",
       "      <td>-0.092460</td>\n",
       "      <td>-0.093116</td>\n",
       "      <td>0.178606</td>\n",
       "      <td>0.006779</td>\n",
       "      <td>0.083500</td>\n",
       "      <td>-0.000353</td>\n",
       "      <td>0.146885</td>\n",
       "      <td>0.020423</td>\n",
       "      <td>0.017907</td>\n",
       "      <td>-0.020367</td>\n",
       "      <td>...</td>\n",
       "      <td>0.015213</td>\n",
       "      <td>-0.003062</td>\n",
       "      <td>0.005986</td>\n",
       "      <td>-0.152157</td>\n",
       "      <td>-0.005441</td>\n",
       "      <td>0.001834</td>\n",
       "      <td>0.000642</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>P0C0L4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9UBI1</th>\n",
       "      <td>0.029154</td>\n",
       "      <td>-0.041047</td>\n",
       "      <td>0.379047</td>\n",
       "      <td>0.001382</td>\n",
       "      <td>-0.022916</td>\n",
       "      <td>0.054253</td>\n",
       "      <td>0.016286</td>\n",
       "      <td>-0.079876</td>\n",
       "      <td>-0.004745</td>\n",
       "      <td>0.369614</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000681</td>\n",
       "      <td>0.021992</td>\n",
       "      <td>-0.017142</td>\n",
       "      <td>-0.087458</td>\n",
       "      <td>-0.018002</td>\n",
       "      <td>0.004626</td>\n",
       "      <td>-0.006150</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>Q9UBI1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9NRL2</th>\n",
       "      <td>-0.075254</td>\n",
       "      <td>-0.009926</td>\n",
       "      <td>0.008734</td>\n",
       "      <td>0.005759</td>\n",
       "      <td>-0.008589</td>\n",
       "      <td>0.002350</td>\n",
       "      <td>0.016089</td>\n",
       "      <td>-0.150416</td>\n",
       "      <td>-0.026265</td>\n",
       "      <td>-0.092472</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.013159</td>\n",
       "      <td>-0.002475</td>\n",
       "      <td>0.001101</td>\n",
       "      <td>0.033993</td>\n",
       "      <td>-0.005234</td>\n",
       "      <td>0.005804</td>\n",
       "      <td>-0.000699</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>Q9NRL2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P37802</th>\n",
       "      <td>0.001871</td>\n",
       "      <td>-0.154841</td>\n",
       "      <td>0.074453</td>\n",
       "      <td>-0.007104</td>\n",
       "      <td>-0.077237</td>\n",
       "      <td>-0.020003</td>\n",
       "      <td>-0.023796</td>\n",
       "      <td>-0.377999</td>\n",
       "      <td>-0.058973</td>\n",
       "      <td>0.061443</td>\n",
       "      <td>...</td>\n",
       "      <td>0.021791</td>\n",
       "      <td>0.006501</td>\n",
       "      <td>0.002317</td>\n",
       "      <td>0.086762</td>\n",
       "      <td>-0.015410</td>\n",
       "      <td>0.012741</td>\n",
       "      <td>-0.008488</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>P37802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000104299</th>\n",
       "      <td>0.030109</td>\n",
       "      <td>0.009456</td>\n",
       "      <td>-0.000050</td>\n",
       "      <td>-0.032007</td>\n",
       "      <td>0.020188</td>\n",
       "      <td>0.016357</td>\n",
       "      <td>0.047936</td>\n",
       "      <td>0.000251</td>\n",
       "      <td>-0.001013</td>\n",
       "      <td>0.000048</td>\n",
       "      <td>...</td>\n",
       "      <td>0.047624</td>\n",
       "      <td>0.001612</td>\n",
       "      <td>0.020146</td>\n",
       "      <td>-0.018153</td>\n",
       "      <td>-0.205680</td>\n",
       "      <td>0.118215</td>\n",
       "      <td>-0.065097</td>\n",
       "      <td>ENSG00000104299</td>\n",
       "      <td>transcriptomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000185532</th>\n",
       "      <td>0.033874</td>\n",
       "      <td>-0.009365</td>\n",
       "      <td>0.000045</td>\n",
       "      <td>0.170626</td>\n",
       "      <td>-0.001765</td>\n",
       "      <td>-0.026931</td>\n",
       "      <td>-0.078983</td>\n",
       "      <td>-0.007647</td>\n",
       "      <td>-0.086124</td>\n",
       "      <td>-0.000040</td>\n",
       "      <td>...</td>\n",
       "      <td>0.074793</td>\n",
       "      <td>-0.037001</td>\n",
       "      <td>-0.024062</td>\n",
       "      <td>-0.019362</td>\n",
       "      <td>0.190330</td>\n",
       "      <td>-0.226037</td>\n",
       "      <td>0.071660</td>\n",
       "      <td>ENSG00000185532</td>\n",
       "      <td>transcriptomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000112715</th>\n",
       "      <td>0.049501</td>\n",
       "      <td>-0.115302</td>\n",
       "      <td>-0.000124</td>\n",
       "      <td>0.016265</td>\n",
       "      <td>-0.077598</td>\n",
       "      <td>0.144006</td>\n",
       "      <td>-0.157112</td>\n",
       "      <td>0.008054</td>\n",
       "      <td>-0.122357</td>\n",
       "      <td>-0.000071</td>\n",
       "      <td>...</td>\n",
       "      <td>0.179531</td>\n",
       "      <td>0.220982</td>\n",
       "      <td>-0.086252</td>\n",
       "      <td>0.262317</td>\n",
       "      <td>-0.195597</td>\n",
       "      <td>-0.027755</td>\n",
       "      <td>0.115978</td>\n",
       "      <td>ENSG00000112715</td>\n",
       "      <td>transcriptomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000174807</th>\n",
       "      <td>0.020886</td>\n",
       "      <td>0.130035</td>\n",
       "      <td>0.000283</td>\n",
       "      <td>-0.036658</td>\n",
       "      <td>-0.041395</td>\n",
       "      <td>-0.169194</td>\n",
       "      <td>0.146575</td>\n",
       "      <td>0.000694</td>\n",
       "      <td>0.198987</td>\n",
       "      <td>-0.000107</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.234533</td>\n",
       "      <td>0.248023</td>\n",
       "      <td>-0.243550</td>\n",
       "      <td>-0.150132</td>\n",
       "      <td>0.208990</td>\n",
       "      <td>-0.253592</td>\n",
       "      <td>0.004672</td>\n",
       "      <td>ENSG00000174807</td>\n",
       "      <td>transcriptomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000136986</th>\n",
       "      <td>-0.001188</td>\n",
       "      <td>-0.025556</td>\n",
       "      <td>-0.000006</td>\n",
       "      <td>0.011816</td>\n",
       "      <td>0.054273</td>\n",
       "      <td>0.077171</td>\n",
       "      <td>-0.107502</td>\n",
       "      <td>0.002933</td>\n",
       "      <td>-0.022823</td>\n",
       "      <td>-0.000134</td>\n",
       "      <td>...</td>\n",
       "      <td>0.024850</td>\n",
       "      <td>-0.032014</td>\n",
       "      <td>-0.068765</td>\n",
       "      <td>-0.010854</td>\n",
       "      <td>0.095898</td>\n",
       "      <td>-0.024257</td>\n",
       "      <td>-0.024088</td>\n",
       "      <td>ENSG00000136986</td>\n",
       "      <td>transcriptomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     LFs1      LFs2      LFs3      LFs4      LFs5      LFs6  \\\n",
       "feature_id                                                                    \n",
       "Q9H792          -0.027243 -0.119244  0.144007  0.001119  0.026899 -0.030268   \n",
       "P0C0L4          -0.092460 -0.093116  0.178606  0.006779  0.083500 -0.000353   \n",
       "Q9UBI1           0.029154 -0.041047  0.379047  0.001382 -0.022916  0.054253   \n",
       "Q9NRL2          -0.075254 -0.009926  0.008734  0.005759 -0.008589  0.002350   \n",
       "P37802           0.001871 -0.154841  0.074453 -0.007104 -0.077237 -0.020003   \n",
       "ENSG00000104299  0.030109  0.009456 -0.000050 -0.032007  0.020188  0.016357   \n",
       "ENSG00000185532  0.033874 -0.009365  0.000045  0.170626 -0.001765 -0.026931   \n",
       "ENSG00000112715  0.049501 -0.115302 -0.000124  0.016265 -0.077598  0.144006   \n",
       "ENSG00000174807  0.020886  0.130035  0.000283 -0.036658 -0.041395 -0.169194   \n",
       "ENSG00000136986 -0.001188 -0.025556 -0.000006  0.011816  0.054273  0.077171   \n",
       "\n",
       "                     LFs7      LFs8      LFs9     LFs10  ...     LFs24  \\\n",
       "feature_id                                               ...             \n",
       "Q9H792           0.033949  0.063107  0.009831  0.016116  ... -0.009719   \n",
       "P0C0L4           0.146885  0.020423  0.017907 -0.020367  ...  0.015213   \n",
       "Q9UBI1           0.016286 -0.079876 -0.004745  0.369614  ... -0.000681   \n",
       "Q9NRL2           0.016089 -0.150416 -0.026265 -0.092472  ... -0.013159   \n",
       "P37802          -0.023796 -0.377999 -0.058973  0.061443  ...  0.021791   \n",
       "ENSG00000104299  0.047936  0.000251 -0.001013  0.000048  ...  0.047624   \n",
       "ENSG00000185532 -0.078983 -0.007647 -0.086124 -0.000040  ...  0.074793   \n",
       "ENSG00000112715 -0.157112  0.008054 -0.122357 -0.000071  ...  0.179531   \n",
       "ENSG00000174807  0.146575  0.000694  0.198987 -0.000107  ... -0.234533   \n",
       "ENSG00000136986 -0.107502  0.002933 -0.022823 -0.000134  ...  0.024850   \n",
       "\n",
       "                    LFs25     LFs26     LFs27     LFs28     LFs29     LFs30  \\\n",
       "feature_id                                                                    \n",
       "Q9H792          -0.003239 -0.001262 -0.214363  0.008808  0.021000 -0.003012   \n",
       "P0C0L4          -0.003062  0.005986 -0.152157 -0.005441  0.001834  0.000642   \n",
       "Q9UBI1           0.021992 -0.017142 -0.087458 -0.018002  0.004626 -0.006150   \n",
       "Q9NRL2          -0.002475  0.001101  0.033993 -0.005234  0.005804 -0.000699   \n",
       "P37802           0.006501  0.002317  0.086762 -0.015410  0.012741 -0.008488   \n",
       "ENSG00000104299  0.001612  0.020146 -0.018153 -0.205680  0.118215 -0.065097   \n",
       "ENSG00000185532 -0.037001 -0.024062 -0.019362  0.190330 -0.226037  0.071660   \n",
       "ENSG00000112715  0.220982 -0.086252  0.262317 -0.195597 -0.027755  0.115978   \n",
       "ENSG00000174807  0.248023 -0.243550 -0.150132  0.208990 -0.253592  0.004672   \n",
       "ENSG00000136986 -0.032014 -0.068765 -0.010854  0.095898 -0.024257 -0.024088   \n",
       "\n",
       "                    ensembl_gene         modality  uniprot  \n",
       "feature_id                                                  \n",
       "Q9H792                       NaN       proteomics   Q9H792  \n",
       "P0C0L4                       NaN       proteomics   P0C0L4  \n",
       "Q9UBI1                       NaN       proteomics   Q9UBI1  \n",
       "Q9NRL2                       NaN       proteomics   Q9NRL2  \n",
       "P37802                       NaN       proteomics   P37802  \n",
       "ENSG00000104299  ENSG00000104299  transcriptomics      NaN  \n",
       "ENSG00000185532  ENSG00000185532  transcriptomics      NaN  \n",
       "ENSG00000112715  ENSG00000112715  transcriptomics      NaN  \n",
       "ENSG00000174807  ENSG00000174807  transcriptomics      NaN  \n",
       "ENSG00000136986  ENSG00000136986  transcriptomics      NaN  \n",
       "\n",
       "[10 rows x 33 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge factors with metadata\n",
    "mofa_dfs_dict = regulation.split_varm_by_modality(mdata)\n",
    "\n",
    "modality = \"transcriptomics\"\n",
    "\n",
    "mofa_df_list = list()\n",
    "for modality in mofa_dfs_dict.keys():\n",
    "\n",
    "    modality_pk = mofa_dfs_dict[modality].index.name\n",
    "    filter_col = [col for col in mofa_dfs_dict[modality] if col.startswith('LF')]\n",
    "    modality_df = mofa_dfs_dict[modality][filter_col].copy()\n",
    "    modality_df.index.name = \"feature_id\"\n",
    "    modality_df[modality_pk] = modality_df.index.to_series()\n",
    "    modality_df[\"modality\"] = modality\n",
    "\n",
    "    mofa_df_list.append(modality_df)\n",
    "\n",
    "mofa_df = pd.concat(mofa_df_list, axis=0)\n",
    "\n",
    "mofa_df.groupby(\"modality\").sample(5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:napistu.sbml_dfs_utils:Running in non-dogmatic mode - genes, transcripts, and proteins will be merged if possible.\n",
      "DEBUG:napistu.mechanism_matching:Validated ontology columns: {'uniprot', 'ensembl_gene'}\n",
      "WARNING:napistu.mechanism_matching:No feature_id column found in DataFrame, creating one\n",
      "INFO:napistu.mechanism_matching:Using columns as results: ['LFs11', 'LFs10', 'LFs13', 'LFs18', 'LFs28', 'LFs29', 'LFs19', 'LFs5', 'LFs7', 'LFs8', 'LFs12', 'LFs27', 'LFs16', 'LFs26', 'LFs9', 'LFs2', 'LFs6', 'LFs14', 'LFs15', 'LFs1', 'LFs4', 'LFs25', 'LFs17', 'LFs21', 'LFs30', 'LFs23', 'LFs24', 'LFs3', 'LFs20', 'modality', 'LFs22', 'feature_id']\n",
      "DEBUG:napistu.mechanism_matching:Final long format shape: (13922, 34)\n",
      "DEBUG:napistu.mechanism_matching:Matching 4788 features to 98 species for ontology uniprot\n",
      "WARNING:napistu.mechanism_matching:No species found for ontology: ensembl_gene\n",
      "INFO:napistu.mechanism_matching:Found 57 total matches across 1 ontologies\n",
      "INFO:napistu.mechanism_matching:0.4% of feature_ids are present one or more times in the output (57/13647)\n",
      "INFO:napistu.mechanism_matching:7 s_id(s) map to more than one feature_id.\n",
      "INFO:napistu.mechanism_matching:Examples of s_id mapping to multiple feature_ids (showing up to 3):\n",
      "s_id       s_name           \n",
      "S00000016  SLC25A12,13                 [9675, 13622]\n",
      "S00000031  enolase dimer        [9967, 10040, 10168]\n",
      "S00000050  aldolase tetramer           [9907, 10068]\n",
      "Name: feature_id, dtype: object\n",
      "INFO:napistu.mechanism_matching:0 feature_id(s) map to more than one s_id.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LFs1</th>\n",
       "      <th>LFs2</th>\n",
       "      <th>LFs3</th>\n",
       "      <th>LFs4</th>\n",
       "      <th>LFs5</th>\n",
       "      <th>LFs6</th>\n",
       "      <th>LFs7</th>\n",
       "      <th>LFs8</th>\n",
       "      <th>LFs9</th>\n",
       "      <th>LFs10</th>\n",
       "      <th>...</th>\n",
       "      <th>LFs23</th>\n",
       "      <th>LFs24</th>\n",
       "      <th>LFs25</th>\n",
       "      <th>LFs26</th>\n",
       "      <th>LFs27</th>\n",
       "      <th>LFs28</th>\n",
       "      <th>LFs29</th>\n",
       "      <th>LFs30</th>\n",
       "      <th>modality</th>\n",
       "      <th>feature_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S00000006</th>\n",
       "      <td>0.051530</td>\n",
       "      <td>0.036362</td>\n",
       "      <td>-0.041906</td>\n",
       "      <td>-0.001347</td>\n",
       "      <td>-0.013508</td>\n",
       "      <td>0.001565</td>\n",
       "      <td>-0.021276</td>\n",
       "      <td>-0.244355</td>\n",
       "      <td>-0.031964</td>\n",
       "      <td>0.057272</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.191047</td>\n",
       "      <td>0.075975</td>\n",
       "      <td>-0.006429</td>\n",
       "      <td>0.009816</td>\n",
       "      <td>0.074251</td>\n",
       "      <td>0.010391</td>\n",
       "      <td>0.000629</td>\n",
       "      <td>-0.004023</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>10646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000012</th>\n",
       "      <td>0.034465</td>\n",
       "      <td>0.115698</td>\n",
       "      <td>-0.047133</td>\n",
       "      <td>-0.001525</td>\n",
       "      <td>-0.006179</td>\n",
       "      <td>0.015768</td>\n",
       "      <td>0.042666</td>\n",
       "      <td>0.118975</td>\n",
       "      <td>0.002245</td>\n",
       "      <td>0.035891</td>\n",
       "      <td>...</td>\n",
       "      <td>0.153040</td>\n",
       "      <td>-0.007040</td>\n",
       "      <td>0.000832</td>\n",
       "      <td>0.005247</td>\n",
       "      <td>0.003451</td>\n",
       "      <td>0.000202</td>\n",
       "      <td>0.001349</td>\n",
       "      <td>-0.002890</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>13559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000013</th>\n",
       "      <td>0.111319</td>\n",
       "      <td>-0.002571</td>\n",
       "      <td>-0.073818</td>\n",
       "      <td>-0.000615</td>\n",
       "      <td>-0.066941</td>\n",
       "      <td>0.040771</td>\n",
       "      <td>-0.002893</td>\n",
       "      <td>0.120330</td>\n",
       "      <td>-0.024400</td>\n",
       "      <td>0.337955</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.075491</td>\n",
       "      <td>-0.005003</td>\n",
       "      <td>0.001831</td>\n",
       "      <td>-0.016633</td>\n",
       "      <td>-0.083339</td>\n",
       "      <td>-0.000686</td>\n",
       "      <td>0.003954</td>\n",
       "      <td>-0.002430</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>11290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000015</th>\n",
       "      <td>0.010020</td>\n",
       "      <td>0.043131</td>\n",
       "      <td>-0.000815</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>-0.032613</td>\n",
       "      <td>0.026158</td>\n",
       "      <td>0.044396</td>\n",
       "      <td>0.171346</td>\n",
       "      <td>0.001049</td>\n",
       "      <td>0.003718</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.071206</td>\n",
       "      <td>0.054874</td>\n",
       "      <td>-0.052803</td>\n",
       "      <td>-0.000263</td>\n",
       "      <td>-0.014903</td>\n",
       "      <td>0.000827</td>\n",
       "      <td>0.036917</td>\n",
       "      <td>-0.002143</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>10647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000016</th>\n",
       "      <td>0.099801</td>\n",
       "      <td>0.064125</td>\n",
       "      <td>-0.015871</td>\n",
       "      <td>-0.001283</td>\n",
       "      <td>-0.085869</td>\n",
       "      <td>0.002985</td>\n",
       "      <td>0.009716</td>\n",
       "      <td>0.216461</td>\n",
       "      <td>-0.005344</td>\n",
       "      <td>0.003299</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.044812</td>\n",
       "      <td>0.032438</td>\n",
       "      <td>-0.001762</td>\n",
       "      <td>0.001995</td>\n",
       "      <td>-0.094416</td>\n",
       "      <td>-0.005864</td>\n",
       "      <td>-0.006293</td>\n",
       "      <td>-0.005239</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>13622,9675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000019</th>\n",
       "      <td>0.064609</td>\n",
       "      <td>0.006714</td>\n",
       "      <td>-0.110616</td>\n",
       "      <td>-0.017464</td>\n",
       "      <td>-0.020428</td>\n",
       "      <td>-0.032641</td>\n",
       "      <td>-0.003856</td>\n",
       "      <td>0.083932</td>\n",
       "      <td>0.000541</td>\n",
       "      <td>-0.054995</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.121061</td>\n",
       "      <td>-0.032175</td>\n",
       "      <td>-0.006698</td>\n",
       "      <td>-0.002780</td>\n",
       "      <td>0.095618</td>\n",
       "      <td>0.002672</td>\n",
       "      <td>0.074969</td>\n",
       "      <td>-0.004676</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>10901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000022</th>\n",
       "      <td>0.034890</td>\n",
       "      <td>0.007629</td>\n",
       "      <td>-0.122095</td>\n",
       "      <td>-0.001708</td>\n",
       "      <td>-0.076321</td>\n",
       "      <td>0.061220</td>\n",
       "      <td>-0.055165</td>\n",
       "      <td>-0.110623</td>\n",
       "      <td>0.014309</td>\n",
       "      <td>0.324787</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.065491</td>\n",
       "      <td>-0.016958</td>\n",
       "      <td>-0.011363</td>\n",
       "      <td>-0.004861</td>\n",
       "      <td>-0.308404</td>\n",
       "      <td>0.006624</td>\n",
       "      <td>-0.031057</td>\n",
       "      <td>0.000834</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>10118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000031</th>\n",
       "      <td>0.014932</td>\n",
       "      <td>-0.100028</td>\n",
       "      <td>-0.056794</td>\n",
       "      <td>-0.015458</td>\n",
       "      <td>-0.039523</td>\n",
       "      <td>0.024284</td>\n",
       "      <td>-0.102462</td>\n",
       "      <td>-0.278852</td>\n",
       "      <td>-0.129555</td>\n",
       "      <td>0.008516</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.225068</td>\n",
       "      <td>0.005913</td>\n",
       "      <td>-0.008280</td>\n",
       "      <td>-0.010984</td>\n",
       "      <td>0.235486</td>\n",
       "      <td>0.014511</td>\n",
       "      <td>0.010286</td>\n",
       "      <td>0.005220</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>10040,10168,9967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000033</th>\n",
       "      <td>0.055625</td>\n",
       "      <td>-0.149947</td>\n",
       "      <td>0.003432</td>\n",
       "      <td>-0.031034</td>\n",
       "      <td>-0.020317</td>\n",
       "      <td>0.054525</td>\n",
       "      <td>-0.158043</td>\n",
       "      <td>-0.283876</td>\n",
       "      <td>-0.124461</td>\n",
       "      <td>0.013506</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.285335</td>\n",
       "      <td>0.016808</td>\n",
       "      <td>-0.004187</td>\n",
       "      <td>-0.003736</td>\n",
       "      <td>0.322549</td>\n",
       "      <td>0.018490</td>\n",
       "      <td>0.048649</td>\n",
       "      <td>-0.000388</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>9864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000036</th>\n",
       "      <td>0.006070</td>\n",
       "      <td>-0.180905</td>\n",
       "      <td>0.004306</td>\n",
       "      <td>0.004061</td>\n",
       "      <td>0.082143</td>\n",
       "      <td>0.124391</td>\n",
       "      <td>0.241261</td>\n",
       "      <td>0.136064</td>\n",
       "      <td>-0.099635</td>\n",
       "      <td>-0.021902</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.151604</td>\n",
       "      <td>0.009547</td>\n",
       "      <td>-0.011912</td>\n",
       "      <td>-0.036209</td>\n",
       "      <td>-0.212754</td>\n",
       "      <td>0.002011</td>\n",
       "      <td>0.004901</td>\n",
       "      <td>-0.001723</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>11794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000037</th>\n",
       "      <td>0.022947</td>\n",
       "      <td>-0.058032</td>\n",
       "      <td>-0.024882</td>\n",
       "      <td>-0.007361</td>\n",
       "      <td>-0.000671</td>\n",
       "      <td>0.019875</td>\n",
       "      <td>-0.090096</td>\n",
       "      <td>-0.323505</td>\n",
       "      <td>-0.037611</td>\n",
       "      <td>-0.081639</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.334929</td>\n",
       "      <td>0.008752</td>\n",
       "      <td>-0.010288</td>\n",
       "      <td>-0.000088</td>\n",
       "      <td>0.198345</td>\n",
       "      <td>-0.003193</td>\n",
       "      <td>0.013340</td>\n",
       "      <td>-0.007567</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>10272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000038</th>\n",
       "      <td>-0.027232</td>\n",
       "      <td>0.062166</td>\n",
       "      <td>0.056516</td>\n",
       "      <td>-0.000365</td>\n",
       "      <td>-0.050971</td>\n",
       "      <td>-0.002178</td>\n",
       "      <td>0.011381</td>\n",
       "      <td>0.121958</td>\n",
       "      <td>-0.002549</td>\n",
       "      <td>-0.016742</td>\n",
       "      <td>...</td>\n",
       "      <td>0.068444</td>\n",
       "      <td>0.005345</td>\n",
       "      <td>-0.030580</td>\n",
       "      <td>0.001823</td>\n",
       "      <td>-0.042488</td>\n",
       "      <td>-0.003180</td>\n",
       "      <td>0.015015</td>\n",
       "      <td>0.004315</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>9862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000040</th>\n",
       "      <td>-0.021196</td>\n",
       "      <td>-0.042606</td>\n",
       "      <td>-0.002211</td>\n",
       "      <td>-0.018687</td>\n",
       "      <td>-0.031417</td>\n",
       "      <td>-0.012370</td>\n",
       "      <td>-0.122853</td>\n",
       "      <td>-0.278146</td>\n",
       "      <td>-0.040692</td>\n",
       "      <td>0.012698</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.144409</td>\n",
       "      <td>0.023224</td>\n",
       "      <td>0.000504</td>\n",
       "      <td>0.005328</td>\n",
       "      <td>0.202808</td>\n",
       "      <td>-0.027104</td>\n",
       "      <td>0.002004</td>\n",
       "      <td>-0.001296</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>9919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000043</th>\n",
       "      <td>0.045661</td>\n",
       "      <td>-0.104146</td>\n",
       "      <td>0.041507</td>\n",
       "      <td>-0.020404</td>\n",
       "      <td>-0.029229</td>\n",
       "      <td>0.018339</td>\n",
       "      <td>-0.098265</td>\n",
       "      <td>-0.335201</td>\n",
       "      <td>-0.117354</td>\n",
       "      <td>-0.038272</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.319510</td>\n",
       "      <td>0.030442</td>\n",
       "      <td>-0.006122</td>\n",
       "      <td>0.000407</td>\n",
       "      <td>0.248316</td>\n",
       "      <td>0.003664</td>\n",
       "      <td>0.101558</td>\n",
       "      <td>-0.005887</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>11005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000049</th>\n",
       "      <td>0.092689</td>\n",
       "      <td>-0.084680</td>\n",
       "      <td>-0.006784</td>\n",
       "      <td>-0.027318</td>\n",
       "      <td>-0.039291</td>\n",
       "      <td>0.007955</td>\n",
       "      <td>-0.171292</td>\n",
       "      <td>-0.255789</td>\n",
       "      <td>-0.042421</td>\n",
       "      <td>0.013457</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.243276</td>\n",
       "      <td>-0.003953</td>\n",
       "      <td>-0.003293</td>\n",
       "      <td>-0.011791</td>\n",
       "      <td>0.166454</td>\n",
       "      <td>0.004979</td>\n",
       "      <td>0.006143</td>\n",
       "      <td>-0.005453</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>9969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000050</th>\n",
       "      <td>-0.016952</td>\n",
       "      <td>-0.074755</td>\n",
       "      <td>0.048621</td>\n",
       "      <td>-0.021370</td>\n",
       "      <td>-0.012469</td>\n",
       "      <td>0.000841</td>\n",
       "      <td>-0.127041</td>\n",
       "      <td>-0.266838</td>\n",
       "      <td>-0.075979</td>\n",
       "      <td>-0.106741</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.066520</td>\n",
       "      <td>0.010346</td>\n",
       "      <td>0.008772</td>\n",
       "      <td>-0.003996</td>\n",
       "      <td>0.271012</td>\n",
       "      <td>0.006833</td>\n",
       "      <td>0.027319</td>\n",
       "      <td>0.003566</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>10068,9907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000052</th>\n",
       "      <td>-0.000008</td>\n",
       "      <td>-0.057172</td>\n",
       "      <td>0.076038</td>\n",
       "      <td>0.000387</td>\n",
       "      <td>0.035516</td>\n",
       "      <td>0.042100</td>\n",
       "      <td>0.065796</td>\n",
       "      <td>-0.229365</td>\n",
       "      <td>-0.121593</td>\n",
       "      <td>0.207655</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.148836</td>\n",
       "      <td>0.035526</td>\n",
       "      <td>-0.015478</td>\n",
       "      <td>-0.003401</td>\n",
       "      <td>-0.081515</td>\n",
       "      <td>0.110851</td>\n",
       "      <td>-0.011741</td>\n",
       "      <td>-0.003127</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>10235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000053</th>\n",
       "      <td>0.009466</td>\n",
       "      <td>-0.047191</td>\n",
       "      <td>0.001708</td>\n",
       "      <td>-0.010746</td>\n",
       "      <td>-0.046002</td>\n",
       "      <td>0.009029</td>\n",
       "      <td>-0.080002</td>\n",
       "      <td>-0.216579</td>\n",
       "      <td>-0.048743</td>\n",
       "      <td>0.183879</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.112031</td>\n",
       "      <td>-0.012816</td>\n",
       "      <td>-0.003062</td>\n",
       "      <td>-0.004971</td>\n",
       "      <td>0.070585</td>\n",
       "      <td>0.000649</td>\n",
       "      <td>0.012074</td>\n",
       "      <td>0.000814</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>10014,10253,11267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000055</th>\n",
       "      <td>0.061756</td>\n",
       "      <td>-0.018183</td>\n",
       "      <td>0.122387</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>-0.003894</td>\n",
       "      <td>0.031492</td>\n",
       "      <td>0.011038</td>\n",
       "      <td>0.067939</td>\n",
       "      <td>-0.011241</td>\n",
       "      <td>-0.801744</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.089089</td>\n",
       "      <td>-0.001327</td>\n",
       "      <td>-0.000028</td>\n",
       "      <td>0.007375</td>\n",
       "      <td>0.043694</td>\n",
       "      <td>-0.008876</td>\n",
       "      <td>0.001447</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>12937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000062</th>\n",
       "      <td>-0.006338</td>\n",
       "      <td>0.022238</td>\n",
       "      <td>0.065426</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>0.002146</td>\n",
       "      <td>0.019219</td>\n",
       "      <td>0.011993</td>\n",
       "      <td>0.063045</td>\n",
       "      <td>-0.008383</td>\n",
       "      <td>0.100418</td>\n",
       "      <td>...</td>\n",
       "      <td>0.107770</td>\n",
       "      <td>0.020965</td>\n",
       "      <td>0.003218</td>\n",
       "      <td>0.007524</td>\n",
       "      <td>0.088904</td>\n",
       "      <td>-0.002463</td>\n",
       "      <td>-0.008899</td>\n",
       "      <td>0.000117</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>10136,10583,10611,10801,10802,10966,10995,1118...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000063</th>\n",
       "      <td>0.098795</td>\n",
       "      <td>0.021605</td>\n",
       "      <td>0.038138</td>\n",
       "      <td>0.002267</td>\n",
       "      <td>-0.018920</td>\n",
       "      <td>0.028503</td>\n",
       "      <td>0.007167</td>\n",
       "      <td>-0.139382</td>\n",
       "      <td>-0.041067</td>\n",
       "      <td>0.129286</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.043905</td>\n",
       "      <td>0.012089</td>\n",
       "      <td>0.020392</td>\n",
       "      <td>-0.002435</td>\n",
       "      <td>0.145502</td>\n",
       "      <td>-0.010664</td>\n",
       "      <td>-0.021019</td>\n",
       "      <td>-0.002354</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>10727,12455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000066</th>\n",
       "      <td>-0.000947</td>\n",
       "      <td>-0.001950</td>\n",
       "      <td>0.111818</td>\n",
       "      <td>-0.004127</td>\n",
       "      <td>-0.043805</td>\n",
       "      <td>-0.002237</td>\n",
       "      <td>-0.056070</td>\n",
       "      <td>-0.013065</td>\n",
       "      <td>0.035485</td>\n",
       "      <td>-0.181588</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.160488</td>\n",
       "      <td>0.003556</td>\n",
       "      <td>-0.019579</td>\n",
       "      <td>-0.007897</td>\n",
       "      <td>0.101904</td>\n",
       "      <td>0.004617</td>\n",
       "      <td>0.014658</td>\n",
       "      <td>-0.001804</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>10281,10890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000069</th>\n",
       "      <td>0.017298</td>\n",
       "      <td>-0.051139</td>\n",
       "      <td>-0.205668</td>\n",
       "      <td>-0.001418</td>\n",
       "      <td>-0.084057</td>\n",
       "      <td>-0.069092</td>\n",
       "      <td>-0.022792</td>\n",
       "      <td>0.021800</td>\n",
       "      <td>-0.035827</td>\n",
       "      <td>0.166415</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.072593</td>\n",
       "      <td>0.001357</td>\n",
       "      <td>-0.001061</td>\n",
       "      <td>0.004199</td>\n",
       "      <td>0.086592</td>\n",
       "      <td>-0.000150</td>\n",
       "      <td>0.004470</td>\n",
       "      <td>-0.000942</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>9996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000071</th>\n",
       "      <td>-0.023183</td>\n",
       "      <td>-0.022540</td>\n",
       "      <td>0.091783</td>\n",
       "      <td>0.002548</td>\n",
       "      <td>0.035444</td>\n",
       "      <td>-0.031894</td>\n",
       "      <td>0.036761</td>\n",
       "      <td>0.020390</td>\n",
       "      <td>-0.000618</td>\n",
       "      <td>0.359657</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.026749</td>\n",
       "      <td>-0.003443</td>\n",
       "      <td>0.002296</td>\n",
       "      <td>-0.000439</td>\n",
       "      <td>-0.018022</td>\n",
       "      <td>-0.002475</td>\n",
       "      <td>-0.001200</td>\n",
       "      <td>-0.000593</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>12023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000073</th>\n",
       "      <td>-0.058798</td>\n",
       "      <td>0.045985</td>\n",
       "      <td>0.062197</td>\n",
       "      <td>-0.000034</td>\n",
       "      <td>0.020415</td>\n",
       "      <td>0.042837</td>\n",
       "      <td>-0.007339</td>\n",
       "      <td>-0.048396</td>\n",
       "      <td>-0.020258</td>\n",
       "      <td>-0.849610</td>\n",
       "      <td>...</td>\n",
       "      <td>0.128460</td>\n",
       "      <td>0.000203</td>\n",
       "      <td>0.002437</td>\n",
       "      <td>0.001223</td>\n",
       "      <td>-0.009079</td>\n",
       "      <td>-0.003331</td>\n",
       "      <td>0.000729</td>\n",
       "      <td>-0.001151</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>9146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000075</th>\n",
       "      <td>-0.001263</td>\n",
       "      <td>0.015386</td>\n",
       "      <td>0.055444</td>\n",
       "      <td>0.001262</td>\n",
       "      <td>0.003094</td>\n",
       "      <td>-0.029072</td>\n",
       "      <td>0.020157</td>\n",
       "      <td>-0.146970</td>\n",
       "      <td>0.002175</td>\n",
       "      <td>0.003175</td>\n",
       "      <td>...</td>\n",
       "      <td>0.164392</td>\n",
       "      <td>-0.000421</td>\n",
       "      <td>-0.001488</td>\n",
       "      <td>-0.000659</td>\n",
       "      <td>0.093662</td>\n",
       "      <td>-0.017453</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>-0.003100</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>11612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000078</th>\n",
       "      <td>-0.007932</td>\n",
       "      <td>-0.004116</td>\n",
       "      <td>-0.164399</td>\n",
       "      <td>-0.001136</td>\n",
       "      <td>-0.025260</td>\n",
       "      <td>0.012151</td>\n",
       "      <td>-0.022979</td>\n",
       "      <td>-0.088234</td>\n",
       "      <td>-0.053451</td>\n",
       "      <td>0.054101</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.033478</td>\n",
       "      <td>-0.027109</td>\n",
       "      <td>0.001039</td>\n",
       "      <td>0.015863</td>\n",
       "      <td>-0.034967</td>\n",
       "      <td>0.000603</td>\n",
       "      <td>0.013189</td>\n",
       "      <td>-0.001627</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>11800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000079</th>\n",
       "      <td>0.042610</td>\n",
       "      <td>-0.034332</td>\n",
       "      <td>0.025165</td>\n",
       "      <td>-0.001721</td>\n",
       "      <td>-0.004204</td>\n",
       "      <td>-0.011743</td>\n",
       "      <td>-0.018016</td>\n",
       "      <td>-0.173415</td>\n",
       "      <td>0.016935</td>\n",
       "      <td>0.087673</td>\n",
       "      <td>...</td>\n",
       "      <td>0.156711</td>\n",
       "      <td>0.002057</td>\n",
       "      <td>0.047498</td>\n",
       "      <td>0.013202</td>\n",
       "      <td>-0.017072</td>\n",
       "      <td>0.004374</td>\n",
       "      <td>-0.009348</td>\n",
       "      <td>0.004558</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>10244</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               LFs1      LFs2      LFs3      LFs4      LFs5      LFs6  \\\n",
       "s_id                                                                    \n",
       "S00000006  0.051530  0.036362 -0.041906 -0.001347 -0.013508  0.001565   \n",
       "S00000012  0.034465  0.115698 -0.047133 -0.001525 -0.006179  0.015768   \n",
       "S00000013  0.111319 -0.002571 -0.073818 -0.000615 -0.066941  0.040771   \n",
       "S00000015  0.010020  0.043131 -0.000815  0.000010 -0.032613  0.026158   \n",
       "S00000016  0.099801  0.064125 -0.015871 -0.001283 -0.085869  0.002985   \n",
       "S00000019  0.064609  0.006714 -0.110616 -0.017464 -0.020428 -0.032641   \n",
       "S00000022  0.034890  0.007629 -0.122095 -0.001708 -0.076321  0.061220   \n",
       "S00000031  0.014932 -0.100028 -0.056794 -0.015458 -0.039523  0.024284   \n",
       "S00000033  0.055625 -0.149947  0.003432 -0.031034 -0.020317  0.054525   \n",
       "S00000036  0.006070 -0.180905  0.004306  0.004061  0.082143  0.124391   \n",
       "S00000037  0.022947 -0.058032 -0.024882 -0.007361 -0.000671  0.019875   \n",
       "S00000038 -0.027232  0.062166  0.056516 -0.000365 -0.050971 -0.002178   \n",
       "S00000040 -0.021196 -0.042606 -0.002211 -0.018687 -0.031417 -0.012370   \n",
       "S00000043  0.045661 -0.104146  0.041507 -0.020404 -0.029229  0.018339   \n",
       "S00000049  0.092689 -0.084680 -0.006784 -0.027318 -0.039291  0.007955   \n",
       "S00000050 -0.016952 -0.074755  0.048621 -0.021370 -0.012469  0.000841   \n",
       "S00000052 -0.000008 -0.057172  0.076038  0.000387  0.035516  0.042100   \n",
       "S00000053  0.009466 -0.047191  0.001708 -0.010746 -0.046002  0.009029   \n",
       "S00000055  0.061756 -0.018183  0.122387  0.000016 -0.003894  0.031492   \n",
       "S00000062 -0.006338  0.022238  0.065426  0.000109  0.002146  0.019219   \n",
       "S00000063  0.098795  0.021605  0.038138  0.002267 -0.018920  0.028503   \n",
       "S00000066 -0.000947 -0.001950  0.111818 -0.004127 -0.043805 -0.002237   \n",
       "S00000069  0.017298 -0.051139 -0.205668 -0.001418 -0.084057 -0.069092   \n",
       "S00000071 -0.023183 -0.022540  0.091783  0.002548  0.035444 -0.031894   \n",
       "S00000073 -0.058798  0.045985  0.062197 -0.000034  0.020415  0.042837   \n",
       "S00000075 -0.001263  0.015386  0.055444  0.001262  0.003094 -0.029072   \n",
       "S00000078 -0.007932 -0.004116 -0.164399 -0.001136 -0.025260  0.012151   \n",
       "S00000079  0.042610 -0.034332  0.025165 -0.001721 -0.004204 -0.011743   \n",
       "\n",
       "               LFs7      LFs8      LFs9     LFs10  ...     LFs23     LFs24  \\\n",
       "s_id                                               ...                       \n",
       "S00000006 -0.021276 -0.244355 -0.031964  0.057272  ... -0.191047  0.075975   \n",
       "S00000012  0.042666  0.118975  0.002245  0.035891  ...  0.153040 -0.007040   \n",
       "S00000013 -0.002893  0.120330 -0.024400  0.337955  ... -0.075491 -0.005003   \n",
       "S00000015  0.044396  0.171346  0.001049  0.003718  ... -0.071206  0.054874   \n",
       "S00000016  0.009716  0.216461 -0.005344  0.003299  ... -0.044812  0.032438   \n",
       "S00000019 -0.003856  0.083932  0.000541 -0.054995  ... -0.121061 -0.032175   \n",
       "S00000022 -0.055165 -0.110623  0.014309  0.324787  ... -0.065491 -0.016958   \n",
       "S00000031 -0.102462 -0.278852 -0.129555  0.008516  ... -0.225068  0.005913   \n",
       "S00000033 -0.158043 -0.283876 -0.124461  0.013506  ... -0.285335  0.016808   \n",
       "S00000036  0.241261  0.136064 -0.099635 -0.021902  ... -0.151604  0.009547   \n",
       "S00000037 -0.090096 -0.323505 -0.037611 -0.081639  ... -0.334929  0.008752   \n",
       "S00000038  0.011381  0.121958 -0.002549 -0.016742  ...  0.068444  0.005345   \n",
       "S00000040 -0.122853 -0.278146 -0.040692  0.012698  ... -0.144409  0.023224   \n",
       "S00000043 -0.098265 -0.335201 -0.117354 -0.038272  ... -0.319510  0.030442   \n",
       "S00000049 -0.171292 -0.255789 -0.042421  0.013457  ... -0.243276 -0.003953   \n",
       "S00000050 -0.127041 -0.266838 -0.075979 -0.106741  ... -0.066520  0.010346   \n",
       "S00000052  0.065796 -0.229365 -0.121593  0.207655  ... -0.148836  0.035526   \n",
       "S00000053 -0.080002 -0.216579 -0.048743  0.183879  ... -0.112031 -0.012816   \n",
       "S00000055  0.011038  0.067939 -0.011241 -0.801744  ... -0.089089 -0.001327   \n",
       "S00000062  0.011993  0.063045 -0.008383  0.100418  ...  0.107770  0.020965   \n",
       "S00000063  0.007167 -0.139382 -0.041067  0.129286  ... -0.043905  0.012089   \n",
       "S00000066 -0.056070 -0.013065  0.035485 -0.181588  ... -0.160488  0.003556   \n",
       "S00000069 -0.022792  0.021800 -0.035827  0.166415  ... -0.072593  0.001357   \n",
       "S00000071  0.036761  0.020390 -0.000618  0.359657  ... -0.026749 -0.003443   \n",
       "S00000073 -0.007339 -0.048396 -0.020258 -0.849610  ...  0.128460  0.000203   \n",
       "S00000075  0.020157 -0.146970  0.002175  0.003175  ...  0.164392 -0.000421   \n",
       "S00000078 -0.022979 -0.088234 -0.053451  0.054101  ... -0.033478 -0.027109   \n",
       "S00000079 -0.018016 -0.173415  0.016935  0.087673  ...  0.156711  0.002057   \n",
       "\n",
       "              LFs25     LFs26     LFs27     LFs28     LFs29     LFs30  \\\n",
       "s_id                                                                    \n",
       "S00000006 -0.006429  0.009816  0.074251  0.010391  0.000629 -0.004023   \n",
       "S00000012  0.000832  0.005247  0.003451  0.000202  0.001349 -0.002890   \n",
       "S00000013  0.001831 -0.016633 -0.083339 -0.000686  0.003954 -0.002430   \n",
       "S00000015 -0.052803 -0.000263 -0.014903  0.000827  0.036917 -0.002143   \n",
       "S00000016 -0.001762  0.001995 -0.094416 -0.005864 -0.006293 -0.005239   \n",
       "S00000019 -0.006698 -0.002780  0.095618  0.002672  0.074969 -0.004676   \n",
       "S00000022 -0.011363 -0.004861 -0.308404  0.006624 -0.031057  0.000834   \n",
       "S00000031 -0.008280 -0.010984  0.235486  0.014511  0.010286  0.005220   \n",
       "S00000033 -0.004187 -0.003736  0.322549  0.018490  0.048649 -0.000388   \n",
       "S00000036 -0.011912 -0.036209 -0.212754  0.002011  0.004901 -0.001723   \n",
       "S00000037 -0.010288 -0.000088  0.198345 -0.003193  0.013340 -0.007567   \n",
       "S00000038 -0.030580  0.001823 -0.042488 -0.003180  0.015015  0.004315   \n",
       "S00000040  0.000504  0.005328  0.202808 -0.027104  0.002004 -0.001296   \n",
       "S00000043 -0.006122  0.000407  0.248316  0.003664  0.101558 -0.005887   \n",
       "S00000049 -0.003293 -0.011791  0.166454  0.004979  0.006143 -0.005453   \n",
       "S00000050  0.008772 -0.003996  0.271012  0.006833  0.027319  0.003566   \n",
       "S00000052 -0.015478 -0.003401 -0.081515  0.110851 -0.011741 -0.003127   \n",
       "S00000053 -0.003062 -0.004971  0.070585  0.000649  0.012074  0.000814   \n",
       "S00000055 -0.000028  0.007375  0.043694 -0.008876  0.001447  0.000043   \n",
       "S00000062  0.003218  0.007524  0.088904 -0.002463 -0.008899  0.000117   \n",
       "S00000063  0.020392 -0.002435  0.145502 -0.010664 -0.021019 -0.002354   \n",
       "S00000066 -0.019579 -0.007897  0.101904  0.004617  0.014658 -0.001804   \n",
       "S00000069 -0.001061  0.004199  0.086592 -0.000150  0.004470 -0.000942   \n",
       "S00000071  0.002296 -0.000439 -0.018022 -0.002475 -0.001200 -0.000593   \n",
       "S00000073  0.002437  0.001223 -0.009079 -0.003331  0.000729 -0.001151   \n",
       "S00000075 -0.001488 -0.000659  0.093662 -0.017453  0.000033 -0.003100   \n",
       "S00000078  0.001039  0.015863 -0.034967  0.000603  0.013189 -0.001627   \n",
       "S00000079  0.047498  0.013202 -0.017072  0.004374 -0.009348  0.004558   \n",
       "\n",
       "             modality                                         feature_id  \n",
       "s_id                                                                      \n",
       "S00000006  proteomics                                              10646  \n",
       "S00000012  proteomics                                              13559  \n",
       "S00000013  proteomics                                              11290  \n",
       "S00000015  proteomics                                              10647  \n",
       "S00000016  proteomics                                         13622,9675  \n",
       "S00000019  proteomics                                              10901  \n",
       "S00000022  proteomics                                              10118  \n",
       "S00000031  proteomics                                   10040,10168,9967  \n",
       "S00000033  proteomics                                               9864  \n",
       "S00000036  proteomics                                              11794  \n",
       "S00000037  proteomics                                              10272  \n",
       "S00000038  proteomics                                               9862  \n",
       "S00000040  proteomics                                               9919  \n",
       "S00000043  proteomics                                              11005  \n",
       "S00000049  proteomics                                               9969  \n",
       "S00000050  proteomics                                         10068,9907  \n",
       "S00000052  proteomics                                              10235  \n",
       "S00000053  proteomics                                  10014,10253,11267  \n",
       "S00000055  proteomics                                              12937  \n",
       "S00000062  proteomics  10136,10583,10611,10801,10802,10966,10995,1118...  \n",
       "S00000063  proteomics                                        10727,12455  \n",
       "S00000066  proteomics                                        10281,10890  \n",
       "S00000069  proteomics                                               9996  \n",
       "S00000071  proteomics                                              12023  \n",
       "S00000073  proteomics                                               9146  \n",
       "S00000075  proteomics                                              11612  \n",
       "S00000078  proteomics                                              11800  \n",
       "S00000079  proteomics                                              10244  \n",
       "\n",
       "[28 rows x 32 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mechanism_matching.bind_wide_results(\n",
    "    sbml_dfs,\n",
    "    mofa_df,\n",
    "    \"mudata_varm_results\",\n",
    "    ontologies = {\"uniprot\", \"ensembl_gene\"},\n",
    "    species_identifiers = identifiers,\n",
    "    dogmatic = False,\n",
    "    verbose = True\n",
    ")\n",
    "\n",
    "sbml_dfs.species_data[\"mudata_varm_results\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:napistu.network.net_create:Organizing all network nodes (compartmentalized species and reactions)\n",
      "INFO:napistu.network.net_create:Formatting edges as a regulatory graph\n",
      "INFO:napistu.network.net_create:Formatting 250 reactions species as tiered edges.\n",
      "INFO:napistu.network.net_create:Adding additional attributes to edges, e.g., # of children and parents.\n",
      "INFO:napistu.network.net_create:Done preparing regulatory graph\n",
      "INFO:napistu.network.net_create:Adding reversibility and other meta-data from reactions_data\n",
      "INFO:napistu.network.net_create:No reactions annotations provided in \"graph_attrs\"; returning None\n",
      "INFO:napistu.network.net_create:Creating reverse reactions for reversible reactions on a directed graph\n",
      "INFO:napistu.network.net_create:Formatting cpr_graph output\n",
      "INFO:napistu.network.net_create:Adding meta-data from species_data\n",
      "INFO:napistu.network.net_create:Adding new attribute LFs5 to vertices\n",
      "INFO:napistu.network.net_create:Adding new attribute effect_case to vertices\n"
     ]
    }
   ],
   "source": [
    "from napistu.network import net_create\n",
    "\n",
    "# now we can pass these species_data attributes to the graph\n",
    "\n",
    "reaction_graph_attrs = {\n",
    "    \"species\": {\n",
    "        \"LFs5\": {\n",
    "            \"table\": \"mudata_varm_results\",\n",
    "            \"variable\": \"LFs5\",\n",
    "            \"trans\": \"abs\",\n",
    "        },\n",
    "        \"effect_case\": {\n",
    "            \"table\": \"var_level_results\",\n",
    "            \"variable\": \"effect_case\",\n",
    "            \"trans\": \"abs\",\n",
    "        },\n",
    "    },\n",
    "}\n",
    "\n",
    "cpr_graph = net_create.create_cpr_graph(\n",
    "    sbml_dfs,\n",
    "    directed=True,\n",
    "    graph_type=\"regulatory\"\n",
    ")\n",
    "\n",
    "# add species attributes\n",
    "# TO DO - this is definitely not a utility function\n",
    "graph_w_annotations = net_create._add_graph_species_attribute(\n",
    "    cpr_graph,\n",
    "    sbml_dfs,\n",
    "    species_graph_attrs = reaction_graph_attrs,\n",
    "    custom_transformations = {\n",
    "        # take the absolute value\n",
    "        \"abs\" : lambda x: abs(x)\n",
    "    }\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_a210a th {\n",
       "  color: limegreen;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_a210a\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th id=\"T_a210a_level0_col0\" class=\"col_heading level0 col0\" >name</th>\n",
       "      <th id=\"T_a210a_level0_col1\" class=\"col_heading level0 col1\" >node_name</th>\n",
       "      <th id=\"T_a210a_level0_col2\" class=\"col_heading level0 col2\" >node_type</th>\n",
       "      <th id=\"T_a210a_level0_col3\" class=\"col_heading level0 col3\" >LFs5</th>\n",
       "      <th id=\"T_a210a_level0_col4\" class=\"col_heading level0 col4\" >effect_case</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th class=\"index_name level0\" >vertex ID</th>\n",
       "      <th class=\"blank col0\" >&nbsp;</th>\n",
       "      <th class=\"blank col1\" >&nbsp;</th>\n",
       "      <th class=\"blank col2\" >&nbsp;</th>\n",
       "      <th class=\"blank col3\" >&nbsp;</th>\n",
       "      <th class=\"blank col4\" >&nbsp;</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_a210a_level0_row0\" class=\"row_heading level0 row0\" >0</th>\n",
       "      <td id=\"T_a210a_row0_col0\" class=\"data row0 col0\" >species_113780</td>\n",
       "      <td id=\"T_a210a_row0_col1\" class=\"data row0 col1\" >Glc [endoplasmic reticulum lumen]</td>\n",
       "      <td id=\"T_a210a_row0_col2\" class=\"data row0 col2\" >species</td>\n",
       "      <td id=\"T_a210a_row0_col3\" class=\"data row0 col3\" >0.000</td>\n",
       "      <td id=\"T_a210a_row0_col4\" class=\"data row0 col4\" >0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_a210a_level0_row1\" class=\"row_heading level0 row1\" >101</th>\n",
       "      <td id=\"T_a210a_row1_col0\" class=\"data row1 col0\" >species_76116</td>\n",
       "      <td id=\"T_a210a_row1_col1\" class=\"data row1 col1\" >Glycerol [cytosol]</td>\n",
       "      <td id=\"T_a210a_row1_col2\" class=\"data row1 col2\" >species</td>\n",
       "      <td id=\"T_a210a_row1_col3\" class=\"data row1 col3\" >0.000</td>\n",
       "      <td id=\"T_a210a_row1_col4\" class=\"data row1 col4\" >0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_a210a_level0_row2\" class=\"row_heading level0 row2\" >103</th>\n",
       "      <td id=\"T_a210a_row2_col0\" class=\"data row2 col0\" >species_163745</td>\n",
       "      <td id=\"T_a210a_row2_col1\" class=\"data row2 col1\" >phosphoPFKFB1 dimer [cytosol]</td>\n",
       "      <td id=\"T_a210a_row2_col2\" class=\"data row2 col2\" >species</td>\n",
       "      <td id=\"T_a210a_row2_col3\" class=\"data row2 col3\" >0.000</td>\n",
       "      <td id=\"T_a210a_row2_col4\" class=\"data row2 col4\" >0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_a210a_level0_row3\" class=\"row_heading level0 row3\" >104</th>\n",
       "      <td id=\"T_a210a_row3_col0\" class=\"data row3 col0\" >species_71786</td>\n",
       "      <td id=\"T_a210a_row3_col1\" class=\"data row3 col1\" >PFKFB1 dimer [cytosol]</td>\n",
       "      <td id=\"T_a210a_row3_col2\" class=\"data row3 col2\" >species</td>\n",
       "      <td id=\"T_a210a_row3_col3\" class=\"data row3 col3\" >0.000</td>\n",
       "      <td id=\"T_a210a_row3_col4\" class=\"data row3 col4\" >0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_a210a_level0_row4\" class=\"row_heading level0 row4\" >107</th>\n",
       "      <td id=\"T_a210a_row4_col0\" class=\"data row4 col0\" >reaction_198458</td>\n",
       "      <td id=\"T_a210a_row4_col1\" class=\"data row4 col1\" >Efflux of glucose from the endoplasmic reticulum</td>\n",
       "      <td id=\"T_a210a_row4_col2\" class=\"data row4 col2\" >reaction</td>\n",
       "      <td id=\"T_a210a_row4_col3\" class=\"data row4 col3\" >0.000</td>\n",
       "      <td id=\"T_a210a_row4_col4\" class=\"data row4 col4\" >0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x166941950>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "napistu_utils.style_df(graph_w_annotations.get_vertex_dataframe().sort_values(\"LFs5\").head(5))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Network Propagation\n",
    "\n",
    "Here we'll implement a workflow for applying network propagation to a cpr_graph's vertex attributes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>pagerank_by_attribute</th>\n",
       "      <th>effect_case</th>\n",
       "      <th>pagerank_uniform</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>species_376856</td>\n",
       "      <td>0.019399</td>\n",
       "      <td>0.290775</td>\n",
       "      <td>0.006498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>species_8955798</td>\n",
       "      <td>0.016729</td>\n",
       "      <td>0.250761</td>\n",
       "      <td>0.006498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>species_372815</td>\n",
       "      <td>0.014503</td>\n",
       "      <td>0.217395</td>\n",
       "      <td>0.006498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>species_8955670</td>\n",
       "      <td>0.014410</td>\n",
       "      <td>0.215993</td>\n",
       "      <td>0.006498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>species_70579</td>\n",
       "      <td>0.014029</td>\n",
       "      <td>0.210290</td>\n",
       "      <td>0.006498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>species_29356</td>\n",
       "      <td>0.015775</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>species_29420</td>\n",
       "      <td>0.002402</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>species_113528</td>\n",
       "      <td>0.002402</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>species_29438</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>reaction_163773</td>\n",
       "      <td>0.005719</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.007572</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>157 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                name  pagerank_by_attribute  effect_case  pagerank_uniform\n",
       "18    species_376856               0.019399     0.290775          0.006498\n",
       "100  species_8955798               0.016729     0.250761          0.006498\n",
       "54    species_372815               0.014503     0.217395          0.006498\n",
       "98   species_8955670               0.014410     0.215993          0.006498\n",
       "72     species_70579               0.014029     0.210290          0.006498\n",
       "..               ...                    ...          ...               ...\n",
       "47     species_29356               0.015775     0.000000          0.014823\n",
       "45     species_29420               0.002402     0.000000          0.001767\n",
       "44    species_113528               0.002402     0.000000          0.001767\n",
       "43     species_29438               0.000000     0.000000          0.000000\n",
       "156  reaction_163773               0.005719     0.000000          0.007572\n",
       "\n",
       "[157 rows x 4 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RESET_PROPORTIONAL_TO = \"effect_case\"\n",
    "\n",
    "net_propagation.personalized_pagerank_by_attribute(\n",
    "    graph_w_annotations,\n",
    "    RESET_PROPORTIONAL_TO\n",
    ").sort_values(\"effect_case\", ascending=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
