{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Network propagation development\n",
    "\n",
    "The notebook currently covers how results from a AnnData/MuData object can be added to a cpr_graph to setup network-based inference. The general strategy is to:\n",
    "1. pull out a pd.DataFrame containing feature-level measures of interest along with feature metadata.\n",
    "2. These are then mapped on the species ids in an sbml_dfs model by shared on ontology, disambiguated (to handle mapping of multiple features to the same s_id), and s_id-indexed results are embedded in the sbml_dfs as a table in species_data\n",
    "3. attributes of interrest are then passed from the sbml_dfs model into the graph.\n",
    "\n",
    "This example uses real MuData results but only a small sbml_dfs object which has uniprot but not ENSG identifiers. This makes things easy to work with but a genome-scale graph will need to be used for a real analysis.\n",
    "\n",
    "Reflecting on the current functionality,\n",
    "\n",
    "(1) is not too hard but the interface can probably be cleaned up as we should have a function which applies 1-3 in a single call.\n",
    "(2) is in pretty good shape following a LOT of new functionality being added to napistu-py for handling many-to-one mappings and wide/nested formats for identifiers.\n",
    "(3) will need some better functionality since the reaction_attrs syntax is pretty cryptic but the core functionality is all there.\n",
    "\n",
    "Next, steps will be develop basic PPR functionality."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "from napistu.ingestion import sbml\n",
    "from napistu import sbml_dfs_core\n",
    "from napistu import mechanism_matching"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DEBUG:h5py._conv:Creating converter from 7 to 5\n",
      "DEBUG:h5py._conv:Creating converter from 5 to 7\n",
      "DEBUG:h5py._conv:Creating converter from 7 to 5\n",
      "DEBUG:h5py._conv:Creating converter from 5 to 7\n"
     ]
    }
   ],
   "source": [
    "# utils\n",
    "\n",
    "from typing import Dict\n",
    "from mudata import MuData\n",
    "\n",
    "def split_varm_by_modality(mdata: MuData, varm_key: str = 'LFs') -> Dict[str, pd.DataFrame]:\n",
    "    \"\"\"\n",
    "    Split a varm matrix by modality and join with corresponding var tables.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    mdata : mudata.MuData\n",
    "        MuData object containing multiple modalities and a varm matrix\n",
    "    varm_key : str, optional\n",
    "        Key in mdata.varm to split by modality. Default is 'LFs'\n",
    "        \n",
    "    Returns\n",
    "    -------\n",
    "    Dict[str, pd.DataFrame]\n",
    "        Dictionary with modality names as keys and DataFrames as values.\n",
    "        Each DataFrame contains the var attributes and corresponding matrix values\n",
    "        for that modality.\n",
    "        \n",
    "    Examples\n",
    "    --------\n",
    "    >>> # Split latent factors\n",
    "    >>> modality_lfs = split_varm_by_modality(mdata, varm_key='LFs')\n",
    "    >>> transcriptomics_lfs = modality_lfs['transcriptomics']\n",
    "    >>> proteomics_lfs = modality_lfs['proteomics']\n",
    "    >>>\n",
    "    >>> # Split other varm matrix\n",
    "    >>> modality_other = split_varm_by_modality(mdata, varm_key='other_matrix')\n",
    "    \"\"\"\n",
    "    if varm_key not in mdata.varm:\n",
    "        raise ValueError(f\"No '{varm_key}' matrix found in varm\")\n",
    "    \n",
    "    # Get the matrix and ensure it has the right index\n",
    "    matrix = pd.DataFrame(\n",
    "        mdata.varm[varm_key],\n",
    "        index=mdata.var_names,\n",
    "        columns=[f'{varm_key}{i+1}' for i in range(mdata.varm[varm_key].shape[1])]\n",
    "    )\n",
    "    \n",
    "    # Initialize results dictionary\n",
    "    results: Dict[str, pd.DataFrame] = {}\n",
    "    \n",
    "    # Process each modality\n",
    "    for modality in mdata.mod.keys():\n",
    "        # Get the var_names for this modality\n",
    "        mod_vars = mdata.mod[modality].var_names\n",
    "        \n",
    "        # Extract matrix values for this modality\n",
    "        mod_matrix = matrix.loc[mod_vars]\n",
    "        \n",
    "        # Get var table and ensure index matches\n",
    "        mod_var = mdata.mod[modality].var.copy()\n",
    "        \n",
    "        # Verify index alignment\n",
    "        if not mod_var.index.equals(mod_matrix.index):\n",
    "            raise ValueError(\n",
    "                f\"Index mismatch in {modality}: var table and matrix subset have different indices\"\n",
    "            )\n",
    "        \n",
    "        # Join with var table on index\n",
    "        mod_results = pd.concat([mod_var, mod_matrix], axis=1)\n",
    "        \n",
    "        # Store in results\n",
    "        results[modality] = mod_results\n",
    "    \n",
    "    return results\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "PATH_TO_TEST_DATA = os.path.expanduser(\"~/Desktop/GITHUB/napistu/lib/napistu-py/src/tests/test_data\")\n",
    "example_pathway = os.path.join(PATH_TO_TEST_DATA, \"reactome_glucose_metabolism.sbml\")\n",
    "assert os.path.exists(example_pathway)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:napistu.utils:creating an edgelist linking index levels s_id, entry and linking it to levels defined by ontology, identifier\n",
      "DEBUG:napistu.utils:label is not defined in table_schema; adding a constant (1)\n"
     ]
    }
   ],
   "source": [
    "sbml_dfs = sbml_dfs_core.SBML_dfs(sbml.SBML(example_pathway))\n",
    "\n",
    "species_identifiers = sbml_dfs.get_identifiers(\"species\").query(\"bqb == 'BQB_IS'\").query(\"ontology != 'reactome'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DEBUG:h5py._conv:Creating converter from 3 to 5\n",
      "/Users/maya/Desktop/GITHUB/napistu/lib/napistu-scrapyard/applications/forny_2023/.venv/lib/python3.11/site-packages/mudata/_core/mudata.py:1531: FutureWarning: From 0.4 .update() will not pull obs/var columns from individual modalities by default anymore. Set mudata.set_options(pull_on_update=False) to adopt the new behaviour, which will become the default. Use new pull_obs/pull_var and push_obs/push_var methods for more flexibility.\n",
      "  self._update_attr(\"var\", axis=0, join_common=join_common)\n",
      "/Users/maya/Desktop/GITHUB/napistu/lib/napistu-scrapyard/applications/forny_2023/.venv/lib/python3.11/site-packages/mudata/_core/mudata.py:1429: FutureWarning: From 0.4 .update() will not pull obs/var columns from individual modalities by default anymore. Set mudata.set_options(pull_on_update=False) to adopt the new behaviour, which will become the default. Use new pull_obs/pull_var and push_obs/push_var methods for more flexibility.\n",
      "  self._update_attr(\"obs\", axis=1, join_common=join_common)\n"
     ]
    }
   ],
   "source": [
    "# lets load the Forny results so we trying adding a few different types of tables to the sbml_dfs\n",
    "\n",
    "import mudata as md\n",
    "OPTIMAL_MODEL_H5MU_PATH = \"/tmp/mofa_optimal_model.h5mu\"\n",
    "\n",
    "mdata = md.read_h5mu(OPTIMAL_MODEL_H5MU_PATH)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:napistu.sbml_dfs_utils:Running in non-dogmatic mode - genes, transcripts, and proteins will be merged if possible.\n",
      "/Users/maya/Desktop/GITHUB/napistu/lib/napistu-scrapyard/applications/forny_2023/.venv/lib/python3.11/site-packages/napistu/sbml_dfs_core.py:1264: FutureWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.\n",
      "  promiscuous_component_identifiers = pd.Series(\n",
      "DEBUG:napistu.mechanism_matching:Validated ontology columns: {'uniprot'}\n",
      "WARNING:napistu.mechanism_matching:No feature_id column found in DataFrame, creating one\n",
      "INFO:napistu.mechanism_matching:Using columns as results: ['feature_id', 'qval_case', 'effect_case']\n",
      "DEBUG:napistu.mechanism_matching:Final long format shape: (4788, 5)\n",
      "DEBUG:napistu.mechanism_matching:Matching 4788 features to 98 species for ontology uniprot\n",
      "INFO:napistu.mechanism_matching:Found 57 total matches across 1 ontologies\n",
      "INFO:napistu.mechanism_matching:1.3% of feature_ids are present one or more times in the output (57/4513)\n",
      "INFO:napistu.mechanism_matching:7 s_id(s) map to more than one feature_id.\n",
      "INFO:napistu.mechanism_matching:Examples of s_id mapping to multiple feature_ids (showing up to 3):\n",
      "s_id       s_name           \n",
      "S00000016  SLC25A12,13               [541, 4488]\n",
      "S00000031  enolase dimer        [833, 906, 1034]\n",
      "S00000050  aldolase tetramer          [773, 934]\n",
      "Name: feature_id, dtype: object\n",
      "INFO:napistu.mechanism_matching:0 feature_id(s) map to more than one s_id.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>effect_case</th>\n",
       "      <th>qval_case</th>\n",
       "      <th>feature_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S00000006</th>\n",
       "      <td>0.064783</td>\n",
       "      <td>0.996212</td>\n",
       "      <td>1512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000012</th>\n",
       "      <td>0.059317</td>\n",
       "      <td>0.996212</td>\n",
       "      <td>4425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000013</th>\n",
       "      <td>0.290775</td>\n",
       "      <td>0.996212</td>\n",
       "      <td>2156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000015</th>\n",
       "      <td>0.079570</td>\n",
       "      <td>0.996212</td>\n",
       "      <td>1513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000016</th>\n",
       "      <td>0.021029</td>\n",
       "      <td>0.996212</td>\n",
       "      <td>4488,541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000019</th>\n",
       "      <td>0.047774</td>\n",
       "      <td>0.996212</td>\n",
       "      <td>1767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000022</th>\n",
       "      <td>0.202818</td>\n",
       "      <td>0.996212</td>\n",
       "      <td>984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000031</th>\n",
       "      <td>0.165435</td>\n",
       "      <td>0.996660</td>\n",
       "      <td>1034,833,906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000033</th>\n",
       "      <td>0.083939</td>\n",
       "      <td>0.996212</td>\n",
       "      <td>730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000036</th>\n",
       "      <td>0.217395</td>\n",
       "      <td>0.996212</td>\n",
       "      <td>2660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000037</th>\n",
       "      <td>-0.001896</td>\n",
       "      <td>0.999501</td>\n",
       "      <td>1138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000038</th>\n",
       "      <td>0.042632</td>\n",
       "      <td>0.996212</td>\n",
       "      <td>728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000040</th>\n",
       "      <td>0.040432</td>\n",
       "      <td>0.996212</td>\n",
       "      <td>785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000043</th>\n",
       "      <td>0.067086</td>\n",
       "      <td>0.996212</td>\n",
       "      <td>1871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000049</th>\n",
       "      <td>0.082118</td>\n",
       "      <td>0.996212</td>\n",
       "      <td>835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000050</th>\n",
       "      <td>0.032988</td>\n",
       "      <td>0.996212</td>\n",
       "      <td>773,934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000052</th>\n",
       "      <td>0.210290</td>\n",
       "      <td>0.996212</td>\n",
       "      <td>1101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000053</th>\n",
       "      <td>0.081554</td>\n",
       "      <td>0.996212</td>\n",
       "      <td>1119,2133,880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000055</th>\n",
       "      <td>0.129561</td>\n",
       "      <td>0.996212</td>\n",
       "      <td>3803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000062</th>\n",
       "      <td>0.017608</td>\n",
       "      <td>0.996412</td>\n",
       "      <td>1002,1449,1477,1667,1668,1832,1861,2052,2253,2...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000063</th>\n",
       "      <td>0.066135</td>\n",
       "      <td>0.996212</td>\n",
       "      <td>1593,3321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000066</th>\n",
       "      <td>0.065708</td>\n",
       "      <td>0.996212</td>\n",
       "      <td>1147,1756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000069</th>\n",
       "      <td>-0.031202</td>\n",
       "      <td>0.996212</td>\n",
       "      <td>862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000071</th>\n",
       "      <td>-0.215993</td>\n",
       "      <td>0.996212</td>\n",
       "      <td>2889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000073</th>\n",
       "      <td>-0.250761</td>\n",
       "      <td>0.996212</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000075</th>\n",
       "      <td>0.002792</td>\n",
       "      <td>0.998624</td>\n",
       "      <td>2478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000078</th>\n",
       "      <td>0.168634</td>\n",
       "      <td>0.996212</td>\n",
       "      <td>2666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000079</th>\n",
       "      <td>0.045571</td>\n",
       "      <td>0.996212</td>\n",
       "      <td>1110</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           effect_case  qval_case  \\\n",
       "s_id                                \n",
       "S00000006     0.064783   0.996212   \n",
       "S00000012     0.059317   0.996212   \n",
       "S00000013     0.290775   0.996212   \n",
       "S00000015     0.079570   0.996212   \n",
       "S00000016     0.021029   0.996212   \n",
       "S00000019     0.047774   0.996212   \n",
       "S00000022     0.202818   0.996212   \n",
       "S00000031     0.165435   0.996660   \n",
       "S00000033     0.083939   0.996212   \n",
       "S00000036     0.217395   0.996212   \n",
       "S00000037    -0.001896   0.999501   \n",
       "S00000038     0.042632   0.996212   \n",
       "S00000040     0.040432   0.996212   \n",
       "S00000043     0.067086   0.996212   \n",
       "S00000049     0.082118   0.996212   \n",
       "S00000050     0.032988   0.996212   \n",
       "S00000052     0.210290   0.996212   \n",
       "S00000053     0.081554   0.996212   \n",
       "S00000055     0.129561   0.996212   \n",
       "S00000062     0.017608   0.996412   \n",
       "S00000063     0.066135   0.996212   \n",
       "S00000066     0.065708   0.996212   \n",
       "S00000069    -0.031202   0.996212   \n",
       "S00000071    -0.215993   0.996212   \n",
       "S00000073    -0.250761   0.996212   \n",
       "S00000075     0.002792   0.998624   \n",
       "S00000078     0.168634   0.996212   \n",
       "S00000079     0.045571   0.996212   \n",
       "\n",
       "                                                  feature_id  \n",
       "s_id                                                          \n",
       "S00000006                                               1512  \n",
       "S00000012                                               4425  \n",
       "S00000013                                               2156  \n",
       "S00000015                                               1513  \n",
       "S00000016                                           4488,541  \n",
       "S00000019                                               1767  \n",
       "S00000022                                                984  \n",
       "S00000031                                       1034,833,906  \n",
       "S00000033                                                730  \n",
       "S00000036                                               2660  \n",
       "S00000037                                               1138  \n",
       "S00000038                                                728  \n",
       "S00000040                                                785  \n",
       "S00000043                                               1871  \n",
       "S00000049                                                835  \n",
       "S00000050                                            773,934  \n",
       "S00000052                                               1101  \n",
       "S00000053                                      1119,2133,880  \n",
       "S00000055                                               3803  \n",
       "S00000062  1002,1449,1477,1667,1668,1832,1861,2052,2253,2...  \n",
       "S00000063                                          1593,3321  \n",
       "S00000066                                          1147,1756  \n",
       "S00000069                                                862  \n",
       "S00000071                                               2889  \n",
       "S00000073                                                 12  \n",
       "S00000075                                               2478  \n",
       "S00000078                                               2666  \n",
       "S00000079                                               1110  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# results from var\n",
    "var_level_results = mdata[\"proteomics\"].var[[\"effect_case\", \"qval_case\"]].copy()\n",
    "var_level_results.index.name = \"feature_id\"\n",
    "var_level_results['uniprot'] = var_level_results.index.to_series()\n",
    "\n",
    "mechanism_matching.bind_wide_results(\n",
    "    sbml_dfs,\n",
    "    var_level_results,\n",
    "    \"var_level_results\",\n",
    "    ontologies = {\"uniprot\"},\n",
    "    dogmatic = False,\n",
    "    verbose = True\n",
    ")\n",
    "\n",
    "sbml_dfs.species_data[\"var_level_results\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LFs1</th>\n",
       "      <th>LFs2</th>\n",
       "      <th>LFs3</th>\n",
       "      <th>LFs4</th>\n",
       "      <th>LFs5</th>\n",
       "      <th>LFs6</th>\n",
       "      <th>LFs7</th>\n",
       "      <th>LFs8</th>\n",
       "      <th>LFs9</th>\n",
       "      <th>LFs10</th>\n",
       "      <th>...</th>\n",
       "      <th>LFs24</th>\n",
       "      <th>LFs25</th>\n",
       "      <th>LFs26</th>\n",
       "      <th>LFs27</th>\n",
       "      <th>LFs28</th>\n",
       "      <th>LFs29</th>\n",
       "      <th>LFs30</th>\n",
       "      <th>ensembl_gene</th>\n",
       "      <th>modality</th>\n",
       "      <th>uniprot</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feature_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Q15056</th>\n",
       "      <td>-0.056613</td>\n",
       "      <td>0.024216</td>\n",
       "      <td>0.043051</td>\n",
       "      <td>0.002981</td>\n",
       "      <td>0.017840</td>\n",
       "      <td>0.038009</td>\n",
       "      <td>0.030697</td>\n",
       "      <td>-0.282958</td>\n",
       "      <td>-0.003660</td>\n",
       "      <td>-0.069065</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.001474</td>\n",
       "      <td>-0.007024</td>\n",
       "      <td>0.001038</td>\n",
       "      <td>0.041995</td>\n",
       "      <td>-0.001967</td>\n",
       "      <td>0.000567</td>\n",
       "      <td>0.001311</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>Q15056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9Y295</th>\n",
       "      <td>-0.065844</td>\n",
       "      <td>0.001523</td>\n",
       "      <td>0.204267</td>\n",
       "      <td>0.008881</td>\n",
       "      <td>0.045325</td>\n",
       "      <td>-0.002938</td>\n",
       "      <td>0.036474</td>\n",
       "      <td>-0.148175</td>\n",
       "      <td>-0.027723</td>\n",
       "      <td>-0.266543</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.009055</td>\n",
       "      <td>-0.002009</td>\n",
       "      <td>0.002134</td>\n",
       "      <td>-0.136886</td>\n",
       "      <td>-0.005834</td>\n",
       "      <td>0.000696</td>\n",
       "      <td>0.003809</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>Q9Y295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q13228</th>\n",
       "      <td>0.150933</td>\n",
       "      <td>0.094928</td>\n",
       "      <td>-0.282249</td>\n",
       "      <td>0.001178</td>\n",
       "      <td>0.020215</td>\n",
       "      <td>-0.042123</td>\n",
       "      <td>-0.036351</td>\n",
       "      <td>-0.240952</td>\n",
       "      <td>0.018441</td>\n",
       "      <td>-0.011496</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001096</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>-0.004922</td>\n",
       "      <td>0.060077</td>\n",
       "      <td>-0.003805</td>\n",
       "      <td>0.008359</td>\n",
       "      <td>-0.004654</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>Q13228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9BY32</th>\n",
       "      <td>-0.024243</td>\n",
       "      <td>0.093431</td>\n",
       "      <td>0.006565</td>\n",
       "      <td>-0.001410</td>\n",
       "      <td>0.000312</td>\n",
       "      <td>-0.006907</td>\n",
       "      <td>-0.001519</td>\n",
       "      <td>-0.049320</td>\n",
       "      <td>0.016266</td>\n",
       "      <td>0.000917</td>\n",
       "      <td>...</td>\n",
       "      <td>0.015919</td>\n",
       "      <td>0.010374</td>\n",
       "      <td>-0.006496</td>\n",
       "      <td>0.076231</td>\n",
       "      <td>-0.001936</td>\n",
       "      <td>-0.018307</td>\n",
       "      <td>0.000494</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>Q9BY32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q08211</th>\n",
       "      <td>-0.001130</td>\n",
       "      <td>0.054894</td>\n",
       "      <td>0.000127</td>\n",
       "      <td>-0.003526</td>\n",
       "      <td>0.000936</td>\n",
       "      <td>0.015544</td>\n",
       "      <td>0.010073</td>\n",
       "      <td>0.151456</td>\n",
       "      <td>-0.009517</td>\n",
       "      <td>0.004257</td>\n",
       "      <td>...</td>\n",
       "      <td>0.034676</td>\n",
       "      <td>-0.004222</td>\n",
       "      <td>-0.023882</td>\n",
       "      <td>0.143186</td>\n",
       "      <td>0.006172</td>\n",
       "      <td>0.008225</td>\n",
       "      <td>0.000093</td>\n",
       "      <td>NaN</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>Q08211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000100567</th>\n",
       "      <td>0.016061</td>\n",
       "      <td>0.039098</td>\n",
       "      <td>0.000143</td>\n",
       "      <td>-0.064764</td>\n",
       "      <td>0.025970</td>\n",
       "      <td>0.066756</td>\n",
       "      <td>-0.060957</td>\n",
       "      <td>0.003253</td>\n",
       "      <td>-0.036549</td>\n",
       "      <td>0.000174</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.113759</td>\n",
       "      <td>0.141184</td>\n",
       "      <td>0.143617</td>\n",
       "      <td>-0.170461</td>\n",
       "      <td>0.073662</td>\n",
       "      <td>-0.003676</td>\n",
       "      <td>0.082745</td>\n",
       "      <td>ENSG00000100567</td>\n",
       "      <td>transcriptomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000175727</th>\n",
       "      <td>0.005928</td>\n",
       "      <td>-0.089782</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.046057</td>\n",
       "      <td>-0.110435</td>\n",
       "      <td>0.007074</td>\n",
       "      <td>0.144012</td>\n",
       "      <td>-0.000737</td>\n",
       "      <td>0.124377</td>\n",
       "      <td>-0.000796</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.061974</td>\n",
       "      <td>-0.133692</td>\n",
       "      <td>0.147478</td>\n",
       "      <td>0.101943</td>\n",
       "      <td>-0.002035</td>\n",
       "      <td>0.150796</td>\n",
       "      <td>-0.362851</td>\n",
       "      <td>ENSG00000175727</td>\n",
       "      <td>transcriptomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000056972</th>\n",
       "      <td>0.133554</td>\n",
       "      <td>0.136796</td>\n",
       "      <td>0.000035</td>\n",
       "      <td>-0.057148</td>\n",
       "      <td>0.035994</td>\n",
       "      <td>-0.095350</td>\n",
       "      <td>-0.031473</td>\n",
       "      <td>-0.000619</td>\n",
       "      <td>0.121508</td>\n",
       "      <td>-0.000028</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.086435</td>\n",
       "      <td>0.036297</td>\n",
       "      <td>-0.038404</td>\n",
       "      <td>0.248996</td>\n",
       "      <td>-0.155451</td>\n",
       "      <td>-0.027898</td>\n",
       "      <td>-0.005025</td>\n",
       "      <td>ENSG00000056972</td>\n",
       "      <td>transcriptomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000065491</th>\n",
       "      <td>-0.009292</td>\n",
       "      <td>-0.049694</td>\n",
       "      <td>-0.000006</td>\n",
       "      <td>0.028420</td>\n",
       "      <td>-0.047005</td>\n",
       "      <td>0.057724</td>\n",
       "      <td>0.001863</td>\n",
       "      <td>-0.009377</td>\n",
       "      <td>-0.035145</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>...</td>\n",
       "      <td>0.096928</td>\n",
       "      <td>0.097767</td>\n",
       "      <td>0.018235</td>\n",
       "      <td>0.132387</td>\n",
       "      <td>-0.157928</td>\n",
       "      <td>0.061822</td>\n",
       "      <td>-0.155265</td>\n",
       "      <td>ENSG00000065491</td>\n",
       "      <td>transcriptomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000971</th>\n",
       "      <td>0.152736</td>\n",
       "      <td>0.061856</td>\n",
       "      <td>-0.000089</td>\n",
       "      <td>0.031364</td>\n",
       "      <td>-0.074854</td>\n",
       "      <td>0.120164</td>\n",
       "      <td>-0.023122</td>\n",
       "      <td>0.003023</td>\n",
       "      <td>0.066621</td>\n",
       "      <td>-0.000042</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.000145</td>\n",
       "      <td>0.013402</td>\n",
       "      <td>-0.093895</td>\n",
       "      <td>-0.116899</td>\n",
       "      <td>-0.085348</td>\n",
       "      <td>-0.009503</td>\n",
       "      <td>0.083474</td>\n",
       "      <td>ENSG00000000971</td>\n",
       "      <td>transcriptomics</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     LFs1      LFs2      LFs3      LFs4      LFs5      LFs6  \\\n",
       "feature_id                                                                    \n",
       "Q15056          -0.056613  0.024216  0.043051  0.002981  0.017840  0.038009   \n",
       "Q9Y295          -0.065844  0.001523  0.204267  0.008881  0.045325 -0.002938   \n",
       "Q13228           0.150933  0.094928 -0.282249  0.001178  0.020215 -0.042123   \n",
       "Q9BY32          -0.024243  0.093431  0.006565 -0.001410  0.000312 -0.006907   \n",
       "Q08211          -0.001130  0.054894  0.000127 -0.003526  0.000936  0.015544   \n",
       "ENSG00000100567  0.016061  0.039098  0.000143 -0.064764  0.025970  0.066756   \n",
       "ENSG00000175727  0.005928 -0.089782  0.000036  0.046057 -0.110435  0.007074   \n",
       "ENSG00000056972  0.133554  0.136796  0.000035 -0.057148  0.035994 -0.095350   \n",
       "ENSG00000065491 -0.009292 -0.049694 -0.000006  0.028420 -0.047005  0.057724   \n",
       "ENSG00000000971  0.152736  0.061856 -0.000089  0.031364 -0.074854  0.120164   \n",
       "\n",
       "                     LFs7      LFs8      LFs9     LFs10  ...     LFs24  \\\n",
       "feature_id                                               ...             \n",
       "Q15056           0.030697 -0.282958 -0.003660 -0.069065  ... -0.001474   \n",
       "Q9Y295           0.036474 -0.148175 -0.027723 -0.266543  ... -0.009055   \n",
       "Q13228          -0.036351 -0.240952  0.018441 -0.011496  ...  0.001096   \n",
       "Q9BY32          -0.001519 -0.049320  0.016266  0.000917  ...  0.015919   \n",
       "Q08211           0.010073  0.151456 -0.009517  0.004257  ...  0.034676   \n",
       "ENSG00000100567 -0.060957  0.003253 -0.036549  0.000174  ... -0.113759   \n",
       "ENSG00000175727  0.144012 -0.000737  0.124377 -0.000796  ... -0.061974   \n",
       "ENSG00000056972 -0.031473 -0.000619  0.121508 -0.000028  ... -0.086435   \n",
       "ENSG00000065491  0.001863 -0.009377 -0.035145  0.000034  ...  0.096928   \n",
       "ENSG00000000971 -0.023122  0.003023  0.066621 -0.000042  ... -0.000145   \n",
       "\n",
       "                    LFs25     LFs26     LFs27     LFs28     LFs29     LFs30  \\\n",
       "feature_id                                                                    \n",
       "Q15056          -0.007024  0.001038  0.041995 -0.001967  0.000567  0.001311   \n",
       "Q9Y295          -0.002009  0.002134 -0.136886 -0.005834  0.000696  0.003809   \n",
       "Q13228           0.000060 -0.004922  0.060077 -0.003805  0.008359 -0.004654   \n",
       "Q9BY32           0.010374 -0.006496  0.076231 -0.001936 -0.018307  0.000494   \n",
       "Q08211          -0.004222 -0.023882  0.143186  0.006172  0.008225  0.000093   \n",
       "ENSG00000100567  0.141184  0.143617 -0.170461  0.073662 -0.003676  0.082745   \n",
       "ENSG00000175727 -0.133692  0.147478  0.101943 -0.002035  0.150796 -0.362851   \n",
       "ENSG00000056972  0.036297 -0.038404  0.248996 -0.155451 -0.027898 -0.005025   \n",
       "ENSG00000065491  0.097767  0.018235  0.132387 -0.157928  0.061822 -0.155265   \n",
       "ENSG00000000971  0.013402 -0.093895 -0.116899 -0.085348 -0.009503  0.083474   \n",
       "\n",
       "                    ensembl_gene         modality  uniprot  \n",
       "feature_id                                                  \n",
       "Q15056                       NaN       proteomics   Q15056  \n",
       "Q9Y295                       NaN       proteomics   Q9Y295  \n",
       "Q13228                       NaN       proteomics   Q13228  \n",
       "Q9BY32                       NaN       proteomics   Q9BY32  \n",
       "Q08211                       NaN       proteomics   Q08211  \n",
       "ENSG00000100567  ENSG00000100567  transcriptomics      NaN  \n",
       "ENSG00000175727  ENSG00000175727  transcriptomics      NaN  \n",
       "ENSG00000056972  ENSG00000056972  transcriptomics      NaN  \n",
       "ENSG00000065491  ENSG00000065491  transcriptomics      NaN  \n",
       "ENSG00000000971  ENSG00000000971  transcriptomics      NaN  \n",
       "\n",
       "[10 rows x 33 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge factors with metadata\n",
    "import utils\n",
    "mofa_dfs_dict = utils.split_varm_by_modality(mdata)\n",
    "\n",
    "modality = \"transcriptomics\"\n",
    "\n",
    "mofa_df_list = list()\n",
    "for modality in mofa_dfs_dict.keys():\n",
    "\n",
    "    modality_pk = mofa_dfs_dict[modality].index.name\n",
    "    filter_col = [col for col in mofa_dfs_dict[modality] if col.startswith('LF')]\n",
    "    modality_df = mofa_dfs_dict[modality][filter_col].copy()\n",
    "    modality_df.index.name = \"feature_id\"\n",
    "    modality_df[modality_pk] = modality_df.index.to_series()\n",
    "    modality_df[\"modality\"] = modality\n",
    "\n",
    "    mofa_df_list.append(modality_df)\n",
    "\n",
    "mofa_df = pd.concat(mofa_df_list, axis=0)\n",
    "\n",
    "mofa_df.groupby(\"modality\").sample(5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:napistu.sbml_dfs_utils:Running in non-dogmatic mode - genes, transcripts, and proteins will be merged if possible.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/maya/Desktop/GITHUB/napistu/lib/napistu-scrapyard/applications/forny_2023/.venv/lib/python3.11/site-packages/napistu/sbml_dfs_core.py:1264: FutureWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.\n",
      "  promiscuous_component_identifiers = pd.Series(\n",
      "DEBUG:napistu.mechanism_matching:Validated ontology columns: {'ensembl_gene', 'uniprot'}\n",
      "WARNING:napistu.mechanism_matching:No feature_id column found in DataFrame, creating one\n",
      "INFO:napistu.mechanism_matching:Using columns as results: ['LFs14', 'LFs16', 'LFs2', 'LFs17', 'LFs13', 'feature_id', 'LFs5', 'LFs29', 'LFs26', 'LFs3', 'LFs22', 'LFs20', 'LFs24', 'LFs15', 'LFs9', 'LFs10', 'LFs28', 'LFs6', 'LFs19', 'LFs7', 'modality', 'LFs12', 'LFs18', 'LFs11', 'LFs25', 'LFs30', 'LFs1', 'LFs27', 'LFs21', 'LFs23', 'LFs8', 'LFs4']\n",
      "DEBUG:napistu.mechanism_matching:Final long format shape: (13922, 34)\n",
      "WARNING:napistu.mechanism_matching:No species found for ontology: ensembl_gene\n",
      "DEBUG:napistu.mechanism_matching:Matching 4788 features to 98 species for ontology uniprot\n",
      "INFO:napistu.mechanism_matching:Found 57 total matches across 1 ontologies\n",
      "INFO:napistu.mechanism_matching:0.4% of feature_ids are present one or more times in the output (57/13647)\n",
      "INFO:napistu.mechanism_matching:7 s_id(s) map to more than one feature_id.\n",
      "INFO:napistu.mechanism_matching:Examples of s_id mapping to multiple feature_ids (showing up to 3):\n",
      "s_id       s_name           \n",
      "S00000016  SLC25A12,13                 [9675, 13622]\n",
      "S00000031  enolase dimer        [9967, 10040, 10168]\n",
      "S00000050  aldolase tetramer           [9907, 10068]\n",
      "Name: feature_id, dtype: object\n",
      "INFO:napistu.mechanism_matching:0 feature_id(s) map to more than one s_id.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LFs1</th>\n",
       "      <th>LFs2</th>\n",
       "      <th>LFs3</th>\n",
       "      <th>LFs4</th>\n",
       "      <th>LFs5</th>\n",
       "      <th>LFs6</th>\n",
       "      <th>LFs7</th>\n",
       "      <th>LFs8</th>\n",
       "      <th>LFs9</th>\n",
       "      <th>LFs10</th>\n",
       "      <th>...</th>\n",
       "      <th>LFs23</th>\n",
       "      <th>LFs24</th>\n",
       "      <th>LFs25</th>\n",
       "      <th>LFs26</th>\n",
       "      <th>LFs27</th>\n",
       "      <th>LFs28</th>\n",
       "      <th>LFs29</th>\n",
       "      <th>LFs30</th>\n",
       "      <th>modality</th>\n",
       "      <th>feature_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>S00000006</th>\n",
       "      <td>0.051530</td>\n",
       "      <td>0.036362</td>\n",
       "      <td>-0.041906</td>\n",
       "      <td>-0.001347</td>\n",
       "      <td>-0.013508</td>\n",
       "      <td>0.001565</td>\n",
       "      <td>-0.021276</td>\n",
       "      <td>-0.244355</td>\n",
       "      <td>-0.031964</td>\n",
       "      <td>0.057272</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.191047</td>\n",
       "      <td>0.075975</td>\n",
       "      <td>-0.006429</td>\n",
       "      <td>0.009816</td>\n",
       "      <td>0.074251</td>\n",
       "      <td>0.010391</td>\n",
       "      <td>0.000629</td>\n",
       "      <td>-0.004023</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>10646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000012</th>\n",
       "      <td>0.034465</td>\n",
       "      <td>0.115698</td>\n",
       "      <td>-0.047133</td>\n",
       "      <td>-0.001525</td>\n",
       "      <td>-0.006179</td>\n",
       "      <td>0.015768</td>\n",
       "      <td>0.042666</td>\n",
       "      <td>0.118975</td>\n",
       "      <td>0.002245</td>\n",
       "      <td>0.035891</td>\n",
       "      <td>...</td>\n",
       "      <td>0.153040</td>\n",
       "      <td>-0.007040</td>\n",
       "      <td>0.000832</td>\n",
       "      <td>0.005247</td>\n",
       "      <td>0.003451</td>\n",
       "      <td>0.000202</td>\n",
       "      <td>0.001349</td>\n",
       "      <td>-0.002890</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>13559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000013</th>\n",
       "      <td>0.111319</td>\n",
       "      <td>-0.002571</td>\n",
       "      <td>-0.073818</td>\n",
       "      <td>-0.000615</td>\n",
       "      <td>-0.066941</td>\n",
       "      <td>0.040771</td>\n",
       "      <td>-0.002893</td>\n",
       "      <td>0.120330</td>\n",
       "      <td>-0.024400</td>\n",
       "      <td>0.337955</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.075491</td>\n",
       "      <td>-0.005003</td>\n",
       "      <td>0.001831</td>\n",
       "      <td>-0.016633</td>\n",
       "      <td>-0.083339</td>\n",
       "      <td>-0.000686</td>\n",
       "      <td>0.003954</td>\n",
       "      <td>-0.002430</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>11290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000015</th>\n",
       "      <td>0.010020</td>\n",
       "      <td>0.043131</td>\n",
       "      <td>-0.000815</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>-0.032613</td>\n",
       "      <td>0.026158</td>\n",
       "      <td>0.044396</td>\n",
       "      <td>0.171346</td>\n",
       "      <td>0.001049</td>\n",
       "      <td>0.003718</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.071206</td>\n",
       "      <td>0.054874</td>\n",
       "      <td>-0.052803</td>\n",
       "      <td>-0.000263</td>\n",
       "      <td>-0.014903</td>\n",
       "      <td>0.000827</td>\n",
       "      <td>0.036917</td>\n",
       "      <td>-0.002143</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>10647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000016</th>\n",
       "      <td>0.099801</td>\n",
       "      <td>0.064125</td>\n",
       "      <td>-0.015871</td>\n",
       "      <td>-0.001283</td>\n",
       "      <td>-0.085869</td>\n",
       "      <td>0.002985</td>\n",
       "      <td>0.009716</td>\n",
       "      <td>0.216461</td>\n",
       "      <td>-0.005344</td>\n",
       "      <td>0.003299</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.044812</td>\n",
       "      <td>0.032438</td>\n",
       "      <td>-0.001762</td>\n",
       "      <td>0.001995</td>\n",
       "      <td>-0.094416</td>\n",
       "      <td>-0.005864</td>\n",
       "      <td>-0.006293</td>\n",
       "      <td>-0.005239</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>13622,9675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000019</th>\n",
       "      <td>0.064609</td>\n",
       "      <td>0.006714</td>\n",
       "      <td>-0.110616</td>\n",
       "      <td>-0.017464</td>\n",
       "      <td>-0.020428</td>\n",
       "      <td>-0.032641</td>\n",
       "      <td>-0.003856</td>\n",
       "      <td>0.083932</td>\n",
       "      <td>0.000541</td>\n",
       "      <td>-0.054995</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.121061</td>\n",
       "      <td>-0.032175</td>\n",
       "      <td>-0.006698</td>\n",
       "      <td>-0.002780</td>\n",
       "      <td>0.095618</td>\n",
       "      <td>0.002672</td>\n",
       "      <td>0.074969</td>\n",
       "      <td>-0.004676</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>10901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000022</th>\n",
       "      <td>0.034890</td>\n",
       "      <td>0.007629</td>\n",
       "      <td>-0.122095</td>\n",
       "      <td>-0.001708</td>\n",
       "      <td>-0.076321</td>\n",
       "      <td>0.061220</td>\n",
       "      <td>-0.055165</td>\n",
       "      <td>-0.110623</td>\n",
       "      <td>0.014309</td>\n",
       "      <td>0.324787</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.065491</td>\n",
       "      <td>-0.016958</td>\n",
       "      <td>-0.011363</td>\n",
       "      <td>-0.004861</td>\n",
       "      <td>-0.308404</td>\n",
       "      <td>0.006624</td>\n",
       "      <td>-0.031057</td>\n",
       "      <td>0.000834</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>10118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000031</th>\n",
       "      <td>0.014932</td>\n",
       "      <td>-0.100028</td>\n",
       "      <td>-0.056794</td>\n",
       "      <td>-0.015458</td>\n",
       "      <td>-0.039523</td>\n",
       "      <td>0.024284</td>\n",
       "      <td>-0.102462</td>\n",
       "      <td>-0.278852</td>\n",
       "      <td>-0.129555</td>\n",
       "      <td>0.008516</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.225068</td>\n",
       "      <td>0.005913</td>\n",
       "      <td>-0.008280</td>\n",
       "      <td>-0.010984</td>\n",
       "      <td>0.235486</td>\n",
       "      <td>0.014511</td>\n",
       "      <td>0.010286</td>\n",
       "      <td>0.005220</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>10040,10168,9967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000033</th>\n",
       "      <td>0.055625</td>\n",
       "      <td>-0.149947</td>\n",
       "      <td>0.003432</td>\n",
       "      <td>-0.031034</td>\n",
       "      <td>-0.020317</td>\n",
       "      <td>0.054525</td>\n",
       "      <td>-0.158043</td>\n",
       "      <td>-0.283876</td>\n",
       "      <td>-0.124461</td>\n",
       "      <td>0.013506</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.285335</td>\n",
       "      <td>0.016808</td>\n",
       "      <td>-0.004187</td>\n",
       "      <td>-0.003736</td>\n",
       "      <td>0.322549</td>\n",
       "      <td>0.018490</td>\n",
       "      <td>0.048649</td>\n",
       "      <td>-0.000388</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>9864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000036</th>\n",
       "      <td>0.006070</td>\n",
       "      <td>-0.180905</td>\n",
       "      <td>0.004306</td>\n",
       "      <td>0.004061</td>\n",
       "      <td>0.082143</td>\n",
       "      <td>0.124391</td>\n",
       "      <td>0.241261</td>\n",
       "      <td>0.136064</td>\n",
       "      <td>-0.099635</td>\n",
       "      <td>-0.021902</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.151604</td>\n",
       "      <td>0.009547</td>\n",
       "      <td>-0.011912</td>\n",
       "      <td>-0.036209</td>\n",
       "      <td>-0.212754</td>\n",
       "      <td>0.002011</td>\n",
       "      <td>0.004901</td>\n",
       "      <td>-0.001723</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>11794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000037</th>\n",
       "      <td>0.022947</td>\n",
       "      <td>-0.058032</td>\n",
       "      <td>-0.024882</td>\n",
       "      <td>-0.007361</td>\n",
       "      <td>-0.000671</td>\n",
       "      <td>0.019875</td>\n",
       "      <td>-0.090096</td>\n",
       "      <td>-0.323505</td>\n",
       "      <td>-0.037611</td>\n",
       "      <td>-0.081639</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.334929</td>\n",
       "      <td>0.008752</td>\n",
       "      <td>-0.010288</td>\n",
       "      <td>-0.000088</td>\n",
       "      <td>0.198345</td>\n",
       "      <td>-0.003193</td>\n",
       "      <td>0.013340</td>\n",
       "      <td>-0.007567</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>10272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000038</th>\n",
       "      <td>-0.027232</td>\n",
       "      <td>0.062166</td>\n",
       "      <td>0.056516</td>\n",
       "      <td>-0.000365</td>\n",
       "      <td>-0.050971</td>\n",
       "      <td>-0.002178</td>\n",
       "      <td>0.011381</td>\n",
       "      <td>0.121958</td>\n",
       "      <td>-0.002549</td>\n",
       "      <td>-0.016742</td>\n",
       "      <td>...</td>\n",
       "      <td>0.068444</td>\n",
       "      <td>0.005345</td>\n",
       "      <td>-0.030580</td>\n",
       "      <td>0.001823</td>\n",
       "      <td>-0.042488</td>\n",
       "      <td>-0.003180</td>\n",
       "      <td>0.015015</td>\n",
       "      <td>0.004315</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>9862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000040</th>\n",
       "      <td>-0.021196</td>\n",
       "      <td>-0.042606</td>\n",
       "      <td>-0.002211</td>\n",
       "      <td>-0.018687</td>\n",
       "      <td>-0.031417</td>\n",
       "      <td>-0.012370</td>\n",
       "      <td>-0.122853</td>\n",
       "      <td>-0.278146</td>\n",
       "      <td>-0.040692</td>\n",
       "      <td>0.012698</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.144409</td>\n",
       "      <td>0.023224</td>\n",
       "      <td>0.000504</td>\n",
       "      <td>0.005328</td>\n",
       "      <td>0.202808</td>\n",
       "      <td>-0.027104</td>\n",
       "      <td>0.002004</td>\n",
       "      <td>-0.001296</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>9919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000043</th>\n",
       "      <td>0.045661</td>\n",
       "      <td>-0.104146</td>\n",
       "      <td>0.041507</td>\n",
       "      <td>-0.020404</td>\n",
       "      <td>-0.029229</td>\n",
       "      <td>0.018339</td>\n",
       "      <td>-0.098265</td>\n",
       "      <td>-0.335201</td>\n",
       "      <td>-0.117354</td>\n",
       "      <td>-0.038272</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.319510</td>\n",
       "      <td>0.030442</td>\n",
       "      <td>-0.006122</td>\n",
       "      <td>0.000407</td>\n",
       "      <td>0.248316</td>\n",
       "      <td>0.003664</td>\n",
       "      <td>0.101558</td>\n",
       "      <td>-0.005887</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>11005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000049</th>\n",
       "      <td>0.092689</td>\n",
       "      <td>-0.084680</td>\n",
       "      <td>-0.006784</td>\n",
       "      <td>-0.027318</td>\n",
       "      <td>-0.039291</td>\n",
       "      <td>0.007955</td>\n",
       "      <td>-0.171292</td>\n",
       "      <td>-0.255789</td>\n",
       "      <td>-0.042421</td>\n",
       "      <td>0.013457</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.243276</td>\n",
       "      <td>-0.003953</td>\n",
       "      <td>-0.003293</td>\n",
       "      <td>-0.011791</td>\n",
       "      <td>0.166454</td>\n",
       "      <td>0.004979</td>\n",
       "      <td>0.006143</td>\n",
       "      <td>-0.005453</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>9969</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000050</th>\n",
       "      <td>-0.016952</td>\n",
       "      <td>-0.074755</td>\n",
       "      <td>0.048621</td>\n",
       "      <td>-0.021370</td>\n",
       "      <td>-0.012469</td>\n",
       "      <td>0.000841</td>\n",
       "      <td>-0.127041</td>\n",
       "      <td>-0.266838</td>\n",
       "      <td>-0.075979</td>\n",
       "      <td>-0.106741</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.066520</td>\n",
       "      <td>0.010346</td>\n",
       "      <td>0.008772</td>\n",
       "      <td>-0.003996</td>\n",
       "      <td>0.271012</td>\n",
       "      <td>0.006833</td>\n",
       "      <td>0.027319</td>\n",
       "      <td>0.003566</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>10068,9907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000052</th>\n",
       "      <td>-0.000008</td>\n",
       "      <td>-0.057172</td>\n",
       "      <td>0.076038</td>\n",
       "      <td>0.000387</td>\n",
       "      <td>0.035516</td>\n",
       "      <td>0.042100</td>\n",
       "      <td>0.065796</td>\n",
       "      <td>-0.229365</td>\n",
       "      <td>-0.121593</td>\n",
       "      <td>0.207655</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.148836</td>\n",
       "      <td>0.035526</td>\n",
       "      <td>-0.015478</td>\n",
       "      <td>-0.003401</td>\n",
       "      <td>-0.081515</td>\n",
       "      <td>0.110851</td>\n",
       "      <td>-0.011741</td>\n",
       "      <td>-0.003127</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>10235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000053</th>\n",
       "      <td>0.009466</td>\n",
       "      <td>-0.047191</td>\n",
       "      <td>0.001708</td>\n",
       "      <td>-0.010746</td>\n",
       "      <td>-0.046002</td>\n",
       "      <td>0.009029</td>\n",
       "      <td>-0.080002</td>\n",
       "      <td>-0.216579</td>\n",
       "      <td>-0.048743</td>\n",
       "      <td>0.183879</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.112031</td>\n",
       "      <td>-0.012816</td>\n",
       "      <td>-0.003062</td>\n",
       "      <td>-0.004971</td>\n",
       "      <td>0.070585</td>\n",
       "      <td>0.000649</td>\n",
       "      <td>0.012074</td>\n",
       "      <td>0.000814</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>10014,10253,11267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000055</th>\n",
       "      <td>0.061756</td>\n",
       "      <td>-0.018183</td>\n",
       "      <td>0.122387</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>-0.003894</td>\n",
       "      <td>0.031492</td>\n",
       "      <td>0.011038</td>\n",
       "      <td>0.067939</td>\n",
       "      <td>-0.011241</td>\n",
       "      <td>-0.801744</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.089089</td>\n",
       "      <td>-0.001327</td>\n",
       "      <td>-0.000028</td>\n",
       "      <td>0.007375</td>\n",
       "      <td>0.043694</td>\n",
       "      <td>-0.008876</td>\n",
       "      <td>0.001447</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>12937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000062</th>\n",
       "      <td>-0.006338</td>\n",
       "      <td>0.022238</td>\n",
       "      <td>0.065426</td>\n",
       "      <td>0.000109</td>\n",
       "      <td>0.002146</td>\n",
       "      <td>0.019219</td>\n",
       "      <td>0.011993</td>\n",
       "      <td>0.063045</td>\n",
       "      <td>-0.008383</td>\n",
       "      <td>0.100418</td>\n",
       "      <td>...</td>\n",
       "      <td>0.107770</td>\n",
       "      <td>0.020965</td>\n",
       "      <td>0.003218</td>\n",
       "      <td>0.007524</td>\n",
       "      <td>0.088904</td>\n",
       "      <td>-0.002463</td>\n",
       "      <td>-0.008899</td>\n",
       "      <td>0.000117</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>10136,10583,10611,10801,10802,10966,10995,1118...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000063</th>\n",
       "      <td>0.098795</td>\n",
       "      <td>0.021605</td>\n",
       "      <td>0.038138</td>\n",
       "      <td>0.002267</td>\n",
       "      <td>-0.018920</td>\n",
       "      <td>0.028503</td>\n",
       "      <td>0.007167</td>\n",
       "      <td>-0.139382</td>\n",
       "      <td>-0.041067</td>\n",
       "      <td>0.129286</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.043905</td>\n",
       "      <td>0.012089</td>\n",
       "      <td>0.020392</td>\n",
       "      <td>-0.002435</td>\n",
       "      <td>0.145502</td>\n",
       "      <td>-0.010664</td>\n",
       "      <td>-0.021019</td>\n",
       "      <td>-0.002354</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>10727,12455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000066</th>\n",
       "      <td>-0.000947</td>\n",
       "      <td>-0.001950</td>\n",
       "      <td>0.111818</td>\n",
       "      <td>-0.004127</td>\n",
       "      <td>-0.043805</td>\n",
       "      <td>-0.002237</td>\n",
       "      <td>-0.056070</td>\n",
       "      <td>-0.013065</td>\n",
       "      <td>0.035485</td>\n",
       "      <td>-0.181588</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.160488</td>\n",
       "      <td>0.003556</td>\n",
       "      <td>-0.019579</td>\n",
       "      <td>-0.007897</td>\n",
       "      <td>0.101904</td>\n",
       "      <td>0.004617</td>\n",
       "      <td>0.014658</td>\n",
       "      <td>-0.001804</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>10281,10890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000069</th>\n",
       "      <td>0.017298</td>\n",
       "      <td>-0.051139</td>\n",
       "      <td>-0.205668</td>\n",
       "      <td>-0.001418</td>\n",
       "      <td>-0.084057</td>\n",
       "      <td>-0.069092</td>\n",
       "      <td>-0.022792</td>\n",
       "      <td>0.021800</td>\n",
       "      <td>-0.035827</td>\n",
       "      <td>0.166415</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.072593</td>\n",
       "      <td>0.001357</td>\n",
       "      <td>-0.001061</td>\n",
       "      <td>0.004199</td>\n",
       "      <td>0.086592</td>\n",
       "      <td>-0.000150</td>\n",
       "      <td>0.004470</td>\n",
       "      <td>-0.000942</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>9996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000071</th>\n",
       "      <td>-0.023183</td>\n",
       "      <td>-0.022540</td>\n",
       "      <td>0.091783</td>\n",
       "      <td>0.002548</td>\n",
       "      <td>0.035444</td>\n",
       "      <td>-0.031894</td>\n",
       "      <td>0.036761</td>\n",
       "      <td>0.020390</td>\n",
       "      <td>-0.000618</td>\n",
       "      <td>0.359657</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.026749</td>\n",
       "      <td>-0.003443</td>\n",
       "      <td>0.002296</td>\n",
       "      <td>-0.000439</td>\n",
       "      <td>-0.018022</td>\n",
       "      <td>-0.002475</td>\n",
       "      <td>-0.001200</td>\n",
       "      <td>-0.000593</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>12023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000073</th>\n",
       "      <td>-0.058798</td>\n",
       "      <td>0.045985</td>\n",
       "      <td>0.062197</td>\n",
       "      <td>-0.000034</td>\n",
       "      <td>0.020415</td>\n",
       "      <td>0.042837</td>\n",
       "      <td>-0.007339</td>\n",
       "      <td>-0.048396</td>\n",
       "      <td>-0.020258</td>\n",
       "      <td>-0.849610</td>\n",
       "      <td>...</td>\n",
       "      <td>0.128460</td>\n",
       "      <td>0.000203</td>\n",
       "      <td>0.002437</td>\n",
       "      <td>0.001223</td>\n",
       "      <td>-0.009079</td>\n",
       "      <td>-0.003331</td>\n",
       "      <td>0.000729</td>\n",
       "      <td>-0.001151</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>9146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000075</th>\n",
       "      <td>-0.001263</td>\n",
       "      <td>0.015386</td>\n",
       "      <td>0.055444</td>\n",
       "      <td>0.001262</td>\n",
       "      <td>0.003094</td>\n",
       "      <td>-0.029072</td>\n",
       "      <td>0.020157</td>\n",
       "      <td>-0.146970</td>\n",
       "      <td>0.002175</td>\n",
       "      <td>0.003175</td>\n",
       "      <td>...</td>\n",
       "      <td>0.164392</td>\n",
       "      <td>-0.000421</td>\n",
       "      <td>-0.001488</td>\n",
       "      <td>-0.000659</td>\n",
       "      <td>0.093662</td>\n",
       "      <td>-0.017453</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>-0.003100</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>11612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000078</th>\n",
       "      <td>-0.007932</td>\n",
       "      <td>-0.004116</td>\n",
       "      <td>-0.164399</td>\n",
       "      <td>-0.001136</td>\n",
       "      <td>-0.025260</td>\n",
       "      <td>0.012151</td>\n",
       "      <td>-0.022979</td>\n",
       "      <td>-0.088234</td>\n",
       "      <td>-0.053451</td>\n",
       "      <td>0.054101</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.033478</td>\n",
       "      <td>-0.027109</td>\n",
       "      <td>0.001039</td>\n",
       "      <td>0.015863</td>\n",
       "      <td>-0.034967</td>\n",
       "      <td>0.000603</td>\n",
       "      <td>0.013189</td>\n",
       "      <td>-0.001627</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>11800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S00000079</th>\n",
       "      <td>0.042610</td>\n",
       "      <td>-0.034332</td>\n",
       "      <td>0.025165</td>\n",
       "      <td>-0.001721</td>\n",
       "      <td>-0.004204</td>\n",
       "      <td>-0.011743</td>\n",
       "      <td>-0.018016</td>\n",
       "      <td>-0.173415</td>\n",
       "      <td>0.016935</td>\n",
       "      <td>0.087673</td>\n",
       "      <td>...</td>\n",
       "      <td>0.156711</td>\n",
       "      <td>0.002057</td>\n",
       "      <td>0.047498</td>\n",
       "      <td>0.013202</td>\n",
       "      <td>-0.017072</td>\n",
       "      <td>0.004374</td>\n",
       "      <td>-0.009348</td>\n",
       "      <td>0.004558</td>\n",
       "      <td>proteomics</td>\n",
       "      <td>10244</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               LFs1      LFs2      LFs3      LFs4      LFs5      LFs6  \\\n",
       "s_id                                                                    \n",
       "S00000006  0.051530  0.036362 -0.041906 -0.001347 -0.013508  0.001565   \n",
       "S00000012  0.034465  0.115698 -0.047133 -0.001525 -0.006179  0.015768   \n",
       "S00000013  0.111319 -0.002571 -0.073818 -0.000615 -0.066941  0.040771   \n",
       "S00000015  0.010020  0.043131 -0.000815  0.000010 -0.032613  0.026158   \n",
       "S00000016  0.099801  0.064125 -0.015871 -0.001283 -0.085869  0.002985   \n",
       "S00000019  0.064609  0.006714 -0.110616 -0.017464 -0.020428 -0.032641   \n",
       "S00000022  0.034890  0.007629 -0.122095 -0.001708 -0.076321  0.061220   \n",
       "S00000031  0.014932 -0.100028 -0.056794 -0.015458 -0.039523  0.024284   \n",
       "S00000033  0.055625 -0.149947  0.003432 -0.031034 -0.020317  0.054525   \n",
       "S00000036  0.006070 -0.180905  0.004306  0.004061  0.082143  0.124391   \n",
       "S00000037  0.022947 -0.058032 -0.024882 -0.007361 -0.000671  0.019875   \n",
       "S00000038 -0.027232  0.062166  0.056516 -0.000365 -0.050971 -0.002178   \n",
       "S00000040 -0.021196 -0.042606 -0.002211 -0.018687 -0.031417 -0.012370   \n",
       "S00000043  0.045661 -0.104146  0.041507 -0.020404 -0.029229  0.018339   \n",
       "S00000049  0.092689 -0.084680 -0.006784 -0.027318 -0.039291  0.007955   \n",
       "S00000050 -0.016952 -0.074755  0.048621 -0.021370 -0.012469  0.000841   \n",
       "S00000052 -0.000008 -0.057172  0.076038  0.000387  0.035516  0.042100   \n",
       "S00000053  0.009466 -0.047191  0.001708 -0.010746 -0.046002  0.009029   \n",
       "S00000055  0.061756 -0.018183  0.122387  0.000016 -0.003894  0.031492   \n",
       "S00000062 -0.006338  0.022238  0.065426  0.000109  0.002146  0.019219   \n",
       "S00000063  0.098795  0.021605  0.038138  0.002267 -0.018920  0.028503   \n",
       "S00000066 -0.000947 -0.001950  0.111818 -0.004127 -0.043805 -0.002237   \n",
       "S00000069  0.017298 -0.051139 -0.205668 -0.001418 -0.084057 -0.069092   \n",
       "S00000071 -0.023183 -0.022540  0.091783  0.002548  0.035444 -0.031894   \n",
       "S00000073 -0.058798  0.045985  0.062197 -0.000034  0.020415  0.042837   \n",
       "S00000075 -0.001263  0.015386  0.055444  0.001262  0.003094 -0.029072   \n",
       "S00000078 -0.007932 -0.004116 -0.164399 -0.001136 -0.025260  0.012151   \n",
       "S00000079  0.042610 -0.034332  0.025165 -0.001721 -0.004204 -0.011743   \n",
       "\n",
       "               LFs7      LFs8      LFs9     LFs10  ...     LFs23     LFs24  \\\n",
       "s_id                                               ...                       \n",
       "S00000006 -0.021276 -0.244355 -0.031964  0.057272  ... -0.191047  0.075975   \n",
       "S00000012  0.042666  0.118975  0.002245  0.035891  ...  0.153040 -0.007040   \n",
       "S00000013 -0.002893  0.120330 -0.024400  0.337955  ... -0.075491 -0.005003   \n",
       "S00000015  0.044396  0.171346  0.001049  0.003718  ... -0.071206  0.054874   \n",
       "S00000016  0.009716  0.216461 -0.005344  0.003299  ... -0.044812  0.032438   \n",
       "S00000019 -0.003856  0.083932  0.000541 -0.054995  ... -0.121061 -0.032175   \n",
       "S00000022 -0.055165 -0.110623  0.014309  0.324787  ... -0.065491 -0.016958   \n",
       "S00000031 -0.102462 -0.278852 -0.129555  0.008516  ... -0.225068  0.005913   \n",
       "S00000033 -0.158043 -0.283876 -0.124461  0.013506  ... -0.285335  0.016808   \n",
       "S00000036  0.241261  0.136064 -0.099635 -0.021902  ... -0.151604  0.009547   \n",
       "S00000037 -0.090096 -0.323505 -0.037611 -0.081639  ... -0.334929  0.008752   \n",
       "S00000038  0.011381  0.121958 -0.002549 -0.016742  ...  0.068444  0.005345   \n",
       "S00000040 -0.122853 -0.278146 -0.040692  0.012698  ... -0.144409  0.023224   \n",
       "S00000043 -0.098265 -0.335201 -0.117354 -0.038272  ... -0.319510  0.030442   \n",
       "S00000049 -0.171292 -0.255789 -0.042421  0.013457  ... -0.243276 -0.003953   \n",
       "S00000050 -0.127041 -0.266838 -0.075979 -0.106741  ... -0.066520  0.010346   \n",
       "S00000052  0.065796 -0.229365 -0.121593  0.207655  ... -0.148836  0.035526   \n",
       "S00000053 -0.080002 -0.216579 -0.048743  0.183879  ... -0.112031 -0.012816   \n",
       "S00000055  0.011038  0.067939 -0.011241 -0.801744  ... -0.089089 -0.001327   \n",
       "S00000062  0.011993  0.063045 -0.008383  0.100418  ...  0.107770  0.020965   \n",
       "S00000063  0.007167 -0.139382 -0.041067  0.129286  ... -0.043905  0.012089   \n",
       "S00000066 -0.056070 -0.013065  0.035485 -0.181588  ... -0.160488  0.003556   \n",
       "S00000069 -0.022792  0.021800 -0.035827  0.166415  ... -0.072593  0.001357   \n",
       "S00000071  0.036761  0.020390 -0.000618  0.359657  ... -0.026749 -0.003443   \n",
       "S00000073 -0.007339 -0.048396 -0.020258 -0.849610  ...  0.128460  0.000203   \n",
       "S00000075  0.020157 -0.146970  0.002175  0.003175  ...  0.164392 -0.000421   \n",
       "S00000078 -0.022979 -0.088234 -0.053451  0.054101  ... -0.033478 -0.027109   \n",
       "S00000079 -0.018016 -0.173415  0.016935  0.087673  ...  0.156711  0.002057   \n",
       "\n",
       "              LFs25     LFs26     LFs27     LFs28     LFs29     LFs30  \\\n",
       "s_id                                                                    \n",
       "S00000006 -0.006429  0.009816  0.074251  0.010391  0.000629 -0.004023   \n",
       "S00000012  0.000832  0.005247  0.003451  0.000202  0.001349 -0.002890   \n",
       "S00000013  0.001831 -0.016633 -0.083339 -0.000686  0.003954 -0.002430   \n",
       "S00000015 -0.052803 -0.000263 -0.014903  0.000827  0.036917 -0.002143   \n",
       "S00000016 -0.001762  0.001995 -0.094416 -0.005864 -0.006293 -0.005239   \n",
       "S00000019 -0.006698 -0.002780  0.095618  0.002672  0.074969 -0.004676   \n",
       "S00000022 -0.011363 -0.004861 -0.308404  0.006624 -0.031057  0.000834   \n",
       "S00000031 -0.008280 -0.010984  0.235486  0.014511  0.010286  0.005220   \n",
       "S00000033 -0.004187 -0.003736  0.322549  0.018490  0.048649 -0.000388   \n",
       "S00000036 -0.011912 -0.036209 -0.212754  0.002011  0.004901 -0.001723   \n",
       "S00000037 -0.010288 -0.000088  0.198345 -0.003193  0.013340 -0.007567   \n",
       "S00000038 -0.030580  0.001823 -0.042488 -0.003180  0.015015  0.004315   \n",
       "S00000040  0.000504  0.005328  0.202808 -0.027104  0.002004 -0.001296   \n",
       "S00000043 -0.006122  0.000407  0.248316  0.003664  0.101558 -0.005887   \n",
       "S00000049 -0.003293 -0.011791  0.166454  0.004979  0.006143 -0.005453   \n",
       "S00000050  0.008772 -0.003996  0.271012  0.006833  0.027319  0.003566   \n",
       "S00000052 -0.015478 -0.003401 -0.081515  0.110851 -0.011741 -0.003127   \n",
       "S00000053 -0.003062 -0.004971  0.070585  0.000649  0.012074  0.000814   \n",
       "S00000055 -0.000028  0.007375  0.043694 -0.008876  0.001447  0.000043   \n",
       "S00000062  0.003218  0.007524  0.088904 -0.002463 -0.008899  0.000117   \n",
       "S00000063  0.020392 -0.002435  0.145502 -0.010664 -0.021019 -0.002354   \n",
       "S00000066 -0.019579 -0.007897  0.101904  0.004617  0.014658 -0.001804   \n",
       "S00000069 -0.001061  0.004199  0.086592 -0.000150  0.004470 -0.000942   \n",
       "S00000071  0.002296 -0.000439 -0.018022 -0.002475 -0.001200 -0.000593   \n",
       "S00000073  0.002437  0.001223 -0.009079 -0.003331  0.000729 -0.001151   \n",
       "S00000075 -0.001488 -0.000659  0.093662 -0.017453  0.000033 -0.003100   \n",
       "S00000078  0.001039  0.015863 -0.034967  0.000603  0.013189 -0.001627   \n",
       "S00000079  0.047498  0.013202 -0.017072  0.004374 -0.009348  0.004558   \n",
       "\n",
       "             modality                                         feature_id  \n",
       "s_id                                                                      \n",
       "S00000006  proteomics                                              10646  \n",
       "S00000012  proteomics                                              13559  \n",
       "S00000013  proteomics                                              11290  \n",
       "S00000015  proteomics                                              10647  \n",
       "S00000016  proteomics                                         13622,9675  \n",
       "S00000019  proteomics                                              10901  \n",
       "S00000022  proteomics                                              10118  \n",
       "S00000031  proteomics                                   10040,10168,9967  \n",
       "S00000033  proteomics                                               9864  \n",
       "S00000036  proteomics                                              11794  \n",
       "S00000037  proteomics                                              10272  \n",
       "S00000038  proteomics                                               9862  \n",
       "S00000040  proteomics                                               9919  \n",
       "S00000043  proteomics                                              11005  \n",
       "S00000049  proteomics                                               9969  \n",
       "S00000050  proteomics                                         10068,9907  \n",
       "S00000052  proteomics                                              10235  \n",
       "S00000053  proteomics                                  10014,10253,11267  \n",
       "S00000055  proteomics                                              12937  \n",
       "S00000062  proteomics  10136,10583,10611,10801,10802,10966,10995,1118...  \n",
       "S00000063  proteomics                                        10727,12455  \n",
       "S00000066  proteomics                                        10281,10890  \n",
       "S00000069  proteomics                                               9996  \n",
       "S00000071  proteomics                                              12023  \n",
       "S00000073  proteomics                                               9146  \n",
       "S00000075  proteomics                                              11612  \n",
       "S00000078  proteomics                                              11800  \n",
       "S00000079  proteomics                                              10244  \n",
       "\n",
       "[28 rows x 32 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mechanism_matching.bind_wide_results(\n",
    "    sbml_dfs,\n",
    "    mofa_df,\n",
    "    \"mudata_varm_results\",\n",
    "    ontologies = {\"uniprot\", \"ensembl_gene\"},\n",
    "    dogmatic = False,\n",
    "    verbose = True\n",
    ")\n",
    "\n",
    "sbml_dfs.species_data[\"mudata_varm_results\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:napistu.network.net_create:Organizing all network nodes (compartmentalized species and reactions)\n",
      "INFO:napistu.network.net_create:Formatting edges as a regulatory graph\n",
      "INFO:napistu.network.net_create:Formatting 250 reactions species as tiered edges.\n",
      "INFO:napistu.network.net_create:Adding additional attributes to edges, e.g., # of children and parents.\n",
      "INFO:napistu.network.net_create:Done preparing regulatory graph\n",
      "INFO:napistu.network.net_create:Adding reversibility and other meta-data from reactions_data\n",
      "INFO:napistu.network.net_create:No reactions annotations provided in \"graph_attrs\"; returning None\n",
      "INFO:napistu.network.net_create:Creating reverse reactions for reversible reactions on a directed graph\n",
      "INFO:napistu.network.net_create:Formatting cpr_graph output\n",
      "INFO:napistu.network.net_create:Adding meta-data from species_data\n",
      "INFO:napistu.network.net_create:Adding new attribute LFs5 to vertices\n",
      "INFO:napistu.network.net_create:Adding new attribute effect_case to vertices\n"
     ]
    }
   ],
   "source": [
    "from napistu.network import net_create\n",
    "\n",
    "# now we can pass these species_data attributes to the graph\n",
    "\n",
    "reaction_graph_attrs = {\n",
    "    \"species\": {\n",
    "        \"LFs5\": {\n",
    "            \"table\": \"mudata_varm_results\",\n",
    "            \"variable\": \"LFs5\",\n",
    "            \"trans\": \"identity\",\n",
    "        },\n",
    "        \"effect_case\": {\n",
    "            \"table\": \"var_level_results\",\n",
    "            \"variable\": \"effect_case\",\n",
    "            \"trans\": \"identity\",\n",
    "        },\n",
    "    },\n",
    "}\n",
    "\n",
    "cpr_graph = net_create.create_cpr_graph(\n",
    "    sbml_dfs,\n",
    "    directed=True,\n",
    "    graph_type=\"regulatory\"\n",
    ")\n",
    "\n",
    "# add species attributes\n",
    "# TO DO - this is definitely not a utility function\n",
    "graph_w_annotations = net_create._add_graph_species_attribute(\n",
    "    cpr_graph,\n",
    "    sbml_dfs,\n",
    "    species_graph_attrs = reaction_graph_attrs,\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_cb916 th {\n",
       "  color: limegreen;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_cb916\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th id=\"T_cb916_level0_col0\" class=\"col_heading level0 col0\" >name</th>\n",
       "      <th id=\"T_cb916_level0_col1\" class=\"col_heading level0 col1\" >node_name</th>\n",
       "      <th id=\"T_cb916_level0_col2\" class=\"col_heading level0 col2\" >node_type</th>\n",
       "      <th id=\"T_cb916_level0_col3\" class=\"col_heading level0 col3\" >LFs5</th>\n",
       "      <th id=\"T_cb916_level0_col4\" class=\"col_heading level0 col4\" >effect_case</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th class=\"index_name level0\" >vertex ID</th>\n",
       "      <th class=\"blank col0\" >&nbsp;</th>\n",
       "      <th class=\"blank col1\" >&nbsp;</th>\n",
       "      <th class=\"blank col2\" >&nbsp;</th>\n",
       "      <th class=\"blank col3\" >&nbsp;</th>\n",
       "      <th class=\"blank col4\" >&nbsp;</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_cb916_level0_row0\" class=\"row_heading level0 row0\" >26</th>\n",
       "      <td id=\"T_cb916_row0_col0\" class=\"data row0 col0\" >species_372469</td>\n",
       "      <td id=\"T_cb916_row0_col1\" class=\"data row0 col1\" >SLC25A12,13 [mitochondrial inner membrane]</td>\n",
       "      <td id=\"T_cb916_row0_col2\" class=\"data row0 col2\" >species</td>\n",
       "      <td id=\"T_cb916_row0_col3\" class=\"data row0 col3\" >-0.086</td>\n",
       "      <td id=\"T_cb916_row0_col4\" class=\"data row0 col4\" >0.021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_cb916_level0_row1\" class=\"row_heading level0 row1\" >96</th>\n",
       "      <td id=\"T_cb916_row1_col0\" class=\"data row1 col0\" >species_6798333</td>\n",
       "      <td id=\"T_cb916_row1_col1\" class=\"data row1 col1\" >BPGM dimer [cytosol]</td>\n",
       "      <td id=\"T_cb916_row1_col2\" class=\"data row1 col2\" >species</td>\n",
       "      <td id=\"T_cb916_row1_col3\" class=\"data row1 col3\" >-0.084</td>\n",
       "      <td id=\"T_cb916_row1_col4\" class=\"data row1 col4\" >-0.031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_cb916_level0_row2\" class=\"row_heading level0 row2\" >36</th>\n",
       "      <td id=\"T_cb916_row2_col0\" class=\"data row2 col0\" >species_70499</td>\n",
       "      <td id=\"T_cb916_row2_col1\" class=\"data row2 col1\" >pyruvate carboxylase holoenzyme [mitochondrial matrix]</td>\n",
       "      <td id=\"T_cb916_row2_col2\" class=\"data row2 col2\" >species</td>\n",
       "      <td id=\"T_cb916_row2_col3\" class=\"data row2 col3\" >-0.076</td>\n",
       "      <td id=\"T_cb916_row2_col4\" class=\"data row2 col4\" >0.203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_cb916_level0_row3\" class=\"row_heading level0 row3\" >18</th>\n",
       "      <td id=\"T_cb916_row3_col0\" class=\"data row3 col0\" >species_376856</td>\n",
       "      <td id=\"T_cb916_row3_col1\" class=\"data row3 col1\" >SLC25A11 homodimer [mitochondrial inner membrane]</td>\n",
       "      <td id=\"T_cb916_row3_col2\" class=\"data row3 col2\" >species</td>\n",
       "      <td id=\"T_cb916_row3_col3\" class=\"data row3 col3\" >-0.067</td>\n",
       "      <td id=\"T_cb916_row3_col4\" class=\"data row3 col4\" >0.291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_cb916_level0_row4\" class=\"row_heading level0 row4\" >58</th>\n",
       "      <td id=\"T_cb916_row4_col0\" class=\"data row4 col0\" >species_70594</td>\n",
       "      <td id=\"T_cb916_row4_col1\" class=\"data row4 col1\" >GOT2 dimer [mitochondrial matrix]</td>\n",
       "      <td id=\"T_cb916_row4_col2\" class=\"data row4 col2\" >species</td>\n",
       "      <td id=\"T_cb916_row4_col3\" class=\"data row4 col3\" >-0.051</td>\n",
       "      <td id=\"T_cb916_row4_col4\" class=\"data row4 col4\" >0.043</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x16ab15ad0>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from napistu import utils as napistu_utils\n",
    "\n",
    "napistu_utils.style_df(graph_w_annotations.get_vertex_dataframe().sort_values(\"LFs5\").head(5))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
